{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import csv\n",
    "import matplotlib.cm as cm\n",
    "import seaborn as sns\n",
    "from matplotlib import gridspec\n",
    "import matplotlib.animation as manimation\n",
    "import time\n",
    "import scipy.io as sio\n",
    "from random import shuffle\n",
    "import time\n",
    "from collections import Counter\n",
    "import sklearn.cluster as cluster\n",
    "import time\n",
    "import hdbscan\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "import os\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from pathlib import Path\n",
    "\n",
    "#import plotly. as py\n",
    "#import plotly.graph_objs as go\n",
    "font = {'family' : 'DejaVu Sans',\n",
    "        'weight' : 'regular',\n",
    "        'size'   : 16}\n",
    "\n",
    "plt.rc('font', **font)\n",
    "plt.rc('lines',lw=2)\n",
    "\n",
    "# dereks function for optimal leaf clustering\n",
    "def _optimal_order(data, **kwargs):\n",
    "    \"\"\" Optimal leaf ordering\n",
    "        **kwargs passed to pdist e.g. metric='correlation'\n",
    "    \"\"\"\n",
    "    d = pdist(data, **kwargs)\n",
    "    link = linkage(d, method='average')\n",
    "    optimal_order = optimal_leaf_ordering(link, d)\n",
    "    return optimal_order\n",
    "\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "#from fastcluster import linkage\n",
    "#from polo import optimal_leaf_ordering\n",
    "\n",
    "from matplotlib_venn import venn2, venn2_circles, venn3, venn3_circles\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniqueValsdf(df):\n",
    "    uniquelist = list(df.iloc[:,0].dropna().unique())\n",
    "    for col in range(1,len(df.columns)):\n",
    "        uniquelist = uniquelist+list(df.iloc[:,col].dropna().unique())\n",
    "    return(uniquelist)\n",
    "\n",
    "def COGmapper(df):\n",
    "    # gets ctab formatted df into mappable one\n",
    "    mapdf = pd.DataFrame(index=COGtab.set_index('COG').index,columns=df.columns)\n",
    "    for names in df.columns:\n",
    "        mapdf.loc[df[names].dropna(how='any').values,names] = 1\n",
    "    return(mapdf)\n",
    "\n",
    "def get_pfam_description(pf_df):\n",
    "    pf_df['pfam_desc'] = ''\n",
    "    listofnans = []\n",
    "    for i in range(0,len(pf_df)):\n",
    "        pf = pf_df.loc[i,'pfam_id']\n",
    "        if len(pfam_lookup[pfam_lookup['pfam_id']==pf][3])>0:\n",
    "            pf_df.loc[i,'pfam_desc'] = pfam_lookup[pfam_lookup['pfam_id']==pf][3].values[0]\n",
    "        else:\n",
    "            listofnans = listofnans+[pf]\n",
    "    return(pf_df,listofnans)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load relevant dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Monomethylamine methyltransferase MtmB'], dtype=object)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cladeDF = pd.read_table('MAGs/extractclades/ObsidianPool_selected_groups.tsv')\n",
    "cladeDF_tot = pd.read_table('MAGs/extractclades/ObsidianPool_allclades.tsv')\n",
    "\n",
    "# COG\n",
    "COGcat = pd.read_table('COGreffiles/fun2003-2014.tab')\n",
    "COGtab = pd.read_table('COGreffiles/cognames2003-2014.tab')\n",
    "COGtab['NameNum'] = COGtab['name']+' - '+COGtab['COG']\n",
    "COGtab['FuncCat'] = [COGcat[COGcat['# Code']==i[0]]['Name'].values[0] for i in COGtab['func']]\n",
    "#COGtab.FuncCat.value_counts()\n",
    "cog_hitdf = pd.read_csv('writeup/COG30_specific_enzymes_round2.table',sep='\\t')\n",
    "\n",
    "#KEGG\n",
    "moddf = pd.read_pickle('KEGG_module_list')\n",
    "KEGGmoddf = pd.read_pickle('KEGG_modules_df')\n",
    "\n",
    "#pfam\n",
    "pfam_lookup = pd.read_table('MAGs/Analysis/pfam_lookuptable',header=None)\n",
    "pfam_lookup['pfam_id'] = 'pfam'+pfam_lookup[0].str[2:7]\n",
    "pfdf_tot = pd.read_csv('MAGs/Analysis/ObsidianAll_pfam.table',sep='\\t')\n",
    "pfdf_tot[pfdf_tot['pfam_desc'].str.contains('onometh')]['pfam_desc'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load gene lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "mancur3 = pd.read_table('MAGs/Analysis/manually_curated list_of_COGs.txt',header=None)\n",
    "mancur_methano = pd.read_table('MAGs/Analysis/manually_curated list_of_COGs_subset.txt',header=None)\n",
    "mancur3 = list(mancur3[0].values)\n",
    "mancur4 = mancur3[20:49]\n",
    "mancur_methano = list(mancur_methano[0].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load threshold dfs (made in 201804_Obsidian_metabolic_reconstruction.ipynb) as these take a while to build"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap_thr = pd.read_csv('MAGs/Analysis/COGplots/COG_percent_ID.table',sep='\\t')#.set_index('NameNum',inplace=True)\n",
    "cmap_clade_thr = pd.read_csv('MAGs/Analysis/COGplots/COG_percent_ID_clade.table',sep='\\t')#.set_index('NameNum',inplace=True)\n",
    "cmap_clade_thr.set_index('NameNum',inplace=True)\n",
    "cmap_thr.set_index('COG',inplace=True)\n",
    "cmap_mcrcodh_thr = pd.read_table('MAGs/methanogens/COG_percent_ID_methanogens.table')\n",
    "cmap_mcrcodh_thr.set_index('COG',inplace=True)\n",
    "cmap_mcrcodh_thr_long = cmap_mcrcodh_thr.copy()\n",
    "cmap_mcrcodh_thr_long.index =  COGtab.set_index('COG').loc[cmap_mcrcodh_thr.index]['NameNum']\n",
    "#cmap_mcrcodh_thr_long.to_csv('MAGs/methanogens/COG_percent_ID_methanogens_withname.table',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "mypath = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'\n",
    "foldernames = [f for f in listdir(mypath) if ('Knumbers') not in f and ('NN_MAGs') not in f]\n",
    "\n",
    "foldername = pd.DataFrame(index=range(len(foldernames)),columns=['folder_name','number','genome'])\n",
    "foldername['folder_name'] = [f for f in foldernames]\n",
    "foldername['number'] = [f[-10:] for f in foldernames]\n",
    "foldername['genome'] = [f[0:15] for f in foldernames]\n",
    "ktab = pd.DataFrame(index=range(10000),columns=list(foldername['genome']))\n",
    "ctab = pd.DataFrame(index=range(10000),columns=list(foldername['genome']))\n",
    "ptab = pd.DataFrame(index=range(10000),columns=list(foldername['genome']))\n",
    "ctab_thr = pd.DataFrame(index=range(10000),columns=list(foldername['genome'])) # threshold hit value\n",
    "\n",
    "# make KO-list table and write to txt file in Knumbers folder\n",
    "for i in range(0,len(foldername)):\n",
    "    kofile = mypath+foldername.loc[i,'folder_name']+'/IMG_Data/'+foldername.loc[i,'number']+'/'+foldername.loc[i,'number']+'.ko.tab.txt'        \n",
    "    ktab[foldername.loc[i,'genome']] = pd.read_csv(kofile,'\\t')['ko_id'].str[3:]\n",
    "    ktab[foldername.loc[i,'genome']].to_csv(mypath+'Knumbers/'+foldername.loc[i,'genome']+'.txt',sep='\\t')\n",
    "    # make COG-list table \n",
    "    cofile = mypath+foldername.loc[i,'folder_name']+'/IMG_Data/'+foldername.loc[i,'number']+'/'+foldername.loc[i,'number']+'.cog.tab.txt'        \n",
    "    ctab[foldername.loc[i,'genome']] = pd.read_csv(cofile,'\\t')['cog_id']\n",
    "    ctab_thr[foldername.loc[i,'genome']] = pd.read_csv(cofile,'\\t')['percent_identity']\n",
    "    #ctab[foldername.loc[i,'genome']].to_csv(mypath+'Knumbers/'+foldername.loc[i,'genome']+'.txt',sep='\\t')\n",
    "    # make pfam-list table \n",
    "    pffile = mypath+foldername.loc[i,'folder_name']+'/IMG_Data/'+foldername.loc[i,'number']+'/'+foldername.loc[i,'number']+'.pfam.tab.txt'        \n",
    "    pfamtemp =  pd.read_csv(pffile,'\\t')[['pfam_id','pfam_name']]\n",
    "    pfamtemp['namenum'] = pfamtemp['pfam_id']+', '+pfamtemp['pfam_name']\n",
    "    ptab[foldername.loc[i,'genome']] = pfamtemp['namenum']\n",
    "\n",
    "\n",
    "# COG dfs with different percent identity threshold values \n",
    "ctab_30 = ctab[ctab_thr>30]\n",
    "ctab_40 = ctab[ctab_thr>40]\n",
    "ctab_50 = ctab[ctab_thr>50]\n",
    "ctab_60 = ctab[ctab_thr>60]\n",
    "ctab_70 = ctab[ctab_thr>70]\n",
    "ctab_80 = ctab[ctab_thr>80]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make one ktab for all clades, including my own Obsidian Pool genomes\n",
    "ktab_clade = pd.DataFrame(index=range(10000),columns=list(cladeDF['Bin_ID']))\n",
    "ctab_clade = pd.DataFrame(index=range(10000),columns=list(cladeDF['Bin_ID']))\n",
    "ptab_clade = pd.DataFrame(index=range(10000),columns=list(cladeDF['Bin_ID']))\n",
    "ctab_clade_thr = pd.DataFrame(index=range(10000),columns=list(cladeDF['Bin_ID']))\n",
    "\n",
    "for bins in ktab_clade.columns:\n",
    "    if bins[0:3]=='Obs':\n",
    "        # this assumes the ktab and ctab dfs have been made \n",
    "        ktab_clade[bins] = ktab[bins]\n",
    "        ctab_clade[bins] = ctab[bins]\n",
    "        ptab_clade[bins] = ptab[bins]\n",
    "        ctab_clade_thr[bins] = ctab_thr[bins]\n",
    "    elif bins[0:3]=='UBA':\n",
    "        ktab_clade.drop(bins,axis=1,inplace=True)\n",
    "        ctab_clade.drop(bins,axis=1,inplace=True)\n",
    "        ptab_clade.drop(bins,axis=1,inplace=True)\n",
    "        ctab_clade_thr.drop(bins,axis=1,inplace=True)\n",
    "    else:\n",
    "        # go to extractclade \n",
    "        path = 'MAGs/extractclades/'\n",
    "        subfolder = cladeDF[cladeDF['Bin_ID']==bins]['Group'].values[0]\n",
    "        kofile = path+subfolder+'/'+bins+'/'+bins+'.ko.tab.txt'\n",
    "        cofile = path+subfolder+'/'+bins+'/'+bins+'.cog.tab.txt'\n",
    "        pffile = path+subfolder+'/'+bins+'/'+bins+'.pfam.tab.txt'\n",
    "        ktab_clade[bins] = pd.read_csv(kofile,'\\t')['ko_id'].str[3:]\n",
    "        ktab_clade[bins].to_csv(path+'Knumbers/'+bins+'.txt',sep='\\t')\n",
    "        ctab_clade[bins] = pd.read_csv(cofile,'\\t')['cog_id']\n",
    "        ctab_clade_thr[bins] = pd.read_csv(cofile,'\\t')['percent_identity']\n",
    "        \n",
    "        pfamtemp =  pd.read_csv(pffile,'\\t')[['pfam_id','pfam_name']]\n",
    "        pfamtemp['namenum'] = pfamtemp['pfam_id']+', '+pfamtemp['pfam_name']\n",
    "        ptab_clade[bins] = pfamtemp['namenum']\n",
    "\n",
    "# COG dfs with different percent identity threshold values \n",
    "ctab_clade_30 = ctab_clade[ctab_clade_thr>30]\n",
    "ctab_clade_40 = ctab_clade[ctab_clade_thr>40]\n",
    "ctab_clade_50 = ctab_clade[ctab_clade_thr>50]\n",
    "ctab_clade_60 = ctab_clade[ctab_clade_thr>60]\n",
    "ctab_clade_70 = ctab_clade[ctab_clade_thr>70]\n",
    "ctab_clade_80 = ctab_clade[ctab_clade_thr>80]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load metadata for all 111 Obsidian MAGs\n",
    "lookup_obsidian = pd.read_table('Tree_Obsidian/FredV2/ObsV2/metadata/ObsidianPool_UNI56_allbins_mafftBMGE_nofilteringnoclustering - Sheet1.tsv')\n",
    "# add CRISPR column\n",
    "lookup_obsidian.set_index('Obsidian Bins ',inplace=True)\n",
    "lookup_obsidian['CRISPR'] = ''\n",
    "clu = cladeDF_tot[cladeDF_tot.Bin_ID.str.contains('Obs')].set_index('Bin_ID')\n",
    "\n",
    "for name in lookup_obsidian.index:\n",
    "    num = lookup_obsidian.loc[name,'IMG_ID'].astype(str)\n",
    "    path = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'+name+'_IMG_'+num+'/IMG_Data/'+num+'/'+num+'.crispr.txt'\n",
    "    file = Path(path)\n",
    "    if file.is_file():\n",
    "        lookup_obsidian.loc[name,'CRISPR'] = 'Yes'\n",
    "    else:\n",
    "        lookup_obsidian.loc[name,'CRISPR'] = 'No'\n",
    "    \n",
    "    if name in clu.index:\n",
    "        lookup_obsidian.loc[name,'Clade'] = clu.loc[name,'Clade_name']\n",
    "    else:\n",
    "        lookup_obsidian.loc[name,'Clade'] = 'other'\n",
    "\n",
    "lookup_obsidian['Effective phylum'] = lookup_obsidian['Phylum ']\n",
    "for binn in lookup_obsidian[lookup_obsidian['Effective phylum']=='unclassified'].index:\n",
    "    if ('Thermopro' in lookup_obsidian.loc[binn,'Clade']) | ('Desulfuro' in lookup_obsidian.loc[binn,'Clade']) | ('Sulfolob' in lookup_obsidian.loc[binn,'Clade']):   \n",
    "        lookup_obsidian.loc[binn, 'Effective phylum'] = 'Crenarchaeota'\n",
    "    elif ('Thaum' in lookup_obsidian.loc[binn,'Clade']):\n",
    "        lookup_obsidian.loc[binn, 'Effective phylum'] = 'Thaumarchaeota'\n",
    "    elif ('Eurya' in lookup_obsidian.loc[binn,'Clade']):\n",
    "        lookup_obsidian.loc[binn, 'Effective phylum'] = 'Euryarchaeota'\n",
    "    else:\n",
    "        lookup_obsidian.loc[binn, 'Effective phylum'] = lookup_obsidian.loc[binn,'Clade']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = COGmapper(ctab)\n",
    "cmap_30 = COGmapper(ctab_30)\n",
    "cmap_40 = COGmapper(ctab_40)\n",
    "cmap_50 = COGmapper(ctab_50)\n",
    "cmap_60 = COGmapper(ctab_60)\n",
    "cmap_70 = COGmapper(ctab_70)\n",
    "cmap_tot = cmap.fillna(0)+cmap_30.fillna(0)+cmap_40.fillna(0)+cmap_50.fillna(0)+cmap_60.fillna(0)+cmap_70.fillna(0)\n",
    "cmap_clade_30 = COGmapper(ctab_clade_30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all the clades in the tree \n",
    "cols = [0,'Class','Order','Family','Genus','Species','IMG_ID']\n",
    "lookup_clades = pd.read_table('Tree_Obsidian/TreesV3/allarcheaMarch2018_V3/archaea_March2018.lookup',sep='|',header=None)\n",
    "lookupGC_cla = pd.read_table('Tree_Obsidian/TreesV3/allarcheaMarch2018_V3/archaea_March2018.GCLen',header=None).set_index(0)\n",
    "keys = lookup_clades.columns\n",
    "lut = dict(zip(keys,cols))\n",
    "lookup_clades = lookup_clades.rename(columns=lut)\n",
    "lc = pd.DataFrame(lookup_clades[0].str.split('\\t',1).tolist(),columns =['IMG','Phylum'])\n",
    "lookup_clades = lc.join(lookup_clades)\n",
    "lookup_clades = lookup_clades.drop([0], axis=1)\n",
    "lookup_clades = lookup_clades.set_index('IMG',drop=True)\n",
    "lookup_clades = lookup_clades.join(lookupGC_cla,how='inner')\n",
    "lookup_clades = lookup_clades.rename(columns={1:'GC',2:'length'})\n",
    "lookup_clades['IMG_ID'] = lookup_clades['IMG_ID'].astype(str).str[0:-2]\n",
    "cladeDF_true = cladeDF[~cladeDF['Bin_ID'].str.contains('UBA')]\n",
    "cladeDF_IMG = cladeDF_true[~(cladeDF_true['Bin_ID'].str.contains('Obs'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "desu_clade = cladeDF_true[cladeDF_true['Group'].str.contains('Desu')]['Bin_ID']\n",
    "bathy_clade = cladeDF_true[cladeDF_true['Group'].str.contains('Bathy')]['Bin_ID']\n",
    "verstr_clade = cladeDF_true[cladeDF_true['Group'].str.contains('Vers')]['Bin_ID']\n",
    "newp_clade = cladeDF_true[cladeDF_true['Group'].str.contains('New')]['Bin_ID']\n",
    "#cmap_clade_30_long = cmap_clade_30.copy()\n",
    "#cmap_clade_30_long.index =  COGtab.set_index('COG').loc[cmap_clade_30.index]['NameNum']\n",
    "finalcolorder_methano = list(verstr_clade[8:11])+list(verstr_clade[0:1])+list(verstr_clade[2:8])+list(desu_clade[11:13])+list(desu_clade[14:22])+list(desu_clade[23:24])+list(bathy_clade)\n",
    "eury = list(cmap_mcrcodh_thr_long.T.index[7:9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap_30_long = cmap_30.copy()\n",
    "cmap_30_long.index =  COGtab.set_index('COG').loc[cmap_30.index]['NameNum']\n",
    "cmap_long = cmap.copy()\n",
    "cmap_long.index =  COGtab.set_index('COG').loc[cmap.index]['NameNum']\n",
    "cmap_thr_long = cmap_thr.copy()\n",
    "cmap_thr_long.index =  COGtab.set_index('COG').loc[cmap_thr.index]['NameNum']\n",
    "cols = list(cladeDF_tot.Bin_ID)\n",
    "cmap_30_archaea = cmap_30_long.loc[:,cols].dropna(how='all')\n",
    "cmap_archaea = cmap_long.loc[:,cols].dropna(how='all')\n",
    "cmap_thr_archaea = cmap_thr_long.loc[:,cols].dropna(how='all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## analysis section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COG= 16, KEGG= 6, pfam= 6\n"
     ]
    }
   ],
   "source": [
    "# search for specific name in KEGG, COG, pfam of OP bins\n",
    "#def searchGenes(bin_name,keyword):\n",
    "keyword = 'F420'\n",
    "bin_name = 'Obs2_genome_008'\n",
    "\n",
    "\n",
    "img_id = lookup_obsidian[lookup_obsidian.index==bin_name]['IMG_ID'].values[0].astype(str)\n",
    "folder = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'+bin_name+'_IMG_'+img_id+'/IMG_Data/'+img_id+'/'        \n",
    "cofile = folder+img_id+'.cog.tab.txt'\n",
    "kofile = folder+img_id+'.ko.tab.txt'\n",
    "pffile = folder+img_id+'.pfam.tab.txt'\n",
    "\n",
    "codf = pd.read_table(cofile,'\\t') #'cog_name','cog_id','percent_identity'\n",
    "kodf = pd.read_table(kofile,'\\t') #'ko_name','ko_id','percent_identity'\n",
    "kodf = kodf[~kodf['ko_name'].isnull()]\n",
    "pfdf = pd.read_table(pffile,'\\t') #'pfam_name','pfam_id'\n",
    "    \n",
    "codf_s = codf[codf['cog_name'].str.contains(keyword)]\n",
    "kodf_s = kodf[kodf['ko_name'].str.contains(keyword)]\n",
    "pfdf_s = pfdf[pfdf['pfam_name'].str.contains(keyword)]\n",
    "print('COG= '+str(len(codf_s))+', KEGG= '+str(len(kodf_s))+', pfam= '+str(len(pfdf_s)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Coenzyme F420-reducing hydrogenase, delta subunit',\n",
       " 'F420-dependent methylenetetrahydromethanopterin dehydrogenase',\n",
       " 'Uncharacterized protein, related to F420-0:gamma-glutamyl ligase',\n",
       " 'Coenzyme F420-reducing hydrogenase, gamma subunit',\n",
       " 'Coenzyme F420-reducing hydrogenase, alpha subunit',\n",
       " 'Coenzyme F420-reducing hydrogenase, beta subunit',\n",
       " ' Creatinine amidohydrolase/Fe(II)-dependent formamide hydrolase involved in riboflavin and F420 biosynthesis',\n",
       " 'F420-0:Gamma-glutamyl ligase (F420 biosynthesis)']"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(codf_s['cog_name'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FAD synthetase']"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(kodf_s['ko_name'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FAD_binding_2',\n",
       " 'FAD_binding_4',\n",
       " 'FAD-oxidase_C',\n",
       " 'FAD_oxidored',\n",
       " 'FAD_binding_6']"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(pfdf_s['pfam_name'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Obs2_genome_000', 'Obs2_genome_001', 'Obs2_genome_005',\n",
       "       'Obs2_genome_007', 'Obs2_genome_008', 'Obs2_genome_009',\n",
       "       'Obs2_genome_010', 'Obs2_genome_011', 'Obs2_genome_012',\n",
       "       'Obs2_genome_013', 'Obs2_genome_014', 'Obs2_genome_015',\n",
       "       'Obs2_genome_016', 'Obs2_genome_017', 'Obs2_genome_018',\n",
       "       'Obs2_genome_020', 'Obs2_genome_021', 'Obs2_genome_022',\n",
       "       'Obs2_genome_023', 'Obs2_genome_025', 'Obs2_genome_027',\n",
       "       'Obs2_genome_031', 'Obs2_genome_034', 'Obs3_genome_035',\n",
       "       'Obs3_genome_036', 'Obs3_genome_037', 'Obs3_genome_038',\n",
       "       'Obs3_genome_039', 'Obs3_genome_040', 'Obs3_genome_041',\n",
       "       'Obs3_genome_046', 'Obs3_genome_047', 'Obs3_genome_048',\n",
       "       'Obs3_genome_053', 'Obs4_genome_057', 'Obs4_genome_058',\n",
       "       'Obs4_genome_061', 'Obs4_genome_062', 'Obs4_genome_063',\n",
       "       'Obs4_genome_065', 'Obs4_genome_066', 'Obs4_genome_067',\n",
       "       'Obs4_genome_068', 'Obs4_genome_069', 'Obs4_genome_070',\n",
       "       'Obs4_genome_071', 'Obs4_genome_072', 'Obs4_genome_073',\n",
       "       'Obs4_genome_074', 'Obs4_genome_075', 'Obs4_genome_076',\n",
       "       'Obs5_genome_078', 'Obs5_genome_080', 'Obs5_genome_082',\n",
       "       'Obs5_genome_083', 'Obs5_genome_084', 'Obs5_genome_088',\n",
       "       'Obs5_genome_090', 'Obs5_genome_092', 'Obs6_genome_096',\n",
       "       'Obs6_genome_097', 'Obs6_genome_099', 'Obs6_genome_100',\n",
       "       'Obs6_genome_101', 'Obs6_genome_104', 'Obs6_genome_109'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# search in OP bins all\n",
    "keyword_all = 'FAD-oxidase_C'# needs to be more specific\n",
    "np.sort(pfdf_tot[pfdf_tot['pfam_name'].str.contains(keyword_all)].bin.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Obs2_genome_000', 'Obs2_genome_001', 'Obs2_genome_005',\n",
       "       'Obs2_genome_007', 'Obs2_genome_008', 'Obs2_genome_009',\n",
       "       'Obs2_genome_010', 'Obs2_genome_011', 'Obs2_genome_012',\n",
       "       'Obs2_genome_013', 'Obs2_genome_014', 'Obs2_genome_015',\n",
       "       'Obs2_genome_016', 'Obs2_genome_017', 'Obs2_genome_018',\n",
       "       'Obs2_genome_020', 'Obs2_genome_021', 'Obs2_genome_022',\n",
       "       'Obs2_genome_023', 'Obs2_genome_025', 'Obs2_genome_027',\n",
       "       'Obs2_genome_031', 'Obs2_genome_034', 'Obs3_genome_035',\n",
       "       'Obs3_genome_036', 'Obs3_genome_037', 'Obs3_genome_039',\n",
       "       'Obs3_genome_040', 'Obs3_genome_041', 'Obs3_genome_046',\n",
       "       'Obs3_genome_047', 'Obs3_genome_048', 'Obs3_genome_053',\n",
       "       'Obs3_genome_055', 'Obs4_genome_057', 'Obs4_genome_058',\n",
       "       'Obs4_genome_061', 'Obs4_genome_062', 'Obs4_genome_063',\n",
       "       'Obs4_genome_065', 'Obs4_genome_066', 'Obs4_genome_067',\n",
       "       'Obs4_genome_068', 'Obs4_genome_069', 'Obs4_genome_070',\n",
       "       'Obs4_genome_071', 'Obs4_genome_072', 'Obs4_genome_073',\n",
       "       'Obs4_genome_074', 'Obs4_genome_075', 'Obs4_genome_076',\n",
       "       'Obs5_genome_078', 'Obs5_genome_080', 'Obs5_genome_082',\n",
       "       'Obs5_genome_083', 'Obs5_genome_084', 'Obs5_genome_088',\n",
       "       'Obs5_genome_090', 'Obs5_genome_092', 'Obs6_genome_096',\n",
       "       'Obs6_genome_099', 'Obs6_genome_100', 'Obs6_genome_101',\n",
       "       'Obs6_genome_104', 'Obs6_genome_109'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmap_long[cmap_long.index.str.contains('FAD/FMN')].dropna(axis=1).columns.sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_oid</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>percent_identity</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subj_start</th>\n",
       "      <th>subj_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>cog_id</th>\n",
       "      <th>cog_name</th>\n",
       "      <th>cog_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>2771094374</td>\n",
       "      <td>620</td>\n",
       "      <td>31.58</td>\n",
       "      <td>209</td>\n",
       "      <td>468</td>\n",
       "      <td>2</td>\n",
       "      <td>261</td>\n",
       "      <td>3.000000e-40</td>\n",
       "      <td>160.0</td>\n",
       "      <td>COG0175</td>\n",
       "      <td>3'-phosphoadenosine 5'-phosphosulfate sulfotra...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>2771094523</td>\n",
       "      <td>478</td>\n",
       "      <td>45.63</td>\n",
       "      <td>3</td>\n",
       "      <td>460</td>\n",
       "      <td>16</td>\n",
       "      <td>484</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>481.0</td>\n",
       "      <td>COG2509</td>\n",
       "      <td>Uncharacterized FAD-dependent dehydrogenase</td>\n",
       "      <td>486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>2771094825</td>\n",
       "      <td>481</td>\n",
       "      <td>26.64</td>\n",
       "      <td>13</td>\n",
       "      <td>475</td>\n",
       "      <td>1</td>\n",
       "      <td>456</td>\n",
       "      <td>2.000000e-57</td>\n",
       "      <td>217.0</td>\n",
       "      <td>COG0277</td>\n",
       "      <td>FAD/FMN-containing dehydrogenase</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>2771094967</td>\n",
       "      <td>422</td>\n",
       "      <td>31.37</td>\n",
       "      <td>195</td>\n",
       "      <td>387</td>\n",
       "      <td>15</td>\n",
       "      <td>214</td>\n",
       "      <td>6.000000e-33</td>\n",
       "      <td>135.0</td>\n",
       "      <td>COG0175</td>\n",
       "      <td>3'-phosphoadenosine 5'-phosphosulfate sulfotra...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>2771095575</td>\n",
       "      <td>463</td>\n",
       "      <td>26.07</td>\n",
       "      <td>9</td>\n",
       "      <td>456</td>\n",
       "      <td>1</td>\n",
       "      <td>452</td>\n",
       "      <td>7.000000e-43</td>\n",
       "      <td>168.0</td>\n",
       "      <td>COG0277</td>\n",
       "      <td>FAD/FMN-containing dehydrogenase</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>2771095731</td>\n",
       "      <td>630</td>\n",
       "      <td>30.43</td>\n",
       "      <td>207</td>\n",
       "      <td>451</td>\n",
       "      <td>13</td>\n",
       "      <td>258</td>\n",
       "      <td>4.000000e-46</td>\n",
       "      <td>180.0</td>\n",
       "      <td>COG0175</td>\n",
       "      <td>3'-phosphoadenosine 5'-phosphosulfate sulfotra...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048</th>\n",
       "      <td>2771095981</td>\n",
       "      <td>462</td>\n",
       "      <td>29.18</td>\n",
       "      <td>11</td>\n",
       "      <td>460</td>\n",
       "      <td>1</td>\n",
       "      <td>455</td>\n",
       "      <td>5.000000e-51</td>\n",
       "      <td>195.0</td>\n",
       "      <td>COG0277</td>\n",
       "      <td>FAD/FMN-containing dehydrogenase</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>2771096306</td>\n",
       "      <td>620</td>\n",
       "      <td>31.58</td>\n",
       "      <td>209</td>\n",
       "      <td>468</td>\n",
       "      <td>2</td>\n",
       "      <td>261</td>\n",
       "      <td>3.000000e-40</td>\n",
       "      <td>160.0</td>\n",
       "      <td>COG0175</td>\n",
       "      <td>3'-phosphoadenosine 5'-phosphosulfate sulfotra...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1317</th>\n",
       "      <td>2771096455</td>\n",
       "      <td>478</td>\n",
       "      <td>45.63</td>\n",
       "      <td>3</td>\n",
       "      <td>460</td>\n",
       "      <td>16</td>\n",
       "      <td>484</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>481.0</td>\n",
       "      <td>COG2509</td>\n",
       "      <td>Uncharacterized FAD-dependent dehydrogenase</td>\n",
       "      <td>486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1531</th>\n",
       "      <td>2771096757</td>\n",
       "      <td>481</td>\n",
       "      <td>26.64</td>\n",
       "      <td>13</td>\n",
       "      <td>475</td>\n",
       "      <td>1</td>\n",
       "      <td>456</td>\n",
       "      <td>2.000000e-57</td>\n",
       "      <td>217.0</td>\n",
       "      <td>COG0277</td>\n",
       "      <td>FAD/FMN-containing dehydrogenase</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620</th>\n",
       "      <td>2771096899</td>\n",
       "      <td>422</td>\n",
       "      <td>31.37</td>\n",
       "      <td>195</td>\n",
       "      <td>387</td>\n",
       "      <td>15</td>\n",
       "      <td>214</td>\n",
       "      <td>6.000000e-33</td>\n",
       "      <td>135.0</td>\n",
       "      <td>COG0175</td>\n",
       "      <td>3'-phosphoadenosine 5'-phosphosulfate sulfotra...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1962</th>\n",
       "      <td>2771097507</td>\n",
       "      <td>463</td>\n",
       "      <td>26.07</td>\n",
       "      <td>9</td>\n",
       "      <td>456</td>\n",
       "      <td>1</td>\n",
       "      <td>452</td>\n",
       "      <td>7.000000e-43</td>\n",
       "      <td>168.0</td>\n",
       "      <td>COG0277</td>\n",
       "      <td>FAD/FMN-containing dehydrogenase</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2060</th>\n",
       "      <td>2771097663</td>\n",
       "      <td>630</td>\n",
       "      <td>30.43</td>\n",
       "      <td>207</td>\n",
       "      <td>451</td>\n",
       "      <td>13</td>\n",
       "      <td>258</td>\n",
       "      <td>4.000000e-46</td>\n",
       "      <td>180.0</td>\n",
       "      <td>COG0175</td>\n",
       "      <td>3'-phosphoadenosine 5'-phosphosulfate sulfotra...</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2166</th>\n",
       "      <td>2771097913</td>\n",
       "      <td>462</td>\n",
       "      <td>29.18</td>\n",
       "      <td>11</td>\n",
       "      <td>460</td>\n",
       "      <td>1</td>\n",
       "      <td>455</td>\n",
       "      <td>5.000000e-51</td>\n",
       "      <td>195.0</td>\n",
       "      <td>COG0277</td>\n",
       "      <td>FAD/FMN-containing dehydrogenase</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_oid  gene_length  percent_identity  query_start  query_end  \\\n",
       "86    2771094374          620             31.58          209        468   \n",
       "199   2771094523          478             45.63            3        460   \n",
       "413   2771094825          481             26.64           13        475   \n",
       "502   2771094967          422             31.37          195        387   \n",
       "844   2771095575          463             26.07            9        456   \n",
       "942   2771095731          630             30.43          207        451   \n",
       "1048  2771095981          462             29.18           11        460   \n",
       "1204  2771096306          620             31.58          209        468   \n",
       "1317  2771096455          478             45.63            3        460   \n",
       "1531  2771096757          481             26.64           13        475   \n",
       "1620  2771096899          422             31.37          195        387   \n",
       "1962  2771097507          463             26.07            9        456   \n",
       "2060  2771097663          630             30.43          207        451   \n",
       "2166  2771097913          462             29.18           11        460   \n",
       "\n",
       "      subj_start  subj_end        evalue  bit_score   cog_id  \\\n",
       "86             2       261  3.000000e-40      160.0  COG0175   \n",
       "199           16       484  0.000000e+00      481.0  COG2509   \n",
       "413            1       456  2.000000e-57      217.0  COG0277   \n",
       "502           15       214  6.000000e-33      135.0  COG0175   \n",
       "844            1       452  7.000000e-43      168.0  COG0277   \n",
       "942           13       258  4.000000e-46      180.0  COG0175   \n",
       "1048           1       455  5.000000e-51      195.0  COG0277   \n",
       "1204           2       261  3.000000e-40      160.0  COG0175   \n",
       "1317          16       484  0.000000e+00      481.0  COG2509   \n",
       "1531           1       456  2.000000e-57      217.0  COG0277   \n",
       "1620          15       214  6.000000e-33      135.0  COG0175   \n",
       "1962           1       452  7.000000e-43      168.0  COG0277   \n",
       "2060          13       258  4.000000e-46      180.0  COG0175   \n",
       "2166           1       455  5.000000e-51      195.0  COG0277   \n",
       "\n",
       "                                               cog_name  cog_length  \n",
       "86    3'-phosphoadenosine 5'-phosphosulfate sulfotra...         261  \n",
       "199         Uncharacterized FAD-dependent dehydrogenase         486  \n",
       "413                    FAD/FMN-containing dehydrogenase         459  \n",
       "502   3'-phosphoadenosine 5'-phosphosulfate sulfotra...         261  \n",
       "844                    FAD/FMN-containing dehydrogenase         459  \n",
       "942   3'-phosphoadenosine 5'-phosphosulfate sulfotra...         261  \n",
       "1048                   FAD/FMN-containing dehydrogenase         459  \n",
       "1204  3'-phosphoadenosine 5'-phosphosulfate sulfotra...         261  \n",
       "1317        Uncharacterized FAD-dependent dehydrogenase         486  \n",
       "1531                   FAD/FMN-containing dehydrogenase         459  \n",
       "1620  3'-phosphoadenosine 5'-phosphosulfate sulfotra...         261  \n",
       "1962                   FAD/FMN-containing dehydrogenase         459  \n",
       "2060  3'-phosphoadenosine 5'-phosphosulfate sulfotra...         261  \n",
       "2166                   FAD/FMN-containing dehydrogenase         459  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codf_s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make pfam table for all bins \n",
    "\"\"\"ptab.count().cumsum()\n",
    "n=0\n",
    "for cols in ptab.columns:\n",
    "    if n==0:\n",
    "        pfdf_tot = pd.DataFrame(index=range(ptab[cols].count()),columns=['pfam_id','pfam_name','bin'])\n",
    "        pfdf_tot['pfam_id'] = [p[0:9] for p in list(ptab[cols].dropna().values)]\n",
    "        pfdf_tot['pfam_name'] = [p[11:] for p in list(ptab[cols].dropna().values)]\n",
    "        pfdf_tot,bla = get_pfam_description(pfdf_tot)\n",
    "        pfdf_tot['bin'] = cols\n",
    "        n=1\n",
    "    else:\n",
    "        pfdf_tem = pd.DataFrame(index=range(ptab[cols].count()),columns=['pfam_id','pfam_name','bin'])\n",
    "        pfdf_tem['pfam_id'] = [p[0:9] for p in list(ptab[cols].dropna().values)]\n",
    "        pfdf_tem['pfam_name'] = [p[11:] for p in list(ptab[cols].dropna().values)]\n",
    "        #print(len(pfdf_tem))\n",
    "        pfdf_tem,bli = get_pfam_description(pfdf_tem)\n",
    "        pfdf_tem['bin'] = cols\n",
    "        pfdf_tot = pfdf_tot.append(pfdf_tem, ignore_index=True)\"\"\"\n",
    "#pfdf_tot.to_csv('MAGs/Analysis/ObsidianAll_pfam.table',sep='\\t')        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_hitdf = cog_hitdf.append(cmap_30_archaea[cmap_30_archaea.index.str.contains(name)])\n",
    "cog_hitdf.to_csv('writeup/COG30_specific_enzymes_round2.table',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mancur = pd.read_csv('writeup/COGlist_manual_curation.csv',sep='\\t',header=None)[0].values\n",
    "mancur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = 'COG0574'\n",
    "name = 'COG1014'\n",
    "addition = cmap_clade_30_long[verstr_clade][cmap_clade_30_long[verstr_clade].index.str.contains(name)].index.str[-7:]\n",
    "cmap_clade_30_long[verstr_clade][cmap_clade_30_long[verstr_clade].index.str.contains(name)]\n",
    "cmap_clade_30_long[cmap_clade_30_long.index.str.contains(name)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_hitdf_clade = cmap_clade_30_long[bathy_clade][cmap_clade_30_long[bathy_clade].index.isin(enzymes)]\n",
    "cog_hitdf_clade.to_csv('writeup/COG30_bathy_clade_specific_enzymes_round1.table',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_hitdf_clade = cog_hitdf_clade.append(cmap_clade_30_long[verstr_clade][cmap_clade_30_long[verstr_clade].index.str.contains(name)])\n",
    "cog_hitdf_clade.to_csv('writeup/COG30_verstraete_clade_specific_enzymes_round1.table',sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## plot heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = plt.figure()\n",
    "f.set_figwidth(16)\n",
    "f.set_figheight(33)\n",
    "gs = gridspec.GridSpec(1,1)\n",
    "ax1 = f.add_subplot(gs[0,0])\n",
    "\n",
    "#sns.heatmap(cmap_30.loc[mancur_short,cmap_30_archaea.columns].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "#sns.heatmap(cmap_30_archaea.loc[list(mancur),:].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "sns.heatmap(cmap_30_arch_elev.loc[list(mancur2),:].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "            ,vmin=-2,vmax=2.5,linewidths=.1,ax=ax1)\n",
    "plt.xlabel('')\n",
    "#plt.show()\n",
    "plt.gcf().subplots_adjust(left=0.25,right=0.99,bottom=0.43,top=0.99)\n",
    "f.savefig('writeup/COG30_specific_enzymes_round36.pdf')\n",
    "f.savefig('writeup/COG30_specific_enzymes_round36.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = list(cladeDF_tot.Bin_ID)\n",
    "# Plot percent identity as heatmap\n",
    "f = plt.figure()\n",
    "f.set_figwidth(16)\n",
    "f.set_figheight(33)\n",
    "gs = gridspec.GridSpec(1,1)\n",
    "ax1 = f.add_subplot(gs[0,0])\n",
    "\n",
    "#sns.heatmap(cmap_30.loc[mancur_short,cmap_30_archaea.columns].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "#sns.heatmap(cmap_30_archaea.loc[list(mancur),:].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "sns.heatmap(cmap_thr_long.loc[mancur_methano,cols].T.fillna(-10),xticklabels=True,yticklabels=True,cbar=False,cmap='Oranges'\n",
    "            ,vmin=0,vmax=80,linewidths=.1,ax=ax1,mask=(cmap_thr_long.loc[mancur_methano,cols].T.fillna(-10)==-10))\n",
    "plt.xlabel('')\n",
    "#plt.show()\n",
    "plt.gcf().subplots_adjust(left=0.25,right=0.99,bottom=0.43,top=0.99)\n",
    "f.savefig('writeup/COGPI_specific_enzymes_methano2.pdf')\n",
    "f.savefig('writeup/COGPI_specific_enzymes_methano2.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap_mcrcodh_thr_long\n",
    "\n",
    "# Plot percent identity as heatmap\n",
    "f = plt.figure()\n",
    "f.set_figwidth(16)\n",
    "f.set_figheight(23)\n",
    "gs = gridspec.GridSpec(1,1)\n",
    "ax1 = f.add_subplot(gs[0,0])\n",
    "\n",
    "#sns.heatmap(cmap_30.loc[mancur_short,cmap_30_archaea.columns].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "#sns.heatmap(cmap_30_archaea.loc[list(mancur),:].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "sns.heatmap(cmap_mcrcodh_thr_long.loc[mancur_methano,:].T.fillna(-10),xticklabels=True,yticklabels=True,cbar=False,cmap='Oranges'\n",
    "            ,vmin=0,vmax=80,linewidths=.1,ax=ax1,mask=(cmap_mcrcodh_thr_long.loc[mancur_methano,:].T.fillna(-10)==-10))\n",
    "plt.xlabel('')\n",
    "#plt.show()\n",
    "plt.gcf().subplots_adjust(left=0.25,right=0.99,bottom=0.43,top=0.99)\n",
    "#f.savefig('writeup/COGPI_euryarch_specific_enzymes.pdf')\n",
    "#f.savefig('writeup/COGPI_euryarch_specific_enzymes.png')\n",
    "\n",
    "f.savefig('writeup/COGPI_euryarch_specific_enzymes_methano3.pdf')\n",
    "f.savefig('writeup/COGPI_euryarch_specific_enzymes_methano3.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot percent identity as heatmap for clades\n",
    "# bathy_clade,verstr_clade,desu_clade,newp_clade\n",
    "cla = 'newp_clade'\n",
    "f = plt.figure()\n",
    "f.set_figwidth(25)\n",
    "f.set_figheight(30)\n",
    "gs = gridspec.GridSpec(1,1)\n",
    "ax1 = f.add_subplot(gs[0,0])\n",
    "#sns.heatmap(cmap_30.loc[mancur_short,cmap_30_archaea.columns].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "#sns.heatmap(cmap_30_archaea.loc[list(mancur),:].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "sns.heatmap(cmap_clade_thr.loc[mancur_methano,finalcolorder_methano].T.fillna(-10),xticklabels=True,yticklabels=True,cbar=True,cmap='Oranges'\n",
    "            ,vmin=0,vmax=80,linewidths=.1,ax=ax1,mask=(cmap_clade_thr.loc[mancur_methano,finalcolorder_methano].T.fillna(-10)==-10))\n",
    "plt.xlabel('')\n",
    "#plt.show()\n",
    "plt.gcf().subplots_adjust(left=0.25,right=0.99,bottom=0.53,top=0.99)\n",
    "#f.savefig('writeup/COGPI_specific_enzymes_'+cla+'_methano.pdf')\n",
    "#f.savefig('writeup/COGPI_specific_enzymes_'+cla+'_methano.png')\n",
    "\n",
    "f.savefig('writeup/COGPI_specific_enzymes_clades_methano2.pdf')\n",
    "f.savefig('writeup/COGPI_specific_enzymes_clades_methano2.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#bathy_clade,verstr_clade,desu_clade,newp_clade\n",
    "\n",
    "plotdf = cmap_clade_30_long.loc[mancur,verstr_clade].fillna(0).copy()\n",
    "cols2elev = [n for n in elevated_nums if n in plotdf.columns]\n",
    "plotdf[cols2elev] = plotdf[cols2elev].multiply(2)\n",
    "\n",
    "f = plt.figure()\n",
    "f.set_figwidth(16)\n",
    "f.set_figheight(33)\n",
    "gs = gridspec.GridSpec(1,1)\n",
    "ax1 = f.add_subplot(gs[0,0])\n",
    "#sns.heatmap(cmap_clade_30_long.loc[mancur,desu_clade].T.fillna(0),xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "#            ,vmin=-0.5,vmax=1,linewidths=.1,ax=ax1)\n",
    "sns.heatmap(plotdf.T,xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "            ,vmin=-2,vmax=2.5,linewidths=.1,ax=ax1)\n",
    "plt.xlabel('')\n",
    "#plt.yticks(rotation=90)\n",
    "#plt.show()\n",
    "plt.gcf().subplots_adjust(left=0.25,right=0.99,bottom=0.43,top=0.99)\n",
    "f.savefig('writeup/COG30_clade_verstr_specific_enzymes_round4.pdf')\n",
    "f.savefig('writeup/COG30_clade_verstr_specific_enzymes_round4.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run through fasta's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import HTSeq\n",
    "#pd.read_table('/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/Obs3_genome_046_IMG_2770939375/IMG_Data/')      \n",
    "#input_file = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/Obs3_genome_046_IMG_2770939375/IMG_Data/2770939375/2770939375.genes.faa'          \n",
    "length = 0\n",
    "#output_file = 'MAGs/ObsidianPool_CODHACS_genes.faa'\n",
    "#output_file = 'MAGs/ObsidianPool_H4MPT_genes.faa'\n",
    "#output_file = 'MAGs/H4MPT/ObsidianPool_H4MPT_genes_subH.faa' #\"tetrahydromethanopterin S-methyltransferase\" & ('subunit H [' in s.descr) \n",
    "#output_file = 'MAGs/Mcr/ObsidianPool_Mcr.faa'\n",
    "output_file = 'MAGs/testfile.faa'\n",
    "description_list = []\n",
    "with open(output_file, 'w') as out_file:\n",
    "    main_dir = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'\n",
    "    for f in lookup_obsidian.index:\n",
    "        IMG_ID = str(lookup_obsidian.loc[f,'IMG_ID'])\n",
    "        input_file = main_dir+f+'_IMG_'+IMG_ID+'/IMG_Data/'+IMG_ID+'/'+IMG_ID+'.genes.faa'\n",
    "        for s in HTSeq.FastaReader(input_file):\n",
    "            if (\"F420\" in s.descr) & (s.descr[19:] not in description_list):\n",
    "                print(s.descr)\n",
    "                #print(s.descr[-16:-1])\n",
    "                description_list = description_list + [s.descr[19:]]\n",
    "                s.name = s.descr[-16:-1]\n",
    "                print(s.name)\n",
    "                length+=1\n",
    "                s.write_to_fasta_file(out_file)\n",
    "    out_file.close()\n",
    "length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length = 0\n",
    "#output_file = 'MAGs/ObsidianPool_CODHACS_genes.faa'\n",
    "#output_file = 'MAGs/ObsidianPool_Clade_H4MPT_genes.faa'\n",
    "#output_file = 'MAGs/H4MPT/ObsidianPool_Clade_H4MPT_genes_subH.faa' #\"tetrahydromethanopterin S-methyltransferase\" & ('subunit H' in s.descr)\n",
    "output_file = 'MAGs/Mcr//ObsidianPool_Clade_Mcr.faa'\n",
    "output_file = 'MAGs/test.faa'\n",
    "description_list = []\n",
    "\n",
    "with open(output_file, 'w') as out_file:\n",
    "    main_dir = 'MAGs/extractclades/'\n",
    "    for bins in ktab_clade.columns:\n",
    "        if (bins[0:3]!='Obs') & (bins[0:3]!='UBA'):\n",
    "            subfolder = cladeDF[cladeDF['Bin_ID']==bins]['Group'].values[0]\n",
    "            input_file = main_dir+subfolder+'/'+bins+'/'+bins+'.genes.faa'\n",
    "            for s in HTSeq.FastaReader(input_file):\n",
    "                if (\"ethyl-co\" in s.descr) & (s.descr[19:] not in description_list) :   \n",
    "                    \n",
    "                    addi = s.descr.split('[')[1][:-1]\n",
    "                    name = s.name+'_'+addi.replace(\" \", \"_\")\n",
    "                    seq = HTSeq.Sequence(s.seq,name)\n",
    "                    seq.descr = s.descr\n",
    "                    description_list = description_list + [s.descr[19:]]\n",
    "                    length+=1\n",
    "                    print(seq.descr)\n",
    "                    print(seq.name)\n",
    "                    \n",
    "                    seq.write_to_fasta_file(out_file)\n",
    "    out_file.close()\n",
    "length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
