{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def make_checkm_output_df(filedir):\n",
    "    # run the read_checkm_tsv and add the outputfile.txt\n",
    "    meta = read_checkm_tsv(filedir)\n",
    "    strhet = pd.read_table(filedir+'outputfile.txt')\n",
    "    strhet = strhet.set_index('Bin Id')\n",
    "    \n",
    "    meta = meta.join(strhet['Strain heterogeneity'])\n",
    "    meta = meta.sort_values('Completeness',ascending=False)\n",
    "    return(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot the contigs of each genome bin seperately and color according to their being selected\n",
    "# for final bin or not (multiple colors for bin splits)\n",
    "# here add the checkM results for a particular bin \n",
    "man_genome_contigs['name'] = man_genome_contigs['Sample']+'_genome_'+man_genome_contigs['OriginalGenomeID']\n",
    "\n",
    "# get the relevant metadata files \n",
    "metaTake2 = make_checkm_output_df('Checkm/Manual_genomes_obs_take2/')\n",
    "metaTake2_oidx = get_old_bin_names(metaTake2,26,29)\n",
    "reassem = make_checkm_output_df('Checkm/reassembly_3kb/')\n",
    "reassem_oldidx = get_old_bin_names(reassem,26,29)\n",
    "orig_bins = make_checkm_output_df('Checkm/obsidian_all_checkm/')\n",
    "orig_bins = orig_bins.loc[alt5.index,:]\n",
    "transl = pd.read_table('fasta/genome_bins_obsidian/info.table')\n",
    "transl['brian_name'] = 'Obsidian'+transl['sample'].str[-1]+'_genome_'+transl['NewName'].str[-3:]\n",
    "Qfilter = pd.DataFrame(index=transl['brian_name'],columns=['old_name','original','reassembled','c1','c2','c3','c4'])\n",
    "Qfilter['old_name'] = [transl[transl['brian_name']==name]['Unnamed: 0'].values[0] for name in Qfilter.index]#[s for s in alt5.index]\n",
    "\n",
    "for name in man_genome_contigs['name'].unique():\n",
    "    tempdf = man_genome_contigs[man_genome_contigs['name']==name]\n",
    "    one = 0;two = 0;three = 0;four = 0\n",
    "    f = plt.figure()\n",
    "    f.set_figwidth(9)\n",
    "    f.set_figheight(9)\n",
    "    \n",
    "    gs = gridspec.GridSpec(4,4)\n",
    "    \n",
    "    ax1 = f.add_subplot(gs[0:2,0:2])\n",
    "    if len(tempdf[tempdf['ManualGenomeID'].str[-1]=='1'])>0:\n",
    "        t1 = tempdf[tempdf['ManualGenomeID'].str[-1]=='1']\n",
    "        t1.plot.scatter('GC','FPK',s=np.array(t1['ContigLength'])/3e2, alpha=0.5,color='blue',ax=ax1,fontsize=10)\n",
    "        one = 1\n",
    "    if len(tempdf[tempdf['ManualGenomeID'].str[-1]=='2'])>0:\n",
    "        t2 = tempdf[tempdf['ManualGenomeID'].str[-1]=='2']\n",
    "        t2.plot.scatter('GC','FPK',s=np.array(t2['ContigLength'])/3e2,alpha=0.5,color='red',ax=ax1,fontsize=10)\n",
    "        two = 1\n",
    "    if len(tempdf[tempdf['ManualGenomeID'].str[-1]=='3'])>0:\n",
    "        t3 = tempdf[tempdf['ManualGenomeID'].str[-1]=='3']\n",
    "        t3.plot.scatter('GC','FPK',s=np.array(t3['ContigLength'])/3e2,alpha=0.5,color='g',ax=ax1,fontsize=10)\n",
    "        three = 1\n",
    "    if len(tempdf[tempdf['ManualGenomeID'].str[-1]=='4'])>0:\n",
    "        t4 = tempdf[tempdf['ManualGenomeID'].str[-1]=='4']\n",
    "        t4.plot.scatter('GC','FPK',s=np.array(t4['ContigLength'])/3e2,alpha=0.5,color='m',ax=ax1,fontsize=10)\n",
    "        four = 1\n",
    "    if len(tempdf[tempdf['IsDropped']=='Yes'])>0:\n",
    "        t0 = tempdf[tempdf['IsDropped']=='Yes']\n",
    "        t0.plot.scatter('GC','FPK',s=np.array(t0['ContigLength'])/3e2,alpha=0.3,color='k',ax=ax1,fontsize=10)\n",
    "    plt.xlabel('GC',fontsize=12)\n",
    "    plt.ylim(1e0,1e6);plt.xlim(0.2,0.8)\n",
    "    plt.yscale('log')\n",
    "    plt.ylabel('Fragments per kbp',fontsize=12)\n",
    "    plt.title(name,fontsize=13)\n",
    "    \n",
    "    if one == 1:\n",
    "        ax2 = f.add_subplot(gs[0,2]) # this is bin #1 \n",
    "        data = metaTake2[metaTake2.index.str[26:29] == name[-3:]]\n",
    "        data = data[data.index.str[-1] == '1']\n",
    "        data[['Completeness','Contamination','Strain heterogeneity']].plot.bar(ax=ax2,fontsize=10)\n",
    "        plt.ylim(0,100)\n",
    "        ax2.legend(bbox_to_anchor=(0.1, 1),loc=3, ncol=1, mode=\"expand\", borderaxespad=0.,fontsize=10)\n",
    "        plt.xticks([]);plt.xlabel('');\n",
    "        ax2.text(0.7,30,\n",
    "                 'lineage:   '+data['marker lineage'][0]\n",
    "                 +'\\nlength (Mb):       '+str(round(data['Genome size'][0]/1e6,2))\n",
    "                 +'\\n# contigs:           '+data['# contigs'].astype(str)[0]\n",
    "                 +'\\nN50 (kb):            '+str(round(data['N50 (contigs)'][0]/1e3,1)),\n",
    "                 fontsize=12,color='b')\n",
    "        if (data['Completeness'][0]>10) & (data['Contamination'][0]<5) & (data['Strain heterogeneity'][0]<5):\n",
    "            Qfilter.loc[name,'c1'] = data['Completeness'][0]\n",
    "    if two == 1:\n",
    "        ax3 = f.add_subplot(gs[1,2]) # this is bin #2\n",
    "        data = metaTake2[metaTake2.index.str[26:29] == name[-3:]]\n",
    "        data = data[data.index.str[-1] == '2']\n",
    "        data[['Completeness','Contamination','Strain heterogeneity']].plot.bar(ax=ax3,legend=False)\n",
    "        plt.xticks([]);plt.ylim(0,100);plt.xlabel('');plt.yticks([])\n",
    "        ax3.text(0.7,30,\n",
    "                 'lineage:   '+data['marker lineage'][0]\n",
    "                 +'\\nlength (Mb):       '+str(round(data['Genome size'][0]/1e6,2))\n",
    "                 +'\\n# contigs:           '+data['# contigs'].astype(str)[0]\n",
    "                 +'\\nN50 (kb):            '+str(round(data['N50 (contigs)'][0]/1e3,1)),\n",
    "                 fontsize=12,color='r')\n",
    "        if (data['Completeness'][0]>10) & (data['Contamination'][0]<5) & (data['Strain heterogeneity'][0]<5):\n",
    "            Qfilter.loc[name,'c2'] = data['Completeness'][0]\n",
    "    if three == 1:\n",
    "        ax4 = f.add_subplot(gs[2,2]) # this is bin #3\n",
    "        data = metaTake2[metaTake2.index.str[26:29] == name[-3:]]\n",
    "        data = data[data.index.str[-1] == '3']\n",
    "        data[['Completeness','Contamination','Strain heterogeneity']].plot.bar(ax=ax4,legend=False,fontsize=10)\n",
    "        plt.xticks([]);plt.ylim(0,100);plt.xlabel('');plt.yticks([])\n",
    "        ax4.text(0.7,30,\n",
    "                 'lineage:   '+data['marker lineage'][0]\n",
    "                 +'\\nlength (Mb):       '+str(round(data['Genome size'][0]/1e6,2))\n",
    "                 +'\\n# contigs:           '+data['# contigs'].astype(str)[0]\n",
    "                 +'\\nN50 (kb):            '+str(round(data['N50 (contigs)'][0]/1e3,1)),\n",
    "                 fontsize=12,color='g')\n",
    "        if (data['Completeness'][0]>10) & (data['Contamination'][0]<5) & (data['Strain heterogeneity'][0]<5):\n",
    "            Qfilter.loc[name,'c3'] = data['Completeness'][0]\n",
    "    if four == 1:\n",
    "        ax5 = f.add_subplot(gs[3,2]) # this is bin #3\n",
    "        data = metaTake2[metaTake2.index.str[26:29] == name[-3:]]\n",
    "        data = data[data.index.str[-1] == '4']\n",
    "        data[['Completeness','Contamination','Strain heterogeneity']].plot.bar(ax=ax5,legend=False)\n",
    "        plt.xticks([]);plt.ylim(0,100);plt.xlabel('');plt.yticks([])\n",
    "        ax5.text(0.7,30,\n",
    "                 'lineage:   '+data['marker lineage'][0]\n",
    "                 +'\\nlength (Mb):       '+str(round(data['Genome size'][0]/1e6,2))\n",
    "                 +'\\n# contigs:           '+data['# contigs'].astype(str)[0]\n",
    "                 +'\\nN50 (kb):            '+str(round(data['N50 (contigs)'][0]/1e3,1)),\n",
    "                 fontsize=12,color='m')\n",
    "        if (data['Completeness'][0]>10) & (data['Contamination'][0]<5) & (data['Strain heterogeneity'][0]<5):\n",
    "            Qfilter.loc[name,'c4'] = data['Completeness'][0]\n",
    "    ax6 = f.add_subplot(gs[2,0])\n",
    "    data = reassem[reassem.index.str[26:29]==name[-3:]]\n",
    "    data[['Completeness','Contamination','Strain heterogeneity']].plot.bar(ax=ax6,legend=False,fontsize=10)\n",
    "    plt.xticks([]);plt.ylim(0,100);plt.xlabel('')#;plt.yticks([])\n",
    "    ax6.text(0.6,20,\n",
    "                 'lineage:   '+data['marker lineage'][0]\n",
    "                 +'\\nlength (Mb):     '+str(round(data['Genome size'][0]/1e6,2))\n",
    "                 +'\\n# contigs:         '+data['# contigs'].astype(str)[0]\n",
    "                 +'\\nN50 (kb):          '+str(round(data['N50 (contigs)'][0]/1e3,1)),\n",
    "                 fontsize=12)\n",
    "    plt.ylabel('Reassembly\\nnon-curated',fontsize=12)\n",
    "    if (data['Completeness'][0]>10) & (data['Contamination'][0]<5) & (data['Strain heterogeneity'][0]<5):\n",
    "        Qfilter.loc[name,'reassembled'] = data['Completeness'][0]\n",
    "    ax7 = f.add_subplot(gs[3,0])\n",
    "    data = pd.DataFrame(orig_bins.loc[reassem_oldidx[reassem_oldidx['new_idx'].str[26:29]==name[-3:]].index[0][0:-2]]).T\n",
    "    data[['Completeness','Contamination','Strain heterogeneity']].plot.bar(ax=ax7,legend=False,fontsize=10)\n",
    "    plt.xticks([]);plt.ylim(0,100);plt.xlabel('');#plt.yticks([])\n",
    "    ax7.text(0.6,20,\n",
    "                 'lineage:   '+data['marker lineage'][0]\n",
    "                 +'\\nlength (Mb):     '+str(round(data['Genome size'][0]/1e6,2))\n",
    "                 +'\\n# contigs:         '+data['# contigs'].astype(str)[0]\n",
    "                 +'\\nN50 (kb):          '+str(round(data['N50 (contigs)'][0]/1e3,1)),\n",
    "                 fontsize=12)\n",
    "    plt.title('Original bin: '+data.index[0],fontsize=12)\n",
    "    if (data['Completeness'][0]>10) & (data['Contamination'][0]<5) & (data['Strain heterogeneity'][0]<5):\n",
    "        Qfilter.loc[name,'original'] = data['Completeness'][0]\n",
    "\n",
    "    gs.update(hspace=.2)#,wspace=0.02)\n",
    "    plt.gcf().subplots_adjust(bottom=0.03)\n",
    "    f.savefig('reassembly/Manual_Genomes_take2//plots_QS/'+name) #this is plots+ quality score\n",
    "    f.clf()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
