{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from matplotlib import gridspec\n",
    "import ast\n",
    "import os\n",
    "\n",
    "#os.chdir('/home/bojk/Data/') #this is when opened in '/home/bojk/Data/notebooks' on linux Quakelab\n",
    "os.chdir('/Users/bojk/Google Drive/QuakeLab/Data/')\n",
    "import sys\n",
    "sys.path.append('/home/bojk/Data/minimeta_pyfiles/')\n",
    "sys.path.append('/Users/bojk/Google Drive/QuakeLab/minimeta_pyfiles/')\n",
    "import imp\n",
    "import HTSeq\n",
    "from scipy.stats import powerlaw\n",
    "import minimetagenome_analysis as minimeta\n",
    "\n",
    "import mock_metagenome_creation as mock\n",
    "import minimetagenome_dimensionality as miniD\n",
    "imp.reload(mock)\n",
    "#import reassembly as rea\n",
    "#import subcluster_bins as sub\n",
    "#import checkm_out_plot as cop\n",
    "#import checkm_out as co\n",
    "\n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['axes.labelsize'] = 15\n",
    "mpl.rcParams['xtick.labelsize'] = 15\n",
    "mpl.rcParams['ytick.labelsize'] = 15\n",
    "mpl.rcParams['axes.titlesize'] = 15\n",
    "mpl.rcParams['figure.titlesize'] = 18\n",
    "mpl.rcParams['legend.fontsize'] = 14\n",
    "\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SagMAT KMERS AND MINIMETA CHIP STUFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#metagenome_fasta = '/home/datastorage/ASSEMBLY_DATA/PermafrostSagMAT/Combined_Analysis/super_contigs.PermafrostSagMAT.fasta'        \n",
    "metagenome_fasta = 'Permafrost/SagMAT/super_contigs.PermafrostSagMAT.fasta'\n",
    "save_dir = 'Permafrost/SagMAT/pickles/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Scaffold ID</th>\n",
       "      <th>Scaffold Name</th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Gene Count</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>GC Content</th>\n",
       "      <th>Read Depth</th>\n",
       "      <th>Lineage Domain</th>\n",
       "      <th>Lineage Phylum</th>\n",
       "      <th>...</th>\n",
       "      <th>y_40_a</th>\n",
       "      <th>y_40_n</th>\n",
       "      <th>y_50_a</th>\n",
       "      <th>y_50_n</th>\n",
       "      <th>y_60_a</th>\n",
       "      <th>y_60_n</th>\n",
       "      <th>y_70_a</th>\n",
       "      <th>y_70_n</th>\n",
       "      <th>y_80_a</th>\n",
       "      <th>y_80_n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SuperContig_k91_275_flag=0_multi=8.2373_len=3205_SegMat_25cm_Bulk</th>\n",
       "      <td>3300022816 assembled Ga0255657_100004</td>\n",
       "      <td>100004</td>\n",
       "      <td>3300022816</td>\n",
       "      <td>Permafrost soil microbial communities from Sag...</td>\n",
       "      <td>4</td>\n",
       "      <td>3205</td>\n",
       "      <td>0.58</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SuperContig_k91_360_flag=1_multi=6.0000_len=3358_SegMat_25cm_Bulk</th>\n",
       "      <td>3300022816 assembled Ga0255657_100006</td>\n",
       "      <td>100006</td>\n",
       "      <td>3300022816</td>\n",
       "      <td>Permafrost soil microbial communities from Sag...</td>\n",
       "      <td>4</td>\n",
       "      <td>3358</td>\n",
       "      <td>0.68</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                              Scaffold ID  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...  3300022816 assembled Ga0255657_100004   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...  3300022816 assembled Ga0255657_100006   \n",
       "\n",
       "                                                   Scaffold Name   Genome ID  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...        100004  3300022816   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...        100006  3300022816   \n",
       "\n",
       "                                                                                               Genome  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...  Permafrost soil microbial communities from Sag...   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...  Permafrost soil microbial communities from Sag...   \n",
       "\n",
       "                                                   Gene Count Sequence Length  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...          4            3205   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...          4            3358   \n",
       "\n",
       "                                                   GC Content Read Depth  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...       0.58          1   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...       0.68          1   \n",
       "\n",
       "                                                   Lineage Domain  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...       Bacteria   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...       Bacteria   \n",
       "\n",
       "                                                   Lineage Phylum  ... y_40_a  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...     Unassigned  ...    NaN   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...     Unassigned  ...    NaN   \n",
       "\n",
       "                                                   y_40_n y_50_a y_50_n  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...    NaN    NaN    NaN   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...    NaN    NaN    NaN   \n",
       "\n",
       "                                                   y_60_a y_60_n y_70_a  \\\n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...    NaN    NaN    NaN   \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...    NaN    NaN    NaN   \n",
       "\n",
       "                                                   y_70_n  y_80_a  y_80_n  \n",
       "SuperContig_k91_275_flag=0_multi=8.2373_len=320...    NaN     NaN     NaN  \n",
       "SuperContig_k91_360_flag=1_multi=6.0000_len=335...    NaN     NaN     NaN  \n",
       "\n",
       "[2 rows x 42 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maindf = pd.read_pickle('Permafrost/SagMAT/pickles/SagMAT_maindf')\n",
    "maindf.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65990, 42)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maindf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "contigdf = maindf[['Sequence Length','GC Content','Read Depth']].copy()\n",
    "contigdf.rename(index=str,columns={'Sequence Length':'Sequence length','GC Content':'GC'},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33580, 365)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chipAbs = pd.read_pickle('Permafrost/SagMAT/pickles/SagMAT_chipdf_bpcounts')\n",
    "chipMini = chipAbs[~chipAbs.index.str.contains('Bulk')].copy()\n",
    "chipMini.drop(['SagMAT_25cm_Bulk','SagMAT_15cm_Bulk'],axis=1,inplace=True)\n",
    "chipMini.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15319, 3)\n",
      "110750347\n",
      "5138.0\n"
     ]
    }
   ],
   "source": [
    "idx = chipMini[chipMini>0\n",
    "              ].dropna(how='all').fillna(0).index\n",
    "print(contigdf.loc[idx].shape)\n",
    "print(contigdf.loc[idx,'Sequence length'].sum())\n",
    "print(contigdf.loc[idx,'Sequence length'].median())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15319, 365)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chipMini[chipMini>0].dropna(how='all').fillna(0).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chimMinifinal = chipMini[chipMini>0].dropna(how='all').fillna(0).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building tSNE of all 365 dimensions\n",
      "finished building main tSNE, this took 526.74 seconds\n",
      "performing cluster sweep of tSNE of all 365 dimensions\n",
      "Performing PCA...\n",
      "building tSNE of 4 PCs\n",
      "finished building tSNE of 4 PCs, this took 396.40 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 6 PCs\n",
      "finished building tSNE of 6 PCs, this took 408.98 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 9 PCs\n",
      "finished building tSNE of 9 PCs, this took 385.24 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 13 PCs\n",
      "finished building tSNE of 13 PCs, this took 384.55 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 18 PCs\n",
      "finished building tSNE of 18 PCs, this took 391.09 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 26 PCs\n",
      "finished building tSNE of 26 PCs, this took 443.27 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 38 PCs\n",
      "finished building tSNE of 38 PCs, this took 458.11 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 55 PCs\n",
      "finished building tSNE of 55 PCs, this took 444.80 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 78 PCs\n",
      "finished building tSNE of 78 PCs, this took 444.09 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 113 PCs\n",
      "finished building tSNE of 113 PCs, this took 452.08 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 162 PCs\n",
      "finished building tSNE of 162 PCs, this took 443.31 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 234 PCs\n",
      "finished building tSNE of 234 PCs, this took 473.00 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 336 PCs\n",
      "finished building tSNE of 336 PCs, this took 526.49 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "Finished everything.\n",
      "Total elapsed time is 102.99 minutes\n"
     ]
    }
   ],
   "source": [
    "\n",
    "imp.reload(miniD)\n",
    "maindir = 'Permafrost/SagMAT/chipPCA/'\n",
    "#SMmini_alltsne = pd.DataFrame('Permafrost/SagMAT/pickles/SMminiCovOnly6mers_all_tSNEs')\n",
    "SMmini_alltsne = miniD.perform_complete_analysis_Coverage(chimMinifinal,6,contigdf,maindir,'SMminiCovOnly')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building tSNE of all 365 dimensions\n",
      "finished building main tSNE, this took 394.43 seconds\n",
      "performing cluster sweep of tSNE of all 365 dimensions\n",
      "Performing PCA...\n",
      "building tSNE of 4 PCs\n",
      "finished building tSNE of 4 PCs, this took 312.12 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 6 PCs\n",
      "finished building tSNE of 6 PCs, this took 334.73 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 9 PCs\n",
      "finished building tSNE of 9 PCs, this took 299.72 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 13 PCs\n",
      "finished building tSNE of 13 PCs, this took 297.62 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 18 PCs\n",
      "finished building tSNE of 18 PCs, this took 296.52 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 26 PCs\n",
      "finished building tSNE of 26 PCs, this took 298.45 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 38 PCs\n",
      "finished building tSNE of 38 PCs, this took 320.96 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 55 PCs\n",
      "finished building tSNE of 55 PCs, this took 308.94 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 78 PCs\n",
      "finished building tSNE of 78 PCs, this took 314.99 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 113 PCs\n",
      "finished building tSNE of 113 PCs, this took 318.79 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 162 PCs\n",
      "finished building tSNE of 162 PCs, this took 338.45 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 234 PCs\n",
      "finished building tSNE of 234 PCs, this took 364.65 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 336 PCs\n",
      "finished building tSNE of 336 PCs, this took 382.40 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "Finished everything.\n",
      "Total elapsed time is 76.39 minutes\n"
     ]
    }
   ],
   "source": [
    "maindir = 'Permafrost/SagMAT/chipPCA/'\n",
    "SMmini_alltsne3kb = miniD.perform_complete_analysis_Coverage(chimMinifinal[chimMinifinal>3000].dropna(how='all').fillna(0),6,contigdf,maindir,'SMminiCovOnly3kbThres')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building tSNE of all 365 dimensions\n",
      "finished building main tSNE, this took 165.61 seconds\n",
      "performing cluster sweep of tSNE of all 365 dimensions\n",
      "Performing PCA...\n",
      "building tSNE of 4 PCs\n",
      "finished building tSNE of 4 PCs, this took 136.71 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 6 PCs\n",
      "finished building tSNE of 6 PCs, this took 136.93 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 9 PCs\n",
      "finished building tSNE of 9 PCs, this took 135.74 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 13 PCs\n",
      "finished building tSNE of 13 PCs, this took 143.23 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 18 PCs\n",
      "finished building tSNE of 18 PCs, this took 136.38 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 26 PCs\n",
      "finished building tSNE of 26 PCs, this took 137.20 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 38 PCs\n",
      "finished building tSNE of 38 PCs, this took 140.37 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 55 PCs\n",
      "finished building tSNE of 55 PCs, this took 141.63 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 78 PCs\n",
      "finished building tSNE of 78 PCs, this took 148.02 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 113 PCs\n",
      "finished building tSNE of 113 PCs, this took 150.71 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 162 PCs\n",
      "finished building tSNE of 162 PCs, this took 159.12 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 234 PCs\n",
      "finished building tSNE of 234 PCs, this took 169.90 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 336 PCs\n",
      "finished building tSNE of 336 PCs, this took 193.19 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "Finished everything.\n",
      "Total elapsed time is 34.92 minutes\n"
     ]
    }
   ],
   "source": [
    "SMmini_alltsne5kb = miniD.perform_complete_analysis_Coverage(chimMinifinal[chimMinifinal>5000].dropna(how='all').fillna(0),6,contigdf,maindir,'SMminiCovOnly5kbThres')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building tSNE of all 365 dimensions\n",
      "finished building main tSNE, this took 118.72 seconds\n",
      "performing cluster sweep of tSNE of all 365 dimensions\n",
      "Performing PCA...\n",
      "building tSNE of 4 PCs\n",
      "finished building tSNE of 4 PCs, this took 81.75 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 6 PCs\n",
      "finished building tSNE of 6 PCs, this took 80.88 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 9 PCs\n",
      "finished building tSNE of 9 PCs, this took 80.57 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 13 PCs\n",
      "finished building tSNE of 13 PCs, this took 86.33 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 18 PCs\n",
      "finished building tSNE of 18 PCs, this took 83.18 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 26 PCs\n",
      "finished building tSNE of 26 PCs, this took 83.45 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 38 PCs\n",
      "finished building tSNE of 38 PCs, this took 85.75 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 55 PCs\n",
      "finished building tSNE of 55 PCs, this took 86.93 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 78 PCs\n",
      "finished building tSNE of 78 PCs, this took 92.41 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 113 PCs\n",
      "finished building tSNE of 113 PCs, this took 102.58 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 162 PCs\n",
      "finished building tSNE of 162 PCs, this took 99.03 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 234 PCs\n",
      "finished building tSNE of 234 PCs, this took 100.43 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 336 PCs\n",
      "finished building tSNE of 336 PCs, this took 109.62 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "Finished everything.\n",
      "Total elapsed time is 21.53 minutes\n"
     ]
    }
   ],
   "source": [
    "SMmini_alltsne7kb = miniD.perform_complete_analysis_Coverage(chimMinifinal[chimMinifinal>7000].dropna(how='all').fillna(0),6,contigdf,maindir,'SMminiCovOnly7kbThres')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building tSNE of all 365 dimensions\n",
      "finished building main tSNE, this took 50.99 seconds\n",
      "performing cluster sweep of tSNE of all 365 dimensions\n",
      "Performing PCA...\n",
      "building tSNE of 4 PCs\n",
      "finished building tSNE of 4 PCs, this took 38.13 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 6 PCs\n",
      "finished building tSNE of 6 PCs, this took 37.17 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 9 PCs\n",
      "finished building tSNE of 9 PCs, this took 39.29 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 13 PCs\n",
      "finished building tSNE of 13 PCs, this took 40.53 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 18 PCs\n",
      "finished building tSNE of 18 PCs, this took 40.60 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 26 PCs\n",
      "finished building tSNE of 26 PCs, this took 41.95 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 38 PCs\n",
      "finished building tSNE of 38 PCs, this took 43.94 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 55 PCs\n",
      "finished building tSNE of 55 PCs, this took 45.05 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 78 PCs\n",
      "finished building tSNE of 78 PCs, this took 49.26 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 113 PCs\n",
      "finished building tSNE of 113 PCs, this took 49.19 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 162 PCs\n",
      "finished building tSNE of 162 PCs, this took 51.73 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 234 PCs\n",
      "finished building tSNE of 234 PCs, this took 55.46 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "building tSNE of 336 PCs\n",
      "finished building tSNE of 336 PCs, this took 56.56 seconds\n",
      "Index([0, 1, 'Sequence length', 'GC'], dtype='object')\n",
      "Finished everything.\n",
      "Total elapsed time is 10.67 minutes\n"
     ]
    }
   ],
   "source": [
    "SMmini_alltsne10kb = miniD.perform_complete_analysis_Coverage(chimMinifinal[chimMinifinal>10000].dropna(how='all').fillna(0),6,contigdf,maindir,'SMminiCovOnly10kbThres')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'SMmini_alltsne' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-56-60c015399c14>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0midx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSMmini_alltsne\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mSMmaindf_mini\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmaindf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mSMmaindf_mini\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSMmaindf_mini\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSMmini_alltsne\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'SMmini_alltsne' is not defined"
     ]
    }
   ],
   "source": [
    "idx = SMmini_alltsne.index\n",
    "SMmaindf_mini = maindf.loc[idx].copy()\n",
    "SMmaindf_mini = SMmaindf_mini.join(SMmini_alltsne)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"SMmini_alltsne\n",
    "#[['x_PC13','y_PC13']])\n",
    "\"\"\"\n",
    "idx = SMmini_alltsne3kb.index\n",
    "SMmaindf_mini3kb = maindf.loc[idx].copy()\n",
    "SMmaindf_mini3kb = SMmaindf_mini3kb.join(SMmini_alltsne3kb)#[['x_PC13','y_PC13']])\n",
    "\n",
    "idx = SMmini_alltsne5kb.index\n",
    "SMmaindf_mini5kb = maindf.loc[idx].copy()\n",
    "SMmaindf_mini5kb = SMmaindf_mini5kb.join(SMmini_alltsne5kb)#[['x_PC13','y_PC13']])\n",
    "\n",
    "idx = SMmini_alltsne7kb.index\n",
    "SMmaindf_mini7kb = maindf.loc[idx].copy()\n",
    "SMmaindf_mini7kb = SMmaindf_mini7kb.join(SMmini_alltsne7kb)#[['x_PC13','y_PC13']])\n",
    "\n",
    "idx = SMmini_alltsne10kb.index\n",
    "SMmaindf_mini10kb = maindf.loc[idx].copy()\n",
    "SMmaindf_mini10kb = SMmaindf_mini10kb.join(SMmini_alltsne10kb)#[['x_PC13','y_PC13']])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "imp.reload(miniD)\n",
    "#main = pd.read_csv('Permafrost/SIMR/WestDock_metadata.table').set_index('Unnamed: 0')\n",
    "#tsnedf_WDminiCov = pd.read_pickle(maindir+'WDminiCovOnly6mers_all_tSNEs')\n",
    "miniD.plotPCs(tsnedf_WDminiCov,main,maindir,'WDminiChipAbsCov') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "minCS = 100\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'leaf',False,'PC13','WestDock_miniPC13',fasta,write,'NO','Permafrost/SagMAT/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'SagMAT_miniPC'+pc+'_0kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(WDmaindf_mini3,**kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Scaffold ID</th>\n",
       "      <th>Scaffold Name</th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Gene Count</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>GC Content</th>\n",
       "      <th>Read Depth</th>\n",
       "      <th>Lineage Domain</th>\n",
       "      <th>Lineage Phylum</th>\n",
       "      <th>...</th>\n",
       "      <th>x_PC113</th>\n",
       "      <th>y_PC113</th>\n",
       "      <th>x_PC162</th>\n",
       "      <th>y_PC162</th>\n",
       "      <th>x_PC234</th>\n",
       "      <th>y_PC234</th>\n",
       "      <th>x_PC336</th>\n",
       "      <th>y_PC336</th>\n",
       "      <th>DBclusternum</th>\n",
       "      <th>Bin</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SuperContig_NODE_3598_length_7534_cov_2.88191</th>\n",
       "      <td>3300022816 assembled Ga0255657_157610</td>\n",
       "      <td>157610</td>\n",
       "      <td>3300022816</td>\n",
       "      <td>Permafrost soil microbial communities from Sag...</td>\n",
       "      <td>13</td>\n",
       "      <td>7534</td>\n",
       "      <td>0.38</td>\n",
       "      <td>1</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>...</td>\n",
       "      <td>26.000343</td>\n",
       "      <td>-13.918259</td>\n",
       "      <td>-23.339903</td>\n",
       "      <td>-27.095160</td>\n",
       "      <td>28.827444</td>\n",
       "      <td>38.500633</td>\n",
       "      <td>52.13921</td>\n",
       "      <td>14.260105</td>\n",
       "      <td>-1</td>\n",
       "      <td>-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SuperContig_NODE_3222_length_8048_cov_4.20292</th>\n",
       "      <td>3300022816 assembled Ga0255657_157234</td>\n",
       "      <td>157234</td>\n",
       "      <td>3300022816</td>\n",
       "      <td>Permafrost soil microbial communities from Sag...</td>\n",
       "      <td>2</td>\n",
       "      <td>8048</td>\n",
       "      <td>0.44</td>\n",
       "      <td>1</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>...</td>\n",
       "      <td>57.267860</td>\n",
       "      <td>7.114372</td>\n",
       "      <td>-9.631462</td>\n",
       "      <td>42.196144</td>\n",
       "      <td>15.303872</td>\n",
       "      <td>-64.097580</td>\n",
       "      <td>39.74527</td>\n",
       "      <td>55.419998</td>\n",
       "      <td>48</td>\n",
       "      <td>048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 74 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         Scaffold ID  \\\n",
       "Unnamed: 0                                                                             \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191  3300022816 assembled Ga0255657_157610   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292  3300022816 assembled Ga0255657_157234   \n",
       "\n",
       "                                              Scaffold Name   Genome ID  \\\n",
       "Unnamed: 0                                                                \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191        157610  3300022816   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292        157234  3300022816   \n",
       "\n",
       "                                                                                          Genome  \\\n",
       "Unnamed: 0                                                                                         \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191  Permafrost soil microbial communities from Sag...   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292  Permafrost soil microbial communities from Sag...   \n",
       "\n",
       "                                              Gene Count  Sequence Length  \\\n",
       "Unnamed: 0                                                                  \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191         13             7534   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292          2             8048   \n",
       "\n",
       "                                              GC Content  Read Depth  \\\n",
       "Unnamed: 0                                                             \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191       0.38           1   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292       0.44           1   \n",
       "\n",
       "                                              Lineage Domain Lineage Phylum  \\\n",
       "Unnamed: 0                                                                    \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191     Unassigned     Unassigned   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292     Unassigned     Unassigned   \n",
       "\n",
       "                                               ...    x_PC113    y_PC113  \\\n",
       "Unnamed: 0                                     ...                         \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191  ...  26.000343 -13.918259   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292  ...  57.267860   7.114372   \n",
       "\n",
       "                                                 x_PC162    y_PC162  \\\n",
       "Unnamed: 0                                                            \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191 -23.339903 -27.095160   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292  -9.631462  42.196144   \n",
       "\n",
       "                                                 x_PC234    y_PC234   x_PC336  \\\n",
       "Unnamed: 0                                                                      \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191  28.827444  38.500633  52.13921   \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292  15.303872 -64.097580  39.74527   \n",
       "\n",
       "                                                 y_PC336  DBclusternum  Bin  \n",
       "Unnamed: 0                                                                   \n",
       "SuperContig_NODE_3598_length_7534_cov_2.88191  14.260105            -1  -01  \n",
       "SuperContig_NODE_3222_length_8048_cov_4.20292  55.419998            48  048  \n",
       "\n",
       "[2 rows x 74 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SMmaindf_mini3kb.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "imp.reload(minimeta)\n",
    "fasta,write = 'YES','YES'\n",
    "minCS = 70\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'leaf',False,'PC13','WestDock_miniPC13',fasta,write,'NO','Permafrost/SagMAT/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'SagMAT_miniPC'+pc+'_3kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(SMmaindf_mini3kb,**kwargs) # cluster whatever pcs you want here.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "minCS = 40\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'eom',False,'PC13','WestDock_miniPC13',fasta,write,'NO','Permafrost/SagMAT/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'SagMAT_miniPC'+pc+'_5kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(SMmaindf_mini5kb,**kwargs) # cluster whatever pcs you want here.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "minCS = 30\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'leaf',False,'PC13','WestDock_miniPC13',fasta,write,'NO','Permafrost/SagMAT/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'SagMAT_miniPC'+pc+'_7kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(SMmaindf_mini7kb,**kwargs) # cluster whatever pcs you want here.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "minCS = 25\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'leaf',False,'PC13','WestDock_miniPC13',fasta,write,'NO','Permafrost/SagMAT/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'SagMAT_miniPC'+pc+'_10kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(SMmaindf_mini10kb,**kwargs) # cluster whatever pcs you want here.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#get tsne cols in the right format (write function to do this, then pass on to \n",
    "#                                  standard minimeta pipeline)\n",
    "imp.reload(minimeta)\n",
    "\"\"\"\n",
    "minCS ['min_cluster_size']\n",
    "minS ['min_samples']\n",
    "CSM ['cluster_selection_method']\n",
    "ASC ['allow_single_cluster']\n",
    "perp ['perplexity_na'] # this is a string of the form '40_a' or '60_n' (so perplexity and absolute or normalized tsne)\n",
    "expt_name ['expt_name']\n",
    "write_fasta ['write_fasta'] # save the fasta Y/n (as in: do dry run clustering first, as fasta writing takes time)\n",
    "writedf ['write_df']\n",
    "series ['is_this_a_series']\n",
    "\"\"\"\n",
    " # cluster whatever pcs you want here.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16619, 51)"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maindf_mini.groupby('DBclusternum').sum()[['Sequence Length','Read Depth']]\n",
    "maindf_mini.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1a3ac98588>"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "size = maindf_mini['Sequence Length']\n",
    "maindf_mini.plot.scatter('x_PC13','y_PC13',s=size.divide(3e2).astype(float),alpha=.05)#,ax=ax1,c=colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of fasta sequences:45613\n",
      "----------------------------------------------------\n",
      "Started collecting kmers, dimensionality = 4096\n",
      "Local time: 22:24\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 500 contigs out of 45613 done, took 762.82 s\n",
      "Progress: 1.1 %, local time: 22:37\n",
      "Total elapsed time is 12.73 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 1000 contigs out of 45613 done, took 779.61 s\n",
      "Progress: 2.19 %, local time: 22:50\n",
      "Total elapsed time is 25.73 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 1500 contigs out of 45613 done, took 780.77 s\n",
      "Progress: 3.29 %, local time: 23:3\n",
      "Total elapsed time is 38.74 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 2000 contigs out of 45613 done, took 851.76 s\n",
      "Progress: 4.38 %, local time: 23:17\n",
      "Total elapsed time is 52.94 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 2500 contigs out of 45613 done, took 1008.50 s\n",
      "Progress: 5.48 %, local time: 23:34\n",
      "Total elapsed time is 69.75 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 3000 contigs out of 45613 done, took 963.19 s\n",
      "Progress: 6.58 %, local time: 23:50\n",
      "Total elapsed time is 85.81 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 3500 contigs out of 45613 done, took 931.38 s\n",
      "Progress: 7.67 %, local time: 0:5\n",
      "Total elapsed time is 101.34 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 4000 contigs out of 45613 done, took 834.46 s\n",
      "Progress: 8.77 %, local time: 0:19\n",
      "Total elapsed time is 115.25 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 4500 contigs out of 45613 done, took 784.07 s\n",
      "Progress: 9.87 %, local time: 0:32\n",
      "Total elapsed time is 128.33 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 5000 contigs out of 45613 done, took 754.28 s\n",
      "Progress: 10.96 %, local time: 0:45\n",
      "Total elapsed time is 140.90 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 5500 contigs out of 45613 done, took 758.36 s\n",
      "Progress: 12.06 %, local time: 0:57\n",
      "Total elapsed time is 153.55 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 6000 contigs out of 45613 done, took 767.04 s\n",
      "Progress: 13.15 %, local time: 1:10\n",
      "Total elapsed time is 166.34 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 6500 contigs out of 45613 done, took 765.41 s\n",
      "Progress: 14.25 %, local time: 1:23\n",
      "Total elapsed time is 179.10 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 7000 contigs out of 45613 done, took 771.02 s\n",
      "Progress: 15.35 %, local time: 1:36\n",
      "Total elapsed time is 191.96 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 7500 contigs out of 45613 done, took 788.60 s\n",
      "Progress: 16.44 %, local time: 1:49\n",
      "Total elapsed time is 205.11 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 8000 contigs out of 45613 done, took 789.20 s\n",
      "Progress: 17.54 %, local time: 2:2\n",
      "Total elapsed time is 218.27 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 8500 contigs out of 45613 done, took 782.31 s\n",
      "Progress: 18.64 %, local time: 2:15\n",
      "Total elapsed time is 231.32 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 9000 contigs out of 45613 done, took 790.70 s\n",
      "Progress: 19.73 %, local time: 2:28\n",
      "Total elapsed time is 244.50 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 9500 contigs out of 45613 done, took 791.99 s\n",
      "Progress: 20.83 %, local time: 2:42\n",
      "Total elapsed time is 257.71 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 10000 contigs out of 45613 done, took 822.11 s\n",
      "Progress: 21.92 %, local time: 2:55\n",
      "Total elapsed time is 271.42 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 10500 contigs out of 45613 done, took 813.14 s\n",
      "Progress: 23.02 %, local time: 3:9\n",
      "Total elapsed time is 284.99 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------\n",
      "Collecting kmers for 11000 contigs out of 45613 done, took 826.96 s\n",
      "Progress: 24.12 %, local time: 3:23\n",
      "Total elapsed time is 298.78 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 11500 contigs out of 45613 done, took 807.19 s\n",
      "Progress: 25.21 %, local time: 3:36\n",
      "Total elapsed time is 312.24 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 12000 contigs out of 45613 done, took 816.47 s\n",
      "Progress: 26.31 %, local time: 3:50\n",
      "Total elapsed time is 325.86 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 12500 contigs out of 45613 done, took 826.33 s\n",
      "Progress: 27.4 %, local time: 4:4\n",
      "Total elapsed time is 339.65 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 13000 contigs out of 45613 done, took 821.54 s\n",
      "Progress: 28.5 %, local time: 4:17\n",
      "Total elapsed time is 353.35 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 13500 contigs out of 45613 done, took 820.80 s\n",
      "Progress: 29.6 %, local time: 4:31\n",
      "Total elapsed time is 367.05 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 14000 contigs out of 45613 done, took 837.86 s\n",
      "Progress: 30.69 %, local time: 4:45\n",
      "Total elapsed time is 381.02 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 14500 contigs out of 45613 done, took 1551.85 s\n",
      "Progress: 31.79 %, local time: 5:11\n",
      "Total elapsed time is 406.90 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 15000 contigs out of 45613 done, took 1267.83 s\n",
      "Progress: 32.89 %, local time: 5:32\n",
      "Total elapsed time is 428.05 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 15500 contigs out of 45613 done, took 1111.31 s\n",
      "Progress: 33.98 %, local time: 5:50\n",
      "Total elapsed time is 446.58 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 16000 contigs out of 45613 done, took 1048.34 s\n",
      "Progress: 35.08 %, local time: 6:8\n",
      "Total elapsed time is 464.07 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 16500 contigs out of 45613 done, took 1010.81 s\n",
      "Progress: 36.17 %, local time: 6:25\n",
      "Total elapsed time is 480.93 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 17000 contigs out of 45613 done, took 989.56 s\n",
      "Progress: 37.27 %, local time: 6:41\n",
      "Total elapsed time is 497.44 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 17500 contigs out of 45613 done, took 970.17 s\n",
      "Progress: 38.37 %, local time: 6:58\n",
      "Total elapsed time is 513.63 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 18000 contigs out of 45613 done, took 955.15 s\n",
      "Progress: 39.46 %, local time: 7:13\n",
      "Total elapsed time is 529.57 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 18500 contigs out of 45613 done, took 942.41 s\n",
      "Progress: 40.56 %, local time: 7:29\n",
      "Total elapsed time is 545.29 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 19000 contigs out of 45613 done, took 928.31 s\n",
      "Progress: 41.65 %, local time: 7:45\n",
      "Total elapsed time is 560.78 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 19500 contigs out of 45613 done, took 917.23 s\n",
      "Progress: 42.75 %, local time: 8:0\n",
      "Total elapsed time is 576.09 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 20000 contigs out of 45613 done, took 912.54 s\n",
      "Progress: 43.85 %, local time: 8:15\n",
      "Total elapsed time is 591.32 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 20500 contigs out of 45613 done, took 913.47 s\n",
      "Progress: 44.94 %, local time: 8:30\n",
      "Total elapsed time is 606.56 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 21000 contigs out of 45613 done, took 1018.61 s\n",
      "Progress: 46.04 %, local time: 8:47\n",
      "Total elapsed time is 623.56 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 21500 contigs out of 45613 done, took 927.23 s\n",
      "Progress: 47.14 %, local time: 9:3\n",
      "Total elapsed time is 639.04 minutes\n",
      "Saving current contig kmers to pickle...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 22000 contigs out of 45613 done, took 942.30 s\n",
      "Progress: 48.23 %, local time: 9:19\n",
      "Total elapsed time is 654.76 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 22500 contigs out of 45613 done, took 987.99 s\n",
      "Progress: 49.33 %, local time: 9:35\n",
      "Total elapsed time is 671.25 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 23000 contigs out of 45613 done, took 1002.79 s\n",
      "Progress: 50.42 %, local time: 9:52\n",
      "Total elapsed time is 687.99 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 23500 contigs out of 45613 done, took 3066.25 s\n",
      "Progress: 51.52 %, local time: 10:43\n",
      "Total elapsed time is 739.12 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 24000 contigs out of 45613 done, took 1062.02 s\n",
      "Progress: 52.62 %, local time: 11:1\n",
      "Total elapsed time is 756.84 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 24500 contigs out of 45613 done, took 1006.29 s\n",
      "Progress: 53.71 %, local time: 11:18\n",
      "Total elapsed time is 773.64 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 25000 contigs out of 45613 done, took 1173.79 s\n",
      "Progress: 54.81 %, local time: 11:37\n",
      "Total elapsed time is 793.23 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 25500 contigs out of 45613 done, took 1221.24 s\n",
      "Progress: 55.91 %, local time: 11:57\n",
      "Total elapsed time is 813.61 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 26000 contigs out of 45613 done, took 1181.54 s\n",
      "Progress: 57.0 %, local time: 12:17\n",
      "Total elapsed time is 833.33 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 26500 contigs out of 45613 done, took 980.42 s\n",
      "Progress: 58.1 %, local time: 12:34\n",
      "Total elapsed time is 849.70 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 27000 contigs out of 45613 done, took 991.60 s\n",
      "Progress: 59.19 %, local time: 12:50\n",
      "Total elapsed time is 866.25 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 27500 contigs out of 45613 done, took 1026.58 s\n",
      "Progress: 60.29 %, local time: 13:7\n",
      "Total elapsed time is 883.39 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 28000 contigs out of 45613 done, took 1024.21 s\n",
      "Progress: 61.39 %, local time: 13:24\n",
      "Total elapsed time is 900.49 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 28500 contigs out of 45613 done, took 979.40 s\n",
      "Progress: 62.48 %, local time: 13:41\n",
      "Total elapsed time is 916.85 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 29000 contigs out of 45613 done, took 940.97 s\n",
      "Progress: 63.58 %, local time: 13:56\n",
      "Total elapsed time is 932.56 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 29500 contigs out of 45613 done, took 1799.81 s\n",
      "Progress: 64.67 %, local time: 14:26\n",
      "Total elapsed time is 962.59 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 30000 contigs out of 45613 done, took 1476.26 s\n",
      "Progress: 65.77 %, local time: 14:51\n",
      "Total elapsed time is 987.23 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 30500 contigs out of 45613 done, took 1352.07 s\n",
      "Progress: 66.87 %, local time: 15:14\n",
      "Total elapsed time is 1009.79 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 31000 contigs out of 45613 done, took 1285.56 s\n",
      "Progress: 67.96 %, local time: 15:35\n",
      "Total elapsed time is 1031.25 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 31500 contigs out of 45613 done, took 1235.81 s\n",
      "Progress: 69.06 %, local time: 15:56\n",
      "Total elapsed time is 1051.87 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 32000 contigs out of 45613 done, took 1222.91 s\n",
      "Progress: 70.16 %, local time: 16:16\n",
      "Total elapsed time is 1072.29 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------\n",
      "Collecting kmers for 32500 contigs out of 45613 done, took 1206.98 s\n",
      "Progress: 71.25 %, local time: 16:36\n",
      "Total elapsed time is 1092.44 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 33000 contigs out of 45613 done, took 1202.59 s\n",
      "Progress: 72.35 %, local time: 16:56\n",
      "Total elapsed time is 1112.51 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 33500 contigs out of 45613 done, took 1138.55 s\n",
      "Progress: 73.44 %, local time: 17:15\n",
      "Total elapsed time is 1131.52 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 34000 contigs out of 45613 done, took 1120.74 s\n",
      "Progress: 74.54 %, local time: 17:34\n",
      "Total elapsed time is 1150.23 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 34500 contigs out of 45613 done, took 1128.31 s\n",
      "Progress: 75.64 %, local time: 17:53\n",
      "Total elapsed time is 1169.07 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 35000 contigs out of 45613 done, took 1112.85 s\n",
      "Progress: 76.73 %, local time: 18:12\n",
      "Total elapsed time is 1187.65 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 35500 contigs out of 45613 done, took 2336.22 s\n",
      "Progress: 77.83 %, local time: 18:51\n",
      "Total elapsed time is 1226.63 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 36000 contigs out of 45613 done, took 1131.81 s\n",
      "Progress: 78.92 %, local time: 19:9\n",
      "Total elapsed time is 1245.52 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 36500 contigs out of 45613 done, took 1104.63 s\n",
      "Progress: 80.02 %, local time: 19:28\n",
      "Total elapsed time is 1263.97 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 37000 contigs out of 45613 done, took 1107.21 s\n",
      "Progress: 81.12 %, local time: 19:46\n",
      "Total elapsed time is 1282.46 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 37500 contigs out of 45613 done, took 1090.20 s\n",
      "Progress: 82.21 %, local time: 20:5\n",
      "Total elapsed time is 1300.67 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 38000 contigs out of 45613 done, took 1156.20 s\n",
      "Progress: 83.31 %, local time: 20:24\n",
      "Total elapsed time is 1319.97 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 38500 contigs out of 45613 done, took 1154.74 s\n",
      "Progress: 84.41 %, local time: 20:43\n",
      "Total elapsed time is 1339.26 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 39000 contigs out of 45613 done, took 1162.60 s\n",
      "Progress: 85.5 %, local time: 21:3\n",
      "Total elapsed time is 1358.67 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 39500 contigs out of 45613 done, took 1197.29 s\n",
      "Progress: 86.6 %, local time: 21:23\n",
      "Total elapsed time is 1378.67 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 40000 contigs out of 45613 done, took 1184.62 s\n",
      "Progress: 87.69 %, local time: 21:42\n",
      "Total elapsed time is 1398.45 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 40500 contigs out of 45613 done, took 1107.77 s\n",
      "Progress: 88.79 %, local time: 22:1\n",
      "Total elapsed time is 1416.96 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 41000 contigs out of 45613 done, took 1066.54 s\n",
      "Progress: 89.89 %, local time: 22:19\n",
      "Total elapsed time is 1434.77 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 41500 contigs out of 45613 done, took 1077.99 s\n",
      "Progress: 90.98 %, local time: 22:37\n",
      "Total elapsed time is 1452.78 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 42000 contigs out of 45613 done, took 1080.65 s\n",
      "Progress: 92.08 %, local time: 22:55\n",
      "Total elapsed time is 1470.83 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 42500 contigs out of 45613 done, took 1093.27 s\n",
      "Progress: 93.18 %, local time: 23:13\n",
      "Total elapsed time is 1489.09 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 43000 contigs out of 45613 done, took 1082.50 s\n",
      "Progress: 94.27 %, local time: 23:31\n",
      "Total elapsed time is 1507.18 minutes\n",
      "Saving current contig kmers to pickle...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 43500 contigs out of 45613 done, took 1076.26 s\n",
      "Progress: 95.37 %, local time: 23:49\n",
      "Total elapsed time is 1525.16 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 44000 contigs out of 45613 done, took 1188.74 s\n",
      "Progress: 96.46 %, local time: 0:9\n",
      "Total elapsed time is 1545.01 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 44500 contigs out of 45613 done, took 1219.45 s\n",
      "Progress: 97.56 %, local time: 0:29\n",
      "Total elapsed time is 1565.38 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 45000 contigs out of 45613 done, took 1174.81 s\n",
      "Progress: 98.66 %, local time: 0:49\n",
      "Total elapsed time is 1585.01 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "----------------------------------------------------\n",
      "Collecting kmers for 45500 contigs out of 45613 done, took 1180.63 s\n",
      "Progress: 99.75 %, local time: 1:9\n",
      "Total elapsed time is 1604.73 minutes\n",
      "Saving current contig kmers to pickle...\n",
      "relative path = Permafrost/WestDock/pickles/kmerdf_from_metagenome_fasta_6mer_incomplete.pickle\n",
      "----------------------------------------------------\n",
      "Constructing 6-mer table from super_contigs.PermafrostWestDock.fasta took 26.82 hours\n",
      "Finished at local time: 1:13\n"
     ]
    }
   ],
   "source": [
    "# compute WD kmers\n",
    "fastapath = 'Permafrost/WestDock/super_contigs.PermafrostWestDock.fasta'\n",
    "savedir = 'Permafrost/WestDock/pickles/'\n",
    "contig6mers,k6mers  = mock.make_kmertable_from_fasta_contigs(fastapath,6,savedir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# writing WD fastas of PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "imp.reload(pd)\n",
    "WDtsnedf3kb = pd.read_pickle('Permafrost/WestDock/pickles/WDminichipCov3kbthres6mers_all_tSNEs')\n",
    "WDtsnedf5kb = pd.read_pickle('Permafrost/WestDock/pickles/WDminichipCov5kbthres6mers_all_tSNEs')\n",
    "WDtsnedf7kb = pd.read_pickle('Permafrost/WestDock/pickles/WDminichipCov7kbthres6mers_all_tSNEs')\n",
    "WDtsnedf10kb = pd.read_pickle('Permafrost/WestDock/pickles/WDminichipCov10kbthres6mers_all_tSNEs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x_160mers</th>\n",
       "      <th>y_160mers</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>GC</th>\n",
       "      <th>x_PC4</th>\n",
       "      <th>y_PC4</th>\n",
       "      <th>x_PC6</th>\n",
       "      <th>y_PC6</th>\n",
       "      <th>x_PC9</th>\n",
       "      <th>y_PC9</th>\n",
       "      <th>...</th>\n",
       "      <th>x_PC26</th>\n",
       "      <th>y_PC26</th>\n",
       "      <th>x_PC38</th>\n",
       "      <th>y_PC38</th>\n",
       "      <th>x_PC55</th>\n",
       "      <th>y_PC55</th>\n",
       "      <th>x_PC78</th>\n",
       "      <th>y_PC78</th>\n",
       "      <th>x_PC113</th>\n",
       "      <th>y_PC113</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SuperContig_NODE_1352_length_18628_cov_1.96179</th>\n",
       "      <td>31.575850</td>\n",
       "      <td>3.469471</td>\n",
       "      <td>18628</td>\n",
       "      <td>0.33</td>\n",
       "      <td>-18.332090</td>\n",
       "      <td>-20.095514</td>\n",
       "      <td>-3.475165</td>\n",
       "      <td>-29.315403</td>\n",
       "      <td>-27.643469</td>\n",
       "      <td>-3.993829</td>\n",
       "      <td>...</td>\n",
       "      <td>-29.258291</td>\n",
       "      <td>-22.604109</td>\n",
       "      <td>6.217392</td>\n",
       "      <td>53.820469</td>\n",
       "      <td>-26.478449</td>\n",
       "      <td>-18.334024</td>\n",
       "      <td>38.173321</td>\n",
       "      <td>-39.896446</td>\n",
       "      <td>21.021051</td>\n",
       "      <td>-39.888367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SuperContig_NODE_8940_length_7683_cov_0.132911</th>\n",
       "      <td>-43.299213</td>\n",
       "      <td>19.329697</td>\n",
       "      <td>7683</td>\n",
       "      <td>0.41</td>\n",
       "      <td>-11.451790</td>\n",
       "      <td>68.012817</td>\n",
       "      <td>-2.638265</td>\n",
       "      <td>55.267048</td>\n",
       "      <td>28.569019</td>\n",
       "      <td>46.891644</td>\n",
       "      <td>...</td>\n",
       "      <td>17.016922</td>\n",
       "      <td>35.047127</td>\n",
       "      <td>-20.102636</td>\n",
       "      <td>30.415081</td>\n",
       "      <td>1.964772</td>\n",
       "      <td>27.980080</td>\n",
       "      <td>17.766685</td>\n",
       "      <td>27.913815</td>\n",
       "      <td>-44.263035</td>\n",
       "      <td>45.189808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SuperContig_NODE_15749_length_5217_cov_0.446463</th>\n",
       "      <td>28.099161</td>\n",
       "      <td>16.418409</td>\n",
       "      <td>5217</td>\n",
       "      <td>0.52</td>\n",
       "      <td>90.308121</td>\n",
       "      <td>3.982028</td>\n",
       "      <td>84.414833</td>\n",
       "      <td>24.387733</td>\n",
       "      <td>81.602951</td>\n",
       "      <td>16.899786</td>\n",
       "      <td>...</td>\n",
       "      <td>78.420700</td>\n",
       "      <td>36.667538</td>\n",
       "      <td>76.324562</td>\n",
       "      <td>41.743462</td>\n",
       "      <td>68.595009</td>\n",
       "      <td>49.237881</td>\n",
       "      <td>62.431118</td>\n",
       "      <td>50.118202</td>\n",
       "      <td>18.289827</td>\n",
       "      <td>-27.063845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SuperContig_NODE_8751_length_7787_cov_54.6094</th>\n",
       "      <td>18.064379</td>\n",
       "      <td>0.208569</td>\n",
       "      <td>7787</td>\n",
       "      <td>0.41</td>\n",
       "      <td>-36.446644</td>\n",
       "      <td>-11.303999</td>\n",
       "      <td>-15.265374</td>\n",
       "      <td>-18.893314</td>\n",
       "      <td>-14.107127</td>\n",
       "      <td>5.644021</td>\n",
       "      <td>...</td>\n",
       "      <td>2.124321</td>\n",
       "      <td>-5.573134</td>\n",
       "      <td>4.644785</td>\n",
       "      <td>1.810426</td>\n",
       "      <td>-0.492613</td>\n",
       "      <td>-28.351114</td>\n",
       "      <td>5.010416</td>\n",
       "      <td>-18.267050</td>\n",
       "      <td>25.398792</td>\n",
       "      <td>24.211378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SuperContig_NODE_2002_length_15947_cov_0.98025</th>\n",
       "      <td>2.331565</td>\n",
       "      <td>2.761369</td>\n",
       "      <td>15947</td>\n",
       "      <td>0.47</td>\n",
       "      <td>36.319847</td>\n",
       "      <td>1.037394</td>\n",
       "      <td>45.580162</td>\n",
       "      <td>-53.018169</td>\n",
       "      <td>15.057109</td>\n",
       "      <td>-38.784721</td>\n",
       "      <td>...</td>\n",
       "      <td>66.262909</td>\n",
       "      <td>-22.474421</td>\n",
       "      <td>13.154781</td>\n",
       "      <td>20.223690</td>\n",
       "      <td>21.144352</td>\n",
       "      <td>-10.222426</td>\n",
       "      <td>29.156239</td>\n",
       "      <td>18.623459</td>\n",
       "      <td>17.449423</td>\n",
       "      <td>-18.079554</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 x_160mers  y_160mers  \\\n",
       "Unnamed: 0                                                              \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179   31.575850   3.469471   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911  -43.299213  19.329697   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  28.099161  16.418409   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094    18.064379   0.208569   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025    2.331565   2.761369   \n",
       "\n",
       "                                                 Sequence length    GC  \\\n",
       "Unnamed: 0                                                               \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179             18628  0.33   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911              7683  0.41   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463             5217  0.52   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094               7787  0.41   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025             15947  0.47   \n",
       "\n",
       "                                                     x_PC4      y_PC4  \\\n",
       "Unnamed: 0                                                              \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179  -18.332090 -20.095514   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911  -11.451790  68.012817   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  90.308121   3.982028   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094   -36.446644 -11.303999   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025   36.319847   1.037394   \n",
       "\n",
       "                                                     x_PC6      y_PC6  \\\n",
       "Unnamed: 0                                                              \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179   -3.475165 -29.315403   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911   -2.638265  55.267048   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  84.414833  24.387733   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094   -15.265374 -18.893314   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025   45.580162 -53.018169   \n",
       "\n",
       "                                                     x_PC9      y_PC9  ...  \\\n",
       "Unnamed: 0                                                             ...   \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179  -27.643469  -3.993829  ...   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911   28.569019  46.891644  ...   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  81.602951  16.899786  ...   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094   -14.107127   5.644021  ...   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025   15.057109 -38.784721  ...   \n",
       "\n",
       "                                                    x_PC26     y_PC26  \\\n",
       "Unnamed: 0                                                              \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179  -29.258291 -22.604109   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911   17.016922  35.047127   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  78.420700  36.667538   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094     2.124321  -5.573134   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025   66.262909 -22.474421   \n",
       "\n",
       "                                                    x_PC38     y_PC38  \\\n",
       "Unnamed: 0                                                              \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179    6.217392  53.820469   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911  -20.102636  30.415081   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  76.324562  41.743462   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094     4.644785   1.810426   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025   13.154781  20.223690   \n",
       "\n",
       "                                                    x_PC55     y_PC55  \\\n",
       "Unnamed: 0                                                              \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179  -26.478449 -18.334024   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911    1.964772  27.980080   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  68.595009  49.237881   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094    -0.492613 -28.351114   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025   21.144352 -10.222426   \n",
       "\n",
       "                                                    x_PC78     y_PC78  \\\n",
       "Unnamed: 0                                                              \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179   38.173321 -39.896446   \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911   17.766685  27.913815   \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  62.431118  50.118202   \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094     5.010416 -18.267050   \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025   29.156239  18.623459   \n",
       "\n",
       "                                                   x_PC113    y_PC113  \n",
       "Unnamed: 0                                                             \n",
       "SuperContig_NODE_1352_length_18628_cov_1.96179   21.021051 -39.888367  \n",
       "SuperContig_NODE_8940_length_7683_cov_0.132911  -44.263035  45.189808  \n",
       "SuperContig_NODE_15749_length_5217_cov_0.446463  18.289827 -27.063845  \n",
       "SuperContig_NODE_8751_length_7787_cov_54.6094    25.398792  24.211378  \n",
       "SuperContig_NODE_2002_length_15947_cov_0.98025   17.449423 -18.079554  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "WDtsnedf3kb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "imp.reload(minimeta)\n",
    "\"\"\"\n",
    "minCS ['min_cluster_size']\n",
    "minS ['min_samples']\n",
    "CSM ['cluster_selection_method']\n",
    "ASC ['allow_single_cluster']\n",
    "perp ['perplexity_na'] # this is a string of the form '40_a' or '60_n' (so perplexity and absolute or normalized tsne)\n",
    "expt_name ['expt_name']\n",
    "write_fasta ['write_fasta'] # save the fasta Y/n (as in: do dry run clustering first, as fasta writing takes time)\n",
    "writedf ['write_df']\n",
    "series ['is_this_a_series']\n",
    "\"\"\"\n",
    "minCS = 40\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'leaf',False,'PC13','FranklinBluffs_miniPC13','YES','YES','NO','Permafrost/FranklinBluffs/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'FranklinBluffs_miniPC'+pc+'_5kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(maindf_mini5,**kwargs) # cluster whatever pcs you want here.\n",
    "minCS = 30\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'leaf',False,'PC13','FranklinBluffs_miniPC13','YES','YES','NO','Permafrost/FranklinBluffs/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'FranklinBluffs_miniPC'+pc+'_5kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(maindf_mini5,**kwargs) # cluster whatever pcs you want here.\n",
    "\n",
    "minCS = 20\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'leaf',False,'PC13','FranklinBluffs_miniPC13','YES','YES','NO','Permafrost/FranklinBluffs/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'FranklinBluffs_miniPC'+pc+'_10kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(maindf_mini10,**kwargs) # cluster whatever pcs you want here.\n",
    "\n",
    "minCS = 30\n",
    "keys = ['min_cluster_size','min_samples','cluster_selection_method','allow_single_cluster'\n",
    "             ,'perplexity_na','expt_name','write_fasta','write_df','is_this_a_series','outdir']\n",
    "values = [minCS,1,'leaf',False,'PC13','FranklinBluffs_miniPC13','YES','YES','NO','Permafrost/FranklinBluffs/']\n",
    "kwargs = dict(zip(keys,values))\n",
    "for pcs in [9,13,18,26,38,55,78,113]:\n",
    "    pc = str(pcs)\n",
    "    kwargs['perplexity_na'] = 'PC'+pc\n",
    "    kwargs['expt_name'] = 'FranklinBluffs_miniPC'+pc+'_10kbthres_minCS'+str(minCS)\n",
    "    maindf_mini = minimeta.cluster_main_tsne(maindf_mini10,**kwargs) # cluster whatever pcs you want here.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
