{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# brian's genome bins\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import csv\n",
    "import matplotlib.cm as cm\n",
    "import seaborn as sns\n",
    "from matplotlib import gridspec\n",
    "import matplotlib.animation as manimation\n",
    "import matplotlib.patches as patches\n",
    "import time\n",
    "from collections import Counter\n",
    "import sklearn.cluster as cluster\n",
    "import time\n",
    "import hdbscan\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import re\n",
    "\n",
    "#import plotly. as py\n",
    "#import plotly.graph_objs as go\n",
    "font = {'family' : 'DejaVu Sans',\n",
    "        'weight' : 'regular',\n",
    "        'size'   : 16}\n",
    "\n",
    "plt.rc('font', **font)\n",
    "plt.rc('lines',lw=2) \n",
    "sns.set(font_scale = 2)\n",
    "\n",
    "# dereks function for optimal leaf clustering\n",
    "def _optimal_order(data, **kwargs):\n",
    "    \"\"\" Optimal leaf ordering\n",
    "        **kwargs passed to pdist e.g. metric='correlation'\n",
    "    \"\"\"\n",
    "    d = pdist(data, **kwargs)\n",
    "    link = linkage(d, method='average')\n",
    "    optimal_order = optimal_leaf_ordering(link, d)\n",
    "    return optimal_order\n",
    "\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from fastcluster import linkage\n",
    "from polo import optimal_leaf_ordering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# make sure this is up-to-date: see the separate notebook for that \n",
    "# current version: Oct 2017.\n",
    "expt_name = 'brian'\n",
    "KEGGmoddf = pd.read_pickle('KEGG_modules_df')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# make csv of .KO file first (and add column titles, it's easier this way)\n",
    "keggdata = list(csv.reader(open('/home/bojk/Data/BrianDataSets/119436.assembled.csv')))\n",
    "keggdf = pd.DataFrame.from_records(keggdata,columns=keggdata[0])\n",
    "keggdf = keggdf.loc[1:]\n",
    "keggdf = keggdf.reset_index(drop=True)\n",
    "keggdf['contig'] = keggdf['contig ID'].str[10:16] # the 6-digit number\n",
    "keggdf['Contig_ID'] = keggdf['contig ID'].str[0:16] # the official IMG contig name\n",
    "keggdf.to_pickle(expt_name+'_KEGG_raw')\n",
    "\n",
    "# load the metafile \n",
    "phyldata = list(csv.reader(open('/home/bojk/Data/BrianDataSets/2017.10.19_grouped_genome_information.csv')))\n",
    "metadata = pd.DataFrame.from_records(phyldata,columns=phyldata[0])\n",
    "metadata = metadata[1:]\n",
    "metadata = metadata.set_index('Contig_ID')\n",
    "\n",
    "# add the genome ID to each KO term\n",
    "keggdf['genome'] = ''\n",
    "keg2 = keggdf.set_index('Contig_ID')\n",
    "keg2['genome']=metadata['Genome_Number']\n",
    "\n",
    "# make KO dataframe containing only the binned contigs \n",
    "KO_bins = keg2.loc[keg2['genome'].dropna().index.unique()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7994040638537204"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fraction clustered into genomes:\n",
    "len(keg2['genome'].dropna())/len(keg2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# make module with KO-terms per genome bin \n",
    "Klist = pd.DataFrame(columns=KO_bins['genome'].unique())\n",
    "knum = np.zeros(len(Klist.columns))\n",
    "bins = KO_bins['genome'].unique()\n",
    "for i in range(0,len(Klist.columns)):\n",
    "    knum[i] = len(KO_bins[KO_bins['genome']==bins[i]]['KO_number'])\n",
    "Klist = pd.DataFrame(index=list(range(0,int(max(knum)))),columns=KO_bins['genome'].unique())    \n",
    "j=0\n",
    "for i in Klist.columns:\n",
    "    Klist.loc[0:(knum[j]-1), i] = KO_bins[KO_bins['genome']==i]['KO_number'].values\n",
    "    j+=1\n",
    "\n",
    "# save to txt\n",
    "for i in Klist.columns:\n",
    "    if int(i)<10:\n",
    "        Klist[i].to_csv('Knumbers/brian/brian0'+i+'.txt',sep='\\t')\n",
    "    else:\n",
    "        Klist[i].to_csv('Knumbers/brian/brian'+i+'.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# time to start running the KEGG crawler\n",
    "from urllib.request import Request, urlopen, URLError\n",
    "from urllib.parse import urlencode\n",
    "from lxml import etree\n",
    "import lxml\n",
    "import html2text as h2t\n",
    "from html.parser import HTMLParser\n",
    "\n",
    "class myhtmlparser(HTMLParser):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        self.reset()\n",
    "\n",
    "        self.NEWTAGS = []\n",
    "        self.NEWATTRS = []\n",
    "        self.HTMLDATA = []\n",
    "    def handle_starttag(self, tag, attrs):\n",
    "        self.NEWTAGS.append(tag)\n",
    "        self.NEWATTRS.append(attrs)\n",
    "    def handle_data(self, data):\n",
    "        self.HTMLDATA.append(data)\n",
    "    def clean(self):\n",
    "        self.NEWTAGS = []\n",
    "        self.NEWATTRS = []\n",
    "        self.HTMLDATA = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def retrieveKEGGmodules(typedata):\n",
    "    #typedata can be: 'all','complete', or 'conservative'\n",
    "    if typedata == 'conservative':\n",
    "        mode = 'complete+ng1'\n",
    "    else:\n",
    "        mode = typedata\n",
    "        \n",
    "    # make dataframe\n",
    "    Mlist = pd.DataFrame(columns=Klist.columns)\n",
    "    # make a datafame containing the KEGG entries for known hits. This is incomplete!\n",
    "    # as I have hits not showing up in the known genomes. \n",
    "    # edit Oct 2017: just make a range of len(max(Modules in KEGG database))\n",
    "    dta = list(range(0,844))\n",
    "    dta2 = ['M0000'+str(dta[s]) for s in range(0,\n",
    "                        10)]+['M000'+str(dta[s]) for s in range(10,\n",
    "                            100)]+['M00'+str(dta[s]) for s in range(100,len(dta))]\n",
    "    #Mmap = pd.DataFrame(index=known_genomes.index,columns=Klist.columns)\n",
    "    Mmap = pd.DataFrame(index=dta2,columns=Klist.columns)\n",
    "    \n",
    "    a=0\n",
    "    start_time = time.time()   \n",
    "    for s in Klist.columns:\n",
    "\n",
    "        #get list of KEGG genes for a given cluster (tab-delimited text)\n",
    "        if int(s)<10:\n",
    "            with open('Knumbers/brian/brian0'+s+'.txt', 'rt') as f:\n",
    "                dat = f.read()\n",
    "        else:\n",
    "            with open('Knumbers/brian/brian'+s+'.txt', 'rt') as f:\n",
    "                dat = f.read()\n",
    "\n",
    "        # retrieve data from KEGG database\n",
    "        url = 'http://www.genome.jp/kegg-bin/find_module_object'\n",
    "        params = urlencode({'unclassified': dat, 'mode': mode}).encode()\n",
    "        html = urlopen(url, params).read()\n",
    "\n",
    "\n",
    "        # save as html file for reference purposes (and look at detailed output)\n",
    "        with open('/home/bojk/Data/Knumbers/brian/html/'+s+'_'+typedata+'.html', 'wb') as f:\n",
    "            f.write(html)\n",
    "\n",
    "        # convert data to list\n",
    "        pstring = html.decode('utf8') \n",
    "        parser = myhtmlparser()\n",
    "        parser.feed(pstring)\n",
    "\n",
    "        # Extract data from parser\n",
    "        tags  = parser.NEWTAGS\n",
    "        attrs = parser.NEWATTRS\n",
    "        data  = parser.HTMLDATA\n",
    "\n",
    "        # Clean the parser\n",
    "        parser.clean()\n",
    "\n",
    "        # extract Module numbers and feed to dataframe\n",
    "        indices = [i for i, s in enumerate(data) if 'M00' in s]\n",
    "        mdata = [data[s] for s in indices]\n",
    "        mdatadf = pd.DataFrame(mdata,columns=[s])\n",
    "\n",
    "        if a==0:\n",
    "            Mlist = mdatadf.copy()\n",
    "            a=1\n",
    "        else:\n",
    "            Mlist = Mlist.join(mdatadf,how='outer')\n",
    "        Mmap.loc[mdatadf.set_index(s).index,s] = 1\n",
    "    \n",
    "    end_time = time.time()            \n",
    "    print('Retrieving modules from KEGG database took {:.2f} s'.format(end_time - start_time)) \n",
    "    Mlist.to_pickle('Knumbers/brian/brian_genome_module_list_'+typedata)\n",
    "    Mmap.to_pickle('Knumbers/brian/brian_genome_module_map_'+typedata)\n",
    "    return (Mlist,Mmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving modules from KEGG database took 161.89 s\n"
     ]
    }
   ],
   "source": [
    "Mlist_all,Mmap_all = retrieveKEGGmodules('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving modules from KEGG database took 113.44 s\n"
     ]
    }
   ],
   "source": [
    "Mlist_cons,Mmap_cons = retrieveKEGGmodules('conservative')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving modules from KEGG database took 80.76 s\n"
     ]
    }
   ],
   "source": [
    "Mlist_comp,Mmap_comp = retrieveKEGGmodules('complete')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# now we need a metadata file (per genome, the stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.6/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "# get some numeric metadata\n",
    "metadata['Contig_Length'] = metadata['Contig_Length'].astype(int)\n",
    "metadata['Gene_Count'] = metadata['Gene_Count'].astype(int)\n",
    "metadata['Bulk_Coverage'] = metadata['Bulk_Coverage'].astype(int)\n",
    "metadata['GC'] = metadata['GC'].astype(float)\n",
    "metadata.groupby('Genome_Number').sum()\n",
    "metadf = pd.DataFrame(index=metadata.groupby('Genome_Number').sum().index,columns=['numofcontigs',       \n",
    "                                                                            'totlength','meanGC','GeneCount','mediancov',\n",
    "                                                                                'Domain', 'Phylum', 'Class', 'Order',\n",
    "                                                                                   'Family','Genus', 'Species'])\n",
    "metadf['totlength'] = metadata.groupby('Genome_Number').sum()['Contig_Length']\n",
    "metadf['meanGC'] = metadata.groupby('Genome_Number').mean()['GC']\n",
    "metadf['GeneCount'] = metadata.groupby('Genome_Number').sum()['Gene_Count']\n",
    "metadf['mediancov'] = metadata.groupby('Genome_Number').median()['Bulk_Coverage']\n",
    "\n",
    "# get consensus phylogenetic information (and make some plots)\n",
    "vals = [6,7,8,11]; keys= list(metadf.columns[vals])\n",
    "lut = dict(zip(keys,vals))\n",
    "\n",
    "for Bin in metadf.index:\n",
    "    df = metadata[metadata['Genome_Number']==Bin].copy()\n",
    "    f = plt.figure()\n",
    "    gs = gridspec.GridSpec(2,3)\n",
    "    \n",
    "    for l in df.columns[6:]:\n",
    "        # set the consensus phylo to the one with max hits\n",
    "        metadf.loc[Bin,l] = df[l].value_counts().index[0]\n",
    "        metadf.loc[Bin,'numofcontigs'] = len(df)\n",
    "        \n",
    "        # do some plotting for reference purposes \n",
    "        if l==keys[0]:\n",
    "            ax1 = f.add_subplot(gs[0,0])\n",
    "            df[l].value_counts(ascending=True).plot(kind='barh',ax=ax1)\n",
    "            plt.title(l)\n",
    "        if l==keys[1]:\n",
    "            ax2 = f.add_subplot(gs[0,2])\n",
    "            df[l].value_counts(ascending=True).plot(kind='barh',ax=ax2)\n",
    "            plt.title(l)\n",
    "        if l==keys[2]:\n",
    "            ax3 = f.add_subplot(gs[1,0])\n",
    "            df[l].value_counts(ascending=True).plot(kind='barh',ax=ax3)\n",
    "            plt.title(l)\n",
    "        if l==keys[3]:\n",
    "            ax4 = f.add_subplot(gs[1,2])\n",
    "            df[l].value_counts(ascending=True).plot(kind='barh',ax=ax4)\n",
    "            plt.title(l)\n",
    "    f.set_figwidth(20)\n",
    "    f.set_figheight(20)\n",
    "    f.savefig('BrianDataSets/figures/PhyloRankGenome_'+Bin)\n",
    "    f.clf()\n",
    "        #df.lineage.value_counts(ascending=True).plot(kind='barh',logx=True,color=barcolorsdom)\n",
    "metadf.to_pickle('BrianDataSets/metadata_df')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# do some coloring \n",
    "# create separate dataframe that assigns a value to each hit based on domain (for colored heatmap)\n",
    "Mmap_all_col = Mmap_all.copy()\n",
    "Mmap_comp_col = Mmap_comp.copy()\n",
    "Mmap_cons_col = Mmap_cons.copy()\n",
    "\n",
    "        \n",
    "domlist = metadf['Domain'].value_counts().index\n",
    "phyllist = metadf['Phylum'].value_counts().index\n",
    "domcol = [.5,1,-1]\n",
    "lut = dict(zip(domlist,domcol))\n",
    "\n",
    "# assign heatmap value (i.e. color) per domain (can do for phylum as well)\n",
    "for k in domlist:\n",
    "        Mmap_all_col.loc[:,metadf[metadf['Domain']==k].index] = Mmap_all_col.loc[:,metadf[metadf['Domain']==k].index]*lut[k]\n",
    "        Mmap_cons_col.loc[:,metadf[metadf['Domain']==k].index] = Mmap_cons_col.loc[:,metadf[metadf['Domain']==k].index]*lut[k]\n",
    "        Mmap_comp_col.loc[:,metadf[metadf['Domain']==k].index] = Mmap_comp_col.loc[:,metadf[metadf['Domain']==k].index]*lut[k]\n",
    "        \n",
    "Mmap_all_col.to_pickle('Knumbers/brian/brian_module_map_all_col')\n",
    "Mmap_cons_col.to_pickle('Knumbers/brian/brian_module_map_cons_col')\n",
    "Mmap_comp_col.to_pickle('Knumbers/brian/brian_module_map_comp_col')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Proteobacteria', 'Unassigned', 'Actinobacteria', 'Deferribacteres',\n",
       "       'Firmicutes', 'Thermotogae', 'Ignavibacteriae', 'Synergistetes',\n",
       "       'Cyanobacteria', 'Spirochaetes', 'Verrucomicrobia', 'Bacteroidetes',\n",
       "       'Aquificae', 'Euryarchaeota'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phyllist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Phylum', 'Class', 'Order', 'Species'], dtype='object')"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>numofcontigs</th>\n",
       "      <th>totlength</th>\n",
       "      <th>meanGC</th>\n",
       "      <th>GeneCount</th>\n",
       "      <th>mediancov</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome_Number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>209</td>\n",
       "      <td>4172654</td>\n",
       "      <td>0.328230</td>\n",
       "      <td>4628</td>\n",
       "      <td>1194.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Aquificae</td>\n",
       "      <td>Aquificae</td>\n",
       "      <td>Aquificales</td>\n",
       "      <td>Hydrogenothermaceae</td>\n",
       "      <td>Sulfurihydrogenibium</td>\n",
       "      <td>Sulfurihydrogenibium azorense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>62</td>\n",
       "      <td>1428612</td>\n",
       "      <td>0.597097</td>\n",
       "      <td>1503</td>\n",
       "      <td>330.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>41</td>\n",
       "      <td>925317</td>\n",
       "      <td>0.684146</td>\n",
       "      <td>939</td>\n",
       "      <td>354.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>Tepidimonas</td>\n",
       "      <td>Tepidimonas taiwanensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>16</td>\n",
       "      <td>560566</td>\n",
       "      <td>0.410000</td>\n",
       "      <td>616</td>\n",
       "      <td>6803.0</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>72</td>\n",
       "      <td>704277</td>\n",
       "      <td>0.330556</td>\n",
       "      <td>806</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>284</td>\n",
       "      <td>5449398</td>\n",
       "      <td>0.349120</td>\n",
       "      <td>5643</td>\n",
       "      <td>1596.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Deferribacteres</td>\n",
       "      <td>Deferribacteres</td>\n",
       "      <td>Deferribacterales</td>\n",
       "      <td>Deferribacteraceae</td>\n",
       "      <td>Calditerrivibrio</td>\n",
       "      <td>Calditerrivibrio nitroreducens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>61</td>\n",
       "      <td>551341</td>\n",
       "      <td>0.319344</td>\n",
       "      <td>628</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>53</td>\n",
       "      <td>593862</td>\n",
       "      <td>0.323585</td>\n",
       "      <td>591</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>338</td>\n",
       "      <td>3088625</td>\n",
       "      <td>0.629290</td>\n",
       "      <td>3312</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Desulfovibrionales</td>\n",
       "      <td>Desulfomicrobiaceae</td>\n",
       "      <td>Desulfomicrobium</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>58</td>\n",
       "      <td>561260</td>\n",
       "      <td>0.436207</td>\n",
       "      <td>694</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Thermococci</td>\n",
       "      <td>Thermococcales</td>\n",
       "      <td>Thermococcaceae</td>\n",
       "      <td>Thermococcus</td>\n",
       "      <td>Thermococcus litoralis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>60</td>\n",
       "      <td>647534</td>\n",
       "      <td>0.450000</td>\n",
       "      <td>662</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Synergistetes</td>\n",
       "      <td>Synergistia</td>\n",
       "      <td>Synergistales</td>\n",
       "      <td>Synergistaceae</td>\n",
       "      <td>Thermovirga</td>\n",
       "      <td>Thermovirga lienii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>218</td>\n",
       "      <td>3510846</td>\n",
       "      <td>0.399037</td>\n",
       "      <td>3850</td>\n",
       "      <td>31.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>197</td>\n",
       "      <td>3622682</td>\n",
       "      <td>0.341015</td>\n",
       "      <td>3687</td>\n",
       "      <td>150.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Deferribacteres</td>\n",
       "      <td>Deferribacteres</td>\n",
       "      <td>Deferribacterales</td>\n",
       "      <td>Deferribacteraceae</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>146</td>\n",
       "      <td>6323282</td>\n",
       "      <td>0.333630</td>\n",
       "      <td>6507</td>\n",
       "      <td>508.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>241</td>\n",
       "      <td>1872876</td>\n",
       "      <td>0.323071</td>\n",
       "      <td>1559</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>129</td>\n",
       "      <td>1811104</td>\n",
       "      <td>0.366822</td>\n",
       "      <td>1787</td>\n",
       "      <td>32.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>669</td>\n",
       "      <td>7244504</td>\n",
       "      <td>0.356413</td>\n",
       "      <td>6984</td>\n",
       "      <td>118.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>Chroococcales</td>\n",
       "      <td>Cyanobacteriaceae</td>\n",
       "      <td>Cyanobacterium</td>\n",
       "      <td>Cyanobacterium aponinum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>157</td>\n",
       "      <td>5700109</td>\n",
       "      <td>0.426051</td>\n",
       "      <td>6265</td>\n",
       "      <td>2678.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>281</td>\n",
       "      <td>2258064</td>\n",
       "      <td>0.687189</td>\n",
       "      <td>2403</td>\n",
       "      <td>42.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Sphingomonadales</td>\n",
       "      <td>Erythrobacteraceae</td>\n",
       "      <td>Porphyrobacter</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>158</td>\n",
       "      <td>1975403</td>\n",
       "      <td>0.343671</td>\n",
       "      <td>2202</td>\n",
       "      <td>44.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Desulfurellales</td>\n",
       "      <td>Desulfurellaceae</td>\n",
       "      <td>Desulfurella</td>\n",
       "      <td>Desulfurella acetivorans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>155</td>\n",
       "      <td>1641158</td>\n",
       "      <td>0.456000</td>\n",
       "      <td>1744</td>\n",
       "      <td>58.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiotrichales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>208</td>\n",
       "      <td>3469125</td>\n",
       "      <td>0.388462</td>\n",
       "      <td>3505</td>\n",
       "      <td>282.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogales</td>\n",
       "      <td>Fervidobacteriaceae</td>\n",
       "      <td>Fervidobacterium</td>\n",
       "      <td>Fervidobacterium pennivorans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>317</td>\n",
       "      <td>5394692</td>\n",
       "      <td>0.477855</td>\n",
       "      <td>5164</td>\n",
       "      <td>284.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Spirochaetes</td>\n",
       "      <td>Spirochaetia</td>\n",
       "      <td>Spirochaetales</td>\n",
       "      <td>Spirochaetaceae</td>\n",
       "      <td>Treponema</td>\n",
       "      <td>Treponema caldaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>103</td>\n",
       "      <td>1213567</td>\n",
       "      <td>0.334757</td>\n",
       "      <td>1306</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Campylobacteraceae</td>\n",
       "      <td>Arcobacter</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>323</td>\n",
       "      <td>5296911</td>\n",
       "      <td>0.645511</td>\n",
       "      <td>5462</td>\n",
       "      <td>3550.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>127</td>\n",
       "      <td>1740346</td>\n",
       "      <td>0.308425</td>\n",
       "      <td>1900</td>\n",
       "      <td>32.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Thermotogae</td>\n",
       "      <td>Petrotogales</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>114</td>\n",
       "      <td>1112598</td>\n",
       "      <td>0.340088</td>\n",
       "      <td>1161</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>66</td>\n",
       "      <td>604817</td>\n",
       "      <td>0.431818</td>\n",
       "      <td>681</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>163</td>\n",
       "      <td>1281619</td>\n",
       "      <td>0.592209</td>\n",
       "      <td>1408</td>\n",
       "      <td>38.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>77</td>\n",
       "      <td>1543874</td>\n",
       "      <td>0.346494</td>\n",
       "      <td>2007</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>106</td>\n",
       "      <td>1299244</td>\n",
       "      <td>0.341509</td>\n",
       "      <td>1240</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>94</td>\n",
       "      <td>1489808</td>\n",
       "      <td>0.388085</td>\n",
       "      <td>1559</td>\n",
       "      <td>18.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Epsilonproteobacteria</td>\n",
       "      <td>Campylobacterales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>169</td>\n",
       "      <td>1695910</td>\n",
       "      <td>0.582899</td>\n",
       "      <td>1788</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>113</td>\n",
       "      <td>1440098</td>\n",
       "      <td>0.596283</td>\n",
       "      <td>1631</td>\n",
       "      <td>102.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>59</td>\n",
       "      <td>561351</td>\n",
       "      <td>0.517458</td>\n",
       "      <td>637</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1257</td>\n",
       "      <td>12920865</td>\n",
       "      <td>0.686889</td>\n",
       "      <td>12296</td>\n",
       "      <td>132.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodospirillales</td>\n",
       "      <td>Rhodospirillaceae</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>83</td>\n",
       "      <td>763744</td>\n",
       "      <td>0.639157</td>\n",
       "      <td>826</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas stutzeri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>94</td>\n",
       "      <td>2585158</td>\n",
       "      <td>0.627128</td>\n",
       "      <td>2562</td>\n",
       "      <td>483.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Rhodocyclales</td>\n",
       "      <td>Rhodocyclaceae</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>87</td>\n",
       "      <td>741951</td>\n",
       "      <td>0.695287</td>\n",
       "      <td>795</td>\n",
       "      <td>94.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>Tepidimonas</td>\n",
       "      <td>Tepidimonas taiwanensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>268</td>\n",
       "      <td>2049565</td>\n",
       "      <td>0.659813</td>\n",
       "      <td>2349</td>\n",
       "      <td>44.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>116</td>\n",
       "      <td>1463421</td>\n",
       "      <td>0.538103</td>\n",
       "      <td>1571</td>\n",
       "      <td>102.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Betaproteobacteria</td>\n",
       "      <td>Burkholderiales</td>\n",
       "      <td>Comamonadaceae</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>210</td>\n",
       "      <td>1603936</td>\n",
       "      <td>0.331952</td>\n",
       "      <td>1494</td>\n",
       "      <td>55.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Ignavibacteriae</td>\n",
       "      <td>Ignavibacteria</td>\n",
       "      <td>Ignavibacteriales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>171</td>\n",
       "      <td>1180663</td>\n",
       "      <td>0.683216</td>\n",
       "      <td>1288</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Desulfovibrionales</td>\n",
       "      <td>Desulfovibrionaceae</td>\n",
       "      <td>Desulfovibrio</td>\n",
       "      <td>Desulfovibrio aminophilus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>125</td>\n",
       "      <td>2253747</td>\n",
       "      <td>0.653760</td>\n",
       "      <td>2161</td>\n",
       "      <td>424.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Chromatiaceae</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>66</td>\n",
       "      <td>1689379</td>\n",
       "      <td>0.635455</td>\n",
       "      <td>1492</td>\n",
       "      <td>832.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Verrucomicrobia</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>237</td>\n",
       "      <td>6891457</td>\n",
       "      <td>0.422236</td>\n",
       "      <td>5857</td>\n",
       "      <td>292.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>44</td>\n",
       "      <td>625347</td>\n",
       "      <td>0.620227</td>\n",
       "      <td>670</td>\n",
       "      <td>85.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas stutzeri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>106</td>\n",
       "      <td>869133</td>\n",
       "      <td>0.296792</td>\n",
       "      <td>930</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>287</td>\n",
       "      <td>2013793</td>\n",
       "      <td>0.716307</td>\n",
       "      <td>2075</td>\n",
       "      <td>36.0</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>Arenimonas</td>\n",
       "      <td>Arenimonas malthae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              numofcontigs  totlength    meanGC  GeneCount  mediancov  \\\n",
       "Genome_Number                                                           \n",
       "10                     209    4172654  0.328230       4628     1194.0   \n",
       "100                     62    1428612  0.597097       1503      330.0   \n",
       "101                     41     925317  0.684146        939      354.0   \n",
       "102                     16     560566  0.410000        616     6803.0   \n",
       "109                     72     704277  0.330556        806        2.0   \n",
       "11                     284    5449398  0.349120       5643     1596.0   \n",
       "112                     61     551341  0.319344        628        4.0   \n",
       "118                     53     593862  0.323585        591        4.0   \n",
       "12                     338    3088625  0.629290       3312       48.0   \n",
       "120                     58     561260  0.436207        694        6.0   \n",
       "126                     60     647534  0.450000        662       12.0   \n",
       "13                     218    3510846  0.399037       3850       31.0   \n",
       "14                     197    3622682  0.341015       3687      150.0   \n",
       "15                     146    6323282  0.333630       6507      508.0   \n",
       "16                     241    1872876  0.323071       1559       48.0   \n",
       "17                     129    1811104  0.366822       1787       32.0   \n",
       "18                     669    7244504  0.356413       6984      118.0   \n",
       "19                     157    5700109  0.426051       6265     2678.0   \n",
       "2                      281    2258064  0.687189       2403       42.0   \n",
       "20                     158    1975403  0.343671       2202       44.0   \n",
       "21                     155    1641158  0.456000       1744       58.0   \n",
       "22                     208    3469125  0.388462       3505      282.0   \n",
       "24                     317    5394692  0.477855       5164      284.0   \n",
       "25                     103    1213567  0.334757       1306       26.0   \n",
       "26                     323    5296911  0.645511       5462     3550.0   \n",
       "27                     127    1740346  0.308425       1900       32.0   \n",
       "28                     114    1112598  0.340088       1161       14.0   \n",
       "29                      66     604817  0.431818        681       17.0   \n",
       "3                      163    1281619  0.592209       1408       38.0   \n",
       "30                      77    1543874  0.346494       2007        0.0   \n",
       "31                     106    1299244  0.341509       1240        4.0   \n",
       "32                      94    1489808  0.388085       1559       18.0   \n",
       "34                     169    1695910  0.582899       1788       66.0   \n",
       "36                     113    1440098  0.596283       1631      102.0   \n",
       "39                      59     561351  0.517458        637       50.0   \n",
       "4                     1257   12920865  0.686889      12296      132.0   \n",
       "40                      83     763744  0.639157        826       66.0   \n",
       "41                      94    2585158  0.627128       2562      483.0   \n",
       "49                      87     741951  0.695287        795       94.0   \n",
       "5                      268    2049565  0.659813       2349       44.0   \n",
       "53                     116    1463421  0.538103       1571      102.0   \n",
       "6                      210    1603936  0.331952       1494       55.0   \n",
       "7                      171    1180663  0.683216       1288       28.0   \n",
       "71                     125    2253747  0.653760       2161      424.0   \n",
       "75                      66    1689379  0.635455       1492      832.0   \n",
       "8                      237    6891457  0.422236       5857      292.0   \n",
       "82                      44     625347  0.620227        670       85.0   \n",
       "88                     106     869133  0.296792        930        2.0   \n",
       "9                      287    2013793  0.716307       2075       36.0   \n",
       "\n",
       "                   Domain           Phylum                  Class  \\\n",
       "Genome_Number                                                       \n",
       "10               Bacteria        Aquificae              Aquificae   \n",
       "100              Bacteria   Proteobacteria     Betaproteobacteria   \n",
       "101              Bacteria   Proteobacteria     Betaproteobacteria   \n",
       "102            Unassigned       Unassigned             Unassigned   \n",
       "109              Bacteria       Unassigned             Unassigned   \n",
       "11               Bacteria  Deferribacteres        Deferribacteres   \n",
       "112              Bacteria       Firmicutes             Clostridia   \n",
       "118              Bacteria       Unassigned             Unassigned   \n",
       "12               Bacteria   Proteobacteria    Deltaproteobacteria   \n",
       "120               Archaea    Euryarchaeota            Thermococci   \n",
       "126              Bacteria    Synergistetes            Synergistia   \n",
       "13               Bacteria   Proteobacteria  Epsilonproteobacteria   \n",
       "14               Bacteria  Deferribacteres        Deferribacteres   \n",
       "15               Bacteria       Firmicutes                Bacilli   \n",
       "16               Bacteria       Unassigned             Unassigned   \n",
       "17               Bacteria       Unassigned             Unassigned   \n",
       "18               Bacteria    Cyanobacteria           unclassified   \n",
       "19               Bacteria   Proteobacteria  Epsilonproteobacteria   \n",
       "2                Bacteria   Proteobacteria    Alphaproteobacteria   \n",
       "20               Bacteria   Proteobacteria    Deltaproteobacteria   \n",
       "21               Bacteria   Proteobacteria    Gammaproteobacteria   \n",
       "22               Bacteria      Thermotogae            Thermotogae   \n",
       "24               Bacteria     Spirochaetes           Spirochaetia   \n",
       "25               Bacteria   Proteobacteria  Epsilonproteobacteria   \n",
       "26               Bacteria   Proteobacteria    Gammaproteobacteria   \n",
       "27               Bacteria      Thermotogae            Thermotogae   \n",
       "28               Bacteria       Unassigned             Unassigned   \n",
       "29               Bacteria   Proteobacteria  Epsilonproteobacteria   \n",
       "3                Bacteria   Actinobacteria         Actinobacteria   \n",
       "30             Unassigned       Unassigned             Unassigned   \n",
       "31               Bacteria       Unassigned             Unassigned   \n",
       "32               Bacteria   Proteobacteria  Epsilonproteobacteria   \n",
       "34               Bacteria   Proteobacteria    Gammaproteobacteria   \n",
       "36               Bacteria   Proteobacteria             Unassigned   \n",
       "39               Bacteria   Proteobacteria             Unassigned   \n",
       "4                Bacteria   Proteobacteria    Alphaproteobacteria   \n",
       "40               Bacteria   Proteobacteria    Gammaproteobacteria   \n",
       "41               Bacteria   Proteobacteria     Betaproteobacteria   \n",
       "49               Bacteria   Proteobacteria     Betaproteobacteria   \n",
       "5                Bacteria   Actinobacteria         Actinobacteria   \n",
       "53               Bacteria   Proteobacteria     Betaproteobacteria   \n",
       "6                Bacteria  Ignavibacteriae         Ignavibacteria   \n",
       "7                Bacteria   Proteobacteria    Deltaproteobacteria   \n",
       "71               Bacteria   Proteobacteria    Gammaproteobacteria   \n",
       "75               Bacteria  Verrucomicrobia             Unassigned   \n",
       "8                Bacteria    Bacteroidetes            Bacteroidia   \n",
       "82               Bacteria   Proteobacteria    Gammaproteobacteria   \n",
       "88               Bacteria       Unassigned             Unassigned   \n",
       "9                Bacteria   Proteobacteria    Gammaproteobacteria   \n",
       "\n",
       "                            Order               Family                 Genus  \\\n",
       "Genome_Number                                                                  \n",
       "10                    Aquificales  Hydrogenothermaceae  Sulfurihydrogenibium   \n",
       "100               Burkholderiales           Unassigned            Unassigned   \n",
       "101               Burkholderiales         unclassified           Tepidimonas   \n",
       "102                    Unassigned           Unassigned            Unassigned   \n",
       "109                    Unassigned           Unassigned            Unassigned   \n",
       "11              Deferribacterales   Deferribacteraceae      Calditerrivibrio   \n",
       "112                 Clostridiales           Unassigned            Unassigned   \n",
       "118                    Unassigned           Unassigned            Unassigned   \n",
       "12             Desulfovibrionales  Desulfomicrobiaceae      Desulfomicrobium   \n",
       "120                Thermococcales      Thermococcaceae          Thermococcus   \n",
       "126                 Synergistales       Synergistaceae           Thermovirga   \n",
       "13              Campylobacterales           Unassigned            Unassigned   \n",
       "14              Deferribacterales   Deferribacteraceae            Unassigned   \n",
       "15                     Bacillales           Unassigned            Unassigned   \n",
       "16                     Unassigned           Unassigned            Unassigned   \n",
       "17                     Unassigned           Unassigned            Unassigned   \n",
       "18                  Chroococcales    Cyanobacteriaceae        Cyanobacterium   \n",
       "19                     Unassigned           Unassigned            Unassigned   \n",
       "2                Sphingomonadales   Erythrobacteraceae        Porphyrobacter   \n",
       "20                Desulfurellales     Desulfurellaceae          Desulfurella   \n",
       "21                  Thiotrichales           Unassigned            Unassigned   \n",
       "22                  Thermotogales  Fervidobacteriaceae      Fervidobacterium   \n",
       "24                 Spirochaetales      Spirochaetaceae             Treponema   \n",
       "25              Campylobacterales   Campylobacteraceae            Arcobacter   \n",
       "26                     Unassigned           Unassigned            Unassigned   \n",
       "27                   Petrotogales         unclassified            Unassigned   \n",
       "28                     Unassigned           Unassigned            Unassigned   \n",
       "29              Campylobacterales           Unassigned            Unassigned   \n",
       "3                      Unassigned           Unassigned            Unassigned   \n",
       "30                     Unassigned           Unassigned            Unassigned   \n",
       "31                     Unassigned           Unassigned            Unassigned   \n",
       "32              Campylobacterales           Unassigned            Unassigned   \n",
       "34                     Unassigned           Unassigned            Unassigned   \n",
       "36                     Unassigned           Unassigned            Unassigned   \n",
       "39                     Unassigned           Unassigned            Unassigned   \n",
       "4                Rhodospirillales    Rhodospirillaceae            Unassigned   \n",
       "40                Pseudomonadales     Pseudomonadaceae           Pseudomonas   \n",
       "41                  Rhodocyclales       Rhodocyclaceae            Unassigned   \n",
       "49                Burkholderiales         unclassified           Tepidimonas   \n",
       "5                      Unassigned           Unassigned            Unassigned   \n",
       "53                Burkholderiales       Comamonadaceae            Unassigned   \n",
       "6               Ignavibacteriales           Unassigned            Unassigned   \n",
       "7              Desulfovibrionales  Desulfovibrionaceae         Desulfovibrio   \n",
       "71                   Chromatiales        Chromatiaceae            Unassigned   \n",
       "75                     Unassigned           Unassigned            Unassigned   \n",
       "8                   Bacteroidales           Unassigned            Unassigned   \n",
       "82                Pseudomonadales     Pseudomonadaceae           Pseudomonas   \n",
       "88                     Unassigned           Unassigned            Unassigned   \n",
       "9                 Xanthomonadales     Xanthomonadaceae            Arenimonas   \n",
       "\n",
       "                                      Species  \n",
       "Genome_Number                                  \n",
       "10              Sulfurihydrogenibium azorense  \n",
       "100                                Unassigned  \n",
       "101                   Tepidimonas taiwanensis  \n",
       "102                                Unassigned  \n",
       "109                                Unassigned  \n",
       "11             Calditerrivibrio nitroreducens  \n",
       "112                                Unassigned  \n",
       "118                                Unassigned  \n",
       "12                                 Unassigned  \n",
       "120                    Thermococcus litoralis  \n",
       "126                        Thermovirga lienii  \n",
       "13                                 Unassigned  \n",
       "14                                 Unassigned  \n",
       "15                                 Unassigned  \n",
       "16                                 Unassigned  \n",
       "17                                 Unassigned  \n",
       "18                    Cyanobacterium aponinum  \n",
       "19                                 Unassigned  \n",
       "2                                  Unassigned  \n",
       "20                   Desulfurella acetivorans  \n",
       "21                                 Unassigned  \n",
       "22               Fervidobacterium pennivorans  \n",
       "24                         Treponema caldaria  \n",
       "25                                 Unassigned  \n",
       "26                                 Unassigned  \n",
       "27                                 Unassigned  \n",
       "28                                 Unassigned  \n",
       "29                                 Unassigned  \n",
       "3                                  Unassigned  \n",
       "30                                 Unassigned  \n",
       "31                                 Unassigned  \n",
       "32                                 Unassigned  \n",
       "34                                 Unassigned  \n",
       "36                                 Unassigned  \n",
       "39                                 Unassigned  \n",
       "4                                  Unassigned  \n",
       "40                       Pseudomonas stutzeri  \n",
       "41                                 Unassigned  \n",
       "49                    Tepidimonas taiwanensis  \n",
       "5                                  Unassigned  \n",
       "53                                 Unassigned  \n",
       "6                                  Unassigned  \n",
       "7                   Desulfovibrio aminophilus  \n",
       "71                                 Unassigned  \n",
       "75                                 Unassigned  \n",
       "8                                  Unassigned  \n",
       "82                       Pseudomonas stutzeri  \n",
       "88                                 Unassigned  \n",
       "9                          Arenimonas malthae  "
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "############################################################################################\n",
    "### plot the overview plots with all KEGG hits and all genomes \n",
    "############################################################################################\n",
    "\n",
    "def plotClusterOverviewTot(KEGGhitType):\n",
    "    \n",
    "    if KEGGhitType == 'all':\n",
    "        dataframe = pd.read_pickle('Knumbers/brian/brian_genome_module_map_all')\n",
    "        dataframe_col = pd.read_pickle('Knumbers/brian/brian_module_map_all_col')\n",
    "    elif KEGGhitType == 'conservative':\n",
    "        dataframe = pd.read_pickle('Knumbers/brian/brian_genome_module_map_conservative')\n",
    "        dataframe_col = pd.read_pickle('Knumbers/brian/brian_module_map_cons_col')\n",
    "    elif KEGGhitType == 'complete':\n",
    "        dataframe = pd.read_pickle('Knumbers/brian/brian_genome_module_map_complete')\n",
    "        dataframe_col = pd.read_pickle('Knumbers/brian/brian_module_map_comp_col')\n",
    "    else:\n",
    "        import sys\n",
    "        sys.exit(\"aborted, enter correct df type\")\n",
    "        \n",
    "            \n",
    "    \"\"\"\n",
    "    0 = Energy metabolism\n",
    "    1 = Carb and lipid metabolism\n",
    "    2 = Nucleic acid and aa metabolism\n",
    "    3 = Secondary metabolism\n",
    "    4 = Genetic info proc.\n",
    "    5 = Env. info proc.\n",
    "    6 = Metabolism\n",
    "    7 = Cellular processes\n",
    "    8 = Gene set\n",
    "    \"\"\"\n",
    "    # some stored dataframes\n",
    "    metadf = pd.read_pickle('BrianDataSets/metadata_df')\n",
    "    KEGGmoddf = pd.read_pickle('KEGG_modules_df') #(from KEGG database -check for updates)\n",
    "    mod_df = pd.read_pickle('KEGG_module_list') #(from KEGG database -check for updates)\n",
    "    completeness = KEGGhitType\n",
    "\n",
    "    \n",
    "    ############################################################################################\n",
    "    # cluster correlation for KEGG B-level = environmental info processing\n",
    "    ############################################################################################\n",
    "\n",
    "    dataC = dataframe.fillna(0).corr().dropna(how='all').T.dropna(how='all')\n",
    "    clustindexlist = dataC.index\n",
    "\n",
    "    D = pdist(dataC, 'euclidean')\n",
    "    Z = linkage(D, 'ward')\n",
    "    row_link = _optimal_order(dataC, metric='correlation')\n",
    "\n",
    "    optimal_Z = optimal_leaf_ordering(Z, D)\n",
    "\n",
    "    cgClust = sns.clustermap(dataC, row_linkage=row_link, col_linkage=row_link,figsize=(20,40))\n",
    "\n",
    "\n",
    "    ############################################################################################\n",
    "    # KEGG-clust correlation for KEGG B-level = environmental info processing\n",
    "    ############################################################################################\n",
    "\n",
    "    #dataK = alt6.ix[idxblevdf['column'].values,:].T.corr() #np.random.choice(10000, (n, 1), replace=False)\n",
    "    dataK = dataframe.fillna(0).T.corr()\n",
    "\n",
    "    dataK = dataK.dropna(how='all').T.dropna(how='all')\n",
    "\n",
    "    D = pdist(dataK, 'euclidean')\n",
    "    Z = linkage(D, 'ward')\n",
    "    row_link = _optimal_order(dataK, metric='correlation')\n",
    "\n",
    "    optimal_Z = optimal_leaf_ordering(Z, D)\n",
    "\n",
    "    cgKEGG_blevel = sns.clustermap(dataK, row_linkage=row_link, col_linkage=row_link,figsize=(10,10))\n",
    "    roundd=0\n",
    "    # sort by clustermap, \n",
    "    for blev in range(0,9):\n",
    "\n",
    "        lut = {'Archaea':'r','Bacteria':'b','Unassigned':'k','Eukaryota':'c','Viruses':'m'}\n",
    "\n",
    "        idx = cgKEGG_blevel.dendrogram_col.reordered_ind\n",
    "        idxcl = cgClust.dendrogram_col.reordered_ind\n",
    "        cmap=ListedColormap([\"#e74c3c\", \"#3498db\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#9b59b6\"])\n",
    "        cmap=ListedColormap([\"#e74c3c\", \"#9b59b6\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#3498db\"])#[red,lightblue,green,GREY,darkblue,purple]\n",
    "        #cmap = LinearSegmentedColormap.from_list('Custom', myColors, len(myColors))\n",
    "\n",
    "        f = plt.figure()\n",
    "        gs = gridspec.GridSpec(45,37)\n",
    "\n",
    "\n",
    "        dataK2 = dataK.copy()\n",
    "        dataK2.index = mod_df.set_index('Module').loc[dataK.index]['Module_combined']\n",
    "\n",
    "        ax1 = f.add_subplot(gs[4:36,0:11])\n",
    "        sns.heatmap(dataK2.iloc[idx,idx],cbar=False,ax=ax1,xticklabels=True,yticklabels=True,cmap=\"RdBu_r\",vmin=-1,vmax=1)\n",
    "        plt.xticks([]);plt.yticks(fontsize=15)\n",
    "        plt.xlabel('KEGG modules',fontsize=30,labelpad=40)\n",
    "\n",
    "        ###############################################\n",
    "        ######### the sorter #####################\n",
    "        ###############################################\n",
    "        #idx2 = alt.sort_values('assembly').index\n",
    "        ###############################################\n",
    "        ###############################################\n",
    "        # save dataframe in ordered way for future examination\n",
    "        clusteredframe = dataframe.loc[dataK.iloc[idx,:].index,dataC.iloc[idxcl,:].index]\n",
    "        if roundd == 0:\n",
    "            clusteredframe.to_pickle('ClusteredDF_brian_'+completeness)\n",
    "            roundd=1 #save only once for a loop through B-levels\n",
    "\n",
    "        ax2 = f.add_subplot(gs[4:36,13:32])\n",
    "        sns.heatmap(dataframe_col.fillna(0).loc[dataK.iloc[idx,:].index,dataC.iloc[idxcl,:].index]\n",
    "                    ,ax=ax2,cbar=False,linewidth=0.5,cmap=cmap,vmin=-1,vmax=1,xticklabels=True)\n",
    "        plt.yticks([])\n",
    "        plt.xlabel('Genome number',fontsize=30)\n",
    "        #plt.xticks(rotation=0)\n",
    "\n",
    "        ax13 = f.add_subplot(gs[37:,13:32])\n",
    "        sns.heatmap(dataC.iloc[idxcl,idxcl],cbar=False,ax=ax13,cmap=\"RdBu_r\",vmin=-1,vmax=1)\n",
    "        plt.yticks([])\n",
    "        plt.xticks([])\n",
    "\n",
    "        ##############################################################################\n",
    "        #################### plot hbars on right #####################################\n",
    "        ##############################################################################\n",
    "        \n",
    "\n",
    "        arc = dataframe.loc[dataK.iloc[idx,:].index,metadf[metadf['Domain']=='Archaea'].index].fillna(0).T.sum()\n",
    "        bac = dataframe.loc[dataK.iloc[idx,:].index,metadf[metadf['Domain']=='Bacteria'].index].fillna(0).T.sum()\n",
    "        #vir = dataframe.loc[dataK.iloc[idx,:].index,metadf[metadf['Domain']=='Viruses'].index].T.sum()\n",
    "        una = dataframe.loc[dataK.iloc[idx,:].index,metadf[metadf['Domain']=='Unassigned'].index].fillna(0).T.sum()\n",
    "            \n",
    "        ax3 = f.add_subplot(gs[4:36,32]) #Archaea\n",
    "        arc = arc/arc.sum()\n",
    "        arc.plot.barh(ax=ax3,sharey=True, color=\"#e74c3c\")\n",
    "        ax3.grid(False)\n",
    "        plt.title('Archaea',fontsize=25,rotation=45,y=1.02)\n",
    "        plt.xticks([])\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        ax4 = f.add_subplot(gs[4:36,33]) #Bacteria\n",
    "        bac = bac/bac.sum()\n",
    "        bac.plot.barh(ax=ax4,sharey=True, color='#34495e')\n",
    "        ax4.grid(False)\n",
    "        plt.title('Bacteria',fontsize=25,rotation=45,y=1.02)\n",
    "        plt.xticks([])\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        \"\"\"ax5 = f.add_subplot(gs[4:36,34]) #Viruses\n",
    "        vir = vir/vir.sum()\n",
    "        vir.plot.barh(ax=ax5,sharey=True, color=\"#9b59b6\")\n",
    "        ax5.grid(False)\n",
    "        plt.title('Viruses',fontsize=25)\n",
    "        plt.xticks(rotation=90)\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\"\"\"\n",
    "\n",
    "        ax6 = f.add_subplot(gs[4:36,34]) #Unassigned\n",
    "        una = una/una.sum()\n",
    "        una.plot.barh(ax=ax6,sharey=True, color=\"#3498db\")\n",
    "        ax6.grid(False)\n",
    "        plt.title('Unassigned',fontsize=25,rotation=45,y=1.025)\n",
    "        plt.xticks([])\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        ax14 = f.add_subplot(gs[4:36,35:]) #Archaea-Bacteria    \n",
    "        arc = arc/arc.max()\n",
    "        bac = bac/bac.max()\n",
    "        diff = arc-bac\n",
    "\n",
    "        diff.plot.barh(ax=ax14,sharey=True, color=\"#e79f3c\")\n",
    "        ax14.grid(False)\n",
    "        plt.title('Arch-Bact\\ndifference',fontsize=25)\n",
    "        #plt.xticks(rotation=90)\n",
    "        plt.axvline(x=0);plt.xlim(-1,1)\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        ##############################################################################\n",
    "        ##########################   plot top bars   #################################\n",
    "        ##############################################################################\n",
    "\n",
    "        ax7 = f.add_subplot(gs[1,13:32]) #phylum\n",
    "        \n",
    "        col = sns.color_palette(\"cubehelix\", len(list(metadf.loc[dataC.index]['Phylum'].unique())))\n",
    "        keys = list(metadf.loc[dataC.index]['Phylum'].unique());values = [i for i in col]\n",
    "        lut = dict(zip(keys,values))\n",
    "        lut.update({'Unassigned':\"#3498db\"})\n",
    "        color = metadf['Phylum'].map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(0., 1.02, 1., .102), loc=3,\n",
    "                   ncol=6, mode=\"expand\", borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax7.set_ylabel('Phylum', rotation=0, fontsize=30, labelpad=90)\n",
    "\n",
    "\n",
    "\n",
    "        #3498db\n",
    "        ax8 = f.add_subplot(gs[2,13:32]) #domain\n",
    "        lut = {'Archaea':\"#e74c3c\",'Bacteria':'#34495e','Unassigned':\"#3498db\"}#'Eukaryota':'c','Viruses':\"#9b59b6\"}\n",
    "        color = metadf['Domain'].map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=2, borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax8.set_ylabel('Domain', rotation=0, fontsize=30, labelpad=90)\n",
    "\n",
    "\n",
    "\n",
    "        ax9 = f.add_subplot(gs[3,13:32]) #mean GC???\n",
    "        \n",
    "        #col = sns.color_palette(\"BrBG\", len(alt['assembly'].unique()))\n",
    "        #keys = np.sort(alt['assembly'].unique())\n",
    "        #values = [i for i in col]\n",
    "        #lut = dict(zip(keys,values))\n",
    "        #color = alt['assembly'].map(lut)\n",
    "\n",
    "        \"\"\"for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=3, borderaxespad=0.,fontsize=20)\n",
    "        \"\"\"\n",
    "        #print(idxcl)\n",
    "        listGC = list(metadf.T.loc['meanGC',dataC.iloc[idxcl,:].index])\n",
    "        #print(listGC)\n",
    "        sns.heatmap([listGC],ax=ax9,cbar=False,cmap=\"RdBu_r\",linewidth=0.5,vmin=0,vmax=1,xticklabels=False)\n",
    "        ax9.set_ylabel('GC-content', rotation=0, fontsize=30, labelpad=90)\n",
    "        \"\"\"plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        ax9.set_ylabel('Assembly', rotation=0, fontsize=30, labelpad=90)\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        \"\"\"ax10 = f.add_subplot(gs[3,13:32]) #experiment\n",
    "        col = sns.color_palette(\"Set2\", 5)\n",
    "        if group == 'YES':\n",
    "            lut = {'Obs2':col[0],'Obs3':col[1],'Obs4':col[2],'Obs5':col[3],'Obs6':col[4]}\n",
    "            color2 = alt5['sample'].map(lut)\n",
    "        elif group == 'SC':\n",
    "            col = sns.color_palette(\"Set2\", 6)\n",
    "            lut = {'Obs2':col[0],'Obs3':col[1],'Obs4':col[2],'Obs5':col[3],'Obs6':col[4],'mix':col[5]}\n",
    "            color2 = altSC['sample'].map(lut)\n",
    "        else:\n",
    "            lut = {'Obsidian2':col[0],'Obsidian3':col[1],'Obsidian4':col[2],'Obsidian5':col[3],'Obsidian6':col[4]}\n",
    "            color2 = alt['sample'].map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=3, borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color2[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax10.set_ylabel('Experiment', rotation=0, fontsize=30, labelpad=100)\"\"\"\n",
    "\n",
    "\n",
    "        ##############################################################################\n",
    "        ##########################   plot left side bars  ############################\n",
    "        ##############################################################################\n",
    "        Blevel = blev\n",
    "\n",
    "        \"\"\"\n",
    "        0 = Energy metabolism\n",
    "        1 = Carb and lipid metabolism\n",
    "        2 = Nucleic acid and aa metabolism\n",
    "        3 = Secondary metabolism\n",
    "        4 = Genetic info proc.\n",
    "        5 = Env. info proc.\n",
    "        6 = Metabolism\n",
    "        7 = Cellular processes\n",
    "        8 = Gene set\n",
    "        \"\"\"\n",
    "        # B-level assignment\n",
    "        ax11 = f.add_subplot(gs[4:36,11])\n",
    "        col = sns.color_palette(\"gist_earth\", len(KEGGmoddf.B.unique())-1)\n",
    "        keys = list(KEGGmoddf.B.unique());keys = keys[0:-1];values = [i for i in col]\n",
    "        keysC = KEGGmoddf[KEGGmoddf['B']==keys[Blevel]].C.unique();\n",
    "\n",
    "        lut = dict(zip(keys,values))\n",
    "        idxlist = dataK.iloc[idx,:].index\n",
    "        #idxlist2 = [i[0:6] for i in idxlist]\n",
    "        D_group = KEGGmoddf.groupby('D-module').sum()\n",
    "\n",
    "        for i in D_group.index:\n",
    "            a = D_group.loc[i,'B']\n",
    "            for j in keys:\n",
    "                if a.find(j)!=-1:\n",
    "                    D_group.loc[i,'B'] = j\n",
    "            c = D_group.loc[i,'C']\n",
    "            for j in KEGGmoddf.C.unique():\n",
    "                if c.find(j)!=-1:\n",
    "                    D_group.loc[i,'C'] = j\n",
    "\n",
    "        B_list  = D_group.loc[idxlist]['B']\n",
    "        #B_list.replace(B_list.index,idxlist)\n",
    "        color = B_list.map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(-20, 1.02, 1., .102), loc=3,\n",
    "                   ncol=1, mode=\"expand\", borderaxespad=0.,fontsize=30)\n",
    "\n",
    "        plt.barh(list(range(0,len(dataK))),list(np.ones(len(dataK))),color=color,height=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.ylim(-.5,len(dataK)-.5)\n",
    "        plt.xlim(0,1)\n",
    "        ax11.set_xlabel('B', rotation=0, fontsize=30, labelpad=40)\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "\n",
    "        #C-level assignment\n",
    "\n",
    "        ax12 = f.add_subplot(gs[4:36,12])\n",
    "        col = sns.color_palette(\"Set3\", len(keysC));values = [i for i in col]\n",
    "        lutC = dict(zip(keysC,values))\n",
    "\n",
    "        C_list  = D_group.loc[idxlist]['C']\n",
    "        C_list.replace(C_list.index,idxlist)\n",
    "        color = C_list.map(lutC)\n",
    "        color = color.fillna('None')\n",
    "\n",
    "        plt.barh(list(range(0,len(dataK))),list(np.ones(len(dataK))),color=color,height=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.ylim(-.5,len(dataK)-.5)\n",
    "        plt.xlim(0,1)\n",
    "        ax12.set_xlabel('C', rotation=0, fontsize=30, labelpad=40)\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        for x,y in lutC.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(-10, 1.02, 1., .102), loc=3,\n",
    "                   ncol=1, mode=\"expand\", borderaxespad=0.,fontsize=30)\n",
    "\n",
    "        idx_blevel5 = C_list.loc[color[color!='None'].index].index\n",
    "        idxblevdf = pd.DataFrame(idx_blevel5,columns=['column'])\n",
    "        idxblevdf.to_pickle('ObsCom_indexblevel_'+completeness+'_'+str(Blevel))\n",
    "\n",
    "        \"\"\"ax5 = f.add_subplot(gs[0,:])\n",
    "        ax5.text(0.5,0.5,'bla',fontsize=28,horizontalalignment='center',verticalalignment='center')\n",
    "        ax5.grid(False)\n",
    "        plt.yticks([])\n",
    "        plt.xticks([])\"\"\"\n",
    "\n",
    "        gs.update(wspace=.05,hspace=.05)\n",
    "        plt.gcf().subplots_adjust(left=0.3)\n",
    "        f.set_figheight(110)\n",
    "        #plt.show()\n",
    "        f.set_figwidth(60)\n",
    "        f.savefig('BrianDataSets/figures/heatmaps/Overview_heatmap_KEGG_'+completeness+'_'+keys[Blevel])\n",
    "        f.savefig('BrianDataSets/figures/heatmaps/pdfs/Overview_heatmap_KEGG_'+completeness+'_'+keys[Blevel]+'.pdf')\n",
    "        f.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.6/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "plotClusterOverviewTot('conservative')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plotClusterOverviewTot('all')\n",
    "plotClusterOverviewTot('complete')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "############################################################################################\n",
    "# plot the separately clustered blevel KEGG modules, without the genomes in which\n",
    "# they do not occur. (so perform cluster correlation there)\n",
    "############################################################################################\n",
    "\n",
    "def plotClusterOverviewBlev_sel(KEGGhitType):\n",
    "    \n",
    "    if KEGGhitType == 'all':\n",
    "        dataframe = pd.read_pickle('Knumbers/brian/brian_genome_module_map_all')\n",
    "        dataframe_col = pd.read_pickle('Knumbers/brian/brian_module_map_all_col')\n",
    "    elif KEGGhitType == 'conservative':\n",
    "        dataframe = pd.read_pickle('Knumbers/brian/brian_genome_module_map_conservative')\n",
    "        dataframe_col = pd.read_pickle('Knumbers/brian/brian_module_map_cons_col')\n",
    "    elif KEGGhitType == 'complete':\n",
    "        dataframe = pd.read_pickle('Knumbers/brian/brian_genome_module_map_complete')\n",
    "        dataframe_col = pd.read_pickle('Knumbers/brian/brian_module_map_comp_col')\n",
    "    else:\n",
    "        import sys\n",
    "        sys.exit(\"aborted, enter correct df type\")         \n",
    "    \n",
    "    \"\"\"\n",
    "    0 = Energy metabolism\n",
    "    1 = Carb and lipid metabolism\n",
    "    2 = Nucleic acid and aa metabolism\n",
    "    3 = Secondary metabolism\n",
    "    4 = Genetic info proc.\n",
    "    5 = Env. info proc.\n",
    "    6 = Metabolism\n",
    "    7 = Cellular processes\n",
    "    8 = Gene set\n",
    "    \"\"\"\n",
    "    # some stored dataframes\n",
    "    metadf = pd.read_pickle('BrianDataSets/metadata_df')\n",
    "    KEGGmoddf = pd.read_pickle('KEGG_modules_df') #(from KEGG database -check for updates)\n",
    "    mod_df = pd.read_pickle('KEGG_module_list') #(from KEGG database -check for updates)\n",
    "    completeness = KEGGhitType\n",
    "\n",
    "    for blev in range(0,7):\n",
    "        Blevel = blev\n",
    "        \n",
    "        idxblevdf = pd.read_pickle('ObsCom_indexblevel_'+completeness+'_'+str(Blevel))\n",
    "        \n",
    "        ############################################################################################\n",
    "        # cluster correlation for KEGG B-level = environmental info processing\n",
    "        ############################################################################################\n",
    "        # selected genomes\n",
    "        dataC = dataframe.loc[idxblevdf['column'].values,:].fillna(0).corr()\n",
    "        dataC = dataC.dropna(how='all').T.dropna(how='all')\n",
    "        clustindexlist = dataC.index\n",
    "        # all genomes\n",
    "        \"\"\"\n",
    "        dataC = dataframe.fillna(0).corr().dropna(how='all').T.dropna(how='all')\n",
    "        clustindexlist = dataC.index        \n",
    "        \"\"\"\n",
    "        D = pdist(dataC, 'euclidean')\n",
    "        Z = linkage(D, 'ward')\n",
    "        row_link = _optimal_order(dataC, metric='correlation')\n",
    "\n",
    "        optimal_Z_clust = optimal_leaf_ordering(Z, D)\n",
    "\n",
    "        cgClust = sns.clustermap(dataC, row_linkage=row_link, col_linkage=row_link,figsize=(20,40))\n",
    "\n",
    "\n",
    "        ############################################################################################\n",
    "        # KEGG-clust correlation for KEGG B-level = environmental info processing\n",
    "        ############################################################################################\n",
    "\n",
    "        dataK = dataframe.loc[idxblevdf['column'].values,:].fillna(0).T.corr() #np.random.choice(10000, (n, 1), replace=False)\n",
    "        #dataK = Mmap_all_NU.fillna(0).T.corr()\n",
    "\n",
    "        dataK = dataK.dropna(how='all').T.dropna(how='all')\n",
    "        if len(dataK)>1:\n",
    "            D = pdist(dataK, 'euclidean')\n",
    "            Z = linkage(D, 'ward')\n",
    "            row_link = _optimal_order(dataK, metric='correlation')\n",
    "\n",
    "            optimal_Z = optimal_leaf_ordering(Z, D)\n",
    "\n",
    "            cgKEGG_blevel = sns.clustermap(dataK, row_linkage=row_link, col_linkage=row_link,figsize=(10,10))\n",
    "\n",
    "            # sort by clustermap, KEGG axis only\n",
    "\n",
    "\n",
    "            lut = {'Archaea':'r','Bacteria':'b','Unassigned':'k','Eukaryota':'c','Viruses':'m'}\n",
    "\n",
    "            idx = cgKEGG_blevel.dendrogram_col.reordered_ind\n",
    "            idxcl = cgClust.dendrogram_col.reordered_ind\n",
    "            cmap=ListedColormap([\"#e74c3c\", \"#3498db\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#9b59b6\"])\n",
    "            cmap=ListedColormap([\"#e74c3c\", \"#9b59b6\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#3498db\"])#[red,lightblue,green,GREY,darkblue,purple]\n",
    "            #cmap = LinearSegmentedColormap.from_list('Custom', myColors, len(myColors))\n",
    "\n",
    "            f = plt.figure()\n",
    "            gs = gridspec.GridSpec(36,37)\n",
    "\n",
    "\n",
    "            dataK2 = dataK.copy()\n",
    "            dataK2.index = mod_df.set_index('Module').loc[dataK.index]['Module_combined']\n",
    "\n",
    "            ax1 = f.add_subplot(gs[5:36,0:5])\n",
    "            sns.heatmap(dataK2.iloc[idx,idx],cbar=False,ax=ax1,xticklabels=True,yticklabels=True,cmap=\"RdBu_r\",vmin=-1,vmax=1)\n",
    "            plt.xticks([]);#plt.yticks(fontsize=11)\n",
    "            plt.ylabel('')\n",
    "            plt.xlabel('KEGG modules')#,fontsize=30,labelpad=40)\n",
    "\n",
    "            ###############################################\n",
    "            ######### the sorter #####################\n",
    "            ###############################################\n",
    "            #idx2 = alt.sort_values('assembly').index\n",
    "            ###############################################\n",
    "            ###############################################\n",
    "\n",
    "            ax2 = f.add_subplot(gs[5:36,7:32])\n",
    "            sns.heatmap(dataframe_col.fillna(0).loc[dataK.iloc[idx,:].index,dataC.iloc[idxcl,:].index]\n",
    "                        ,ax=ax2,cbar=False,linewidth=0.5,cmap=cmap,vmin=-1,vmax=1,xticklabels=True)\n",
    "            plt.yticks([])\n",
    "            #plt.yticks(fontsize=35)\n",
    "            plt.xlabel('Genome number')\n",
    "            #plt.xticks(rotation=0)\n",
    "\n",
    "            \"\"\"ax13 = f.add_subplot(gs[39:,13:32])\n",
    "            sns.heatmap(dataC.iloc[idxcl,idxcl],cbar=False,ax=ax13,cmap=\"RdBu_r\",vmin=-1,vmax=1)\n",
    "            plt.yticks([])\n",
    "            plt.xticks([])\"\"\"\n",
    "\n",
    "            ##############################################################################\n",
    "            #################### plot hbars on right #####################################\n",
    "            ##############################################################################\n",
    "            arc = dataframe.loc[dataK.iloc[idx,:].index,metadf[metadf['Domain']=='Archaea'].index].fillna(0).T.sum()\n",
    "            bac = dataframe.loc[dataK.iloc[idx,:].index,metadf[metadf['Domain']=='Bacteria'].index].fillna(0).T.sum()\n",
    "            #vir = dataframe.loc[dataK.iloc[idx,:].index,metadf[metadf['Domain']=='Viruses'].index].T.sum()\n",
    "            una = dataframe.loc[dataK.iloc[idx,:].index,metadf[metadf['Domain']=='Unassigned'].index].fillna(0).T.sum()\n",
    "\n",
    "            ax3 = f.add_subplot(gs[5:36,32]) #Archaea\n",
    "            g = arc/arc.sum()\n",
    "            g.plot.barh(ax=ax3,sharey=True, color=\"#e74c3c\")\n",
    "            ax3.grid(False)\n",
    "            plt.title('Archaea',fontsize=15,rotation=45,y=1.035)\n",
    "            #plt.xticks(rotation=90)\n",
    "            #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "            plt.gca().invert_yaxis()\n",
    "            plt.xticks([])\n",
    "\n",
    "            ax4 = f.add_subplot(gs[5:36,33]) #Bacteria\n",
    "            g = bac/bac.sum()\n",
    "            g.plot.barh(ax=ax4,sharey=True, color='#34495e')\n",
    "            ax4.grid(False)\n",
    "            plt.title('Bacteria',fontsize=15,rotation=45,y=1.035)\n",
    "            #plt.xticks(rotation=90)\n",
    "            #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "            plt.gca().invert_yaxis()\n",
    "            plt.xticks([])\n",
    "\n",
    "            \"\"\"ax5 = f.add_subplot(gs[4:36,34]) #Viruses            \n",
    "            g = vir/vir.sum()\n",
    "            g.plot.barh(ax=ax5,sharey=True, color=\"#9b59b6\")\n",
    "            ax5.grid(False)\n",
    "            plt.title('Viruses',fontsize=25)\n",
    "            plt.xticks(rotation=90)\n",
    "            #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "            plt.gca().invert_yaxis()\"\"\"\n",
    "\n",
    "            ax6 = f.add_subplot(gs[5:36,34]) #Unassigned\n",
    "            g = una/una.sum()\n",
    "            g.plot.barh(ax=ax6,sharey=True, color=\"#3498db\")\n",
    "            ax6.grid(False)\n",
    "            plt.title('Unassigned',fontsize=15,rotation=45,y=1.04)\n",
    "            #plt.xticks(rotation=90)\n",
    "            #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "            plt.gca().invert_yaxis()\n",
    "            plt.xticks([])\n",
    "\n",
    "            ax14 = f.add_subplot(gs[5:36,35:]) #Archaea-Bacteria\n",
    "            g = arc/arc.max()\n",
    "            h = bac/bac.max()\n",
    "            diff = g-h\n",
    "\n",
    "            diff.plot.barh(ax=ax14,sharey=True, color=\"#e79f3c\")\n",
    "            ax14.grid(False)\n",
    "            plt.title('Arch-Bact\\ndifference',fontsize=15)\n",
    "            #plt.xticks(rotation=90)\n",
    "            plt.axvline(x=0);plt.xlim(-1,1)\n",
    "            plt.gca().invert_yaxis()\n",
    "\n",
    "            \n",
    "            ##############################################################################\n",
    "            ##########################   plot top bars   #################################\n",
    "            ##############################################################################\n",
    "\n",
    "            ax7 = f.add_subplot(gs[0:2,7:32]) #phylum\n",
    "            \n",
    "            col = sns.color_palette(\"cubehelix\", len(list(metadf.loc[dataC.index]['Phylum'].unique())))\n",
    "            keys = list(metadf.loc[dataC.index]['Phylum'].unique());values = [i for i in col]\n",
    "            lut = dict(zip(keys,values))\n",
    "            lut.update({'Unassigned':\"#3498db\"})\n",
    "            color = metadf['Phylum'].map(lut)\n",
    "            \n",
    "            for x,y in lut.items():\n",
    "                plt.bar(0,0,color=y,label=x,width=1)\n",
    "            plt.legend(bbox_to_anchor=(0., 1.02, 1., .102), loc=3,\n",
    "                       ncol=5, mode=\"expand\", borderaxespad=0.)#,fontsize=30)\n",
    "\n",
    "            plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "            plt.xticks([]);plt.yticks([])\n",
    "            plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "            ax7.set_ylabel('Phylum', rotation=0, fontsize=20, labelpad=90)\n",
    "\n",
    "\n",
    "\n",
    "            #3498db\n",
    "            ax8 = f.add_subplot(gs[2:3,7:32]) #domain\n",
    "            lut = {'Archaea':\"#e74c3c\",'Bacteria':'#34495e','Unassigned':\"#3498db\"}#'Eukaryota':'c','Viruses':\"#9b59b6\"}\n",
    "            color = metadf['Domain'].map(lut)\n",
    "\n",
    "            for x,y in lut.items():\n",
    "                plt.bar(0,0,color=y,label=x,width=1)\n",
    "            plt.legend(bbox_to_anchor=(1.05, 4, 1., .102), loc=2,\n",
    "                       ncol=1, borderaxespad=0.,fontsize=15)\n",
    "\n",
    "            plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "            plt.xticks([]);plt.yticks([])\n",
    "            plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "            ax8.set_ylabel('Kingdom', rotation=0, fontsize=20, labelpad=90)\n",
    "\n",
    "\n",
    "\n",
    "            ax9 = f.add_subplot(gs[3:5,7:32]) # GC\n",
    "            listGC = list(metadf.T.loc['meanGC',dataC.iloc[idxcl,:].index])\n",
    "            #print(listGC)\n",
    "            sns.heatmap([listGC],ax=ax9,cbar=False,cmap=\"RdBu_r\",linewidth=0.5,vmin=0,vmax=1,xticklabels=False)\n",
    "            ax9.set_ylabel('GC-content', rotation=0, fontsize=20, labelpad=90)\n",
    "\n",
    "\n",
    "            \"\"\"\n",
    "\n",
    "            ax10 = f.add_subplot(gs[3,7:32]) #experiment\n",
    "            col = sns.color_palette(\"Set2\", 5)\n",
    "            if group == 'YES':\n",
    "                lut = {'Obs2':col[0],'Obs3':col[1],'Obs4':col[2],'Obs5':col[3],'Obs6':col[4]}\n",
    "                color2 = alt5['sample'].map(lut)\n",
    "            elif group == 'SC':\n",
    "                col = sns.color_palette(\"Set2\", 6)\n",
    "                lut = {'Obs2':col[0],'Obs3':col[1],'Obs4':col[2],'Obs5':col[3],'Obs6':col[4],'mix':col[5]}\n",
    "                color2 = altSC['sample'].map(lut)\n",
    "            else:\n",
    "                lut = {'Obsidian2':col[0],'Obsidian3':col[1],'Obsidian4':col[2],'Obsidian5':col[3],'Obsidian6':col[4]}\n",
    "                color2 = alt['sample'].map(lut)\n",
    "\n",
    "            for x,y in lut.items():\n",
    "                plt.bar(0,0,color=y,label=x,width=1)\n",
    "            plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                       ncol=3, borderaxespad=0.,fontsize=20)\n",
    "\n",
    "            plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color2[dataC.iloc[idxcl,:].index],width=1)\n",
    "            plt.xticks([]);plt.yticks([])\n",
    "            plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "            ax10.set_ylabel('Experiment', rotation=0, fontsize=20, labelpad=100)\"\"\"\n",
    "\n",
    "\n",
    "            ##############################################################################\n",
    "            ##########################   plot left side bars  ############################\n",
    "            ##############################################################################\n",
    "            Blevel = blev\n",
    "\n",
    "            \"\"\"\n",
    "            0 = Energy metabolism\n",
    "            1 = Carb and lipid metabolism\n",
    "            2 = Nucleic acid and aa metabolism\n",
    "            3 = Secondary metabolism\n",
    "            4 = Genetic info proc.\n",
    "            5 = Env. info proc.\n",
    "            6 = Metabolism\n",
    "            7 = Cellular processes\n",
    "            8 = Gene set\n",
    "            \"\"\"\n",
    "            # B-level assignment\n",
    "            ax11 = f.add_subplot(gs[5:36,5])\n",
    "            col = sns.color_palette(\"gist_earth\", len(KEGGmoddf.B.unique())-1)\n",
    "            keys = list(KEGGmoddf.B.unique());keys = keys[0:-1];values = [i for i in col]\n",
    "            keysC = KEGGmoddf[KEGGmoddf['B']==keys[Blevel]].C.unique();\n",
    "\n",
    "            lut = dict(zip(keys,values))\n",
    "            idxlist = dataK.iloc[idx,:].index\n",
    "            #idxlist2 = [i[0:6] for i in idxlist]\n",
    "            D_group = KEGGmoddf.groupby('D-module').sum()\n",
    "\n",
    "            for i in D_group.index:\n",
    "                a = D_group.loc[i,'B']\n",
    "                for j in keys:\n",
    "                    if a.find(j)!=-1:\n",
    "                        D_group.loc[i,'B'] = j\n",
    "                c = D_group.loc[i,'C']\n",
    "                for j in KEGGmoddf.C.unique():\n",
    "                    if c.find(j)!=-1:\n",
    "                        D_group.loc[i,'C'] = j\n",
    "\n",
    "            B_list  = D_group.loc[idxlist]['B']\n",
    "            #B_list.replace(B_list.index,idxlist)\n",
    "            color = B_list.map(lut)\n",
    "\n",
    "            \"\"\"for x,y in lut.items():\n",
    "                plt.bar(0,0,color=y,label=x,width=1)\n",
    "            plt.legend(bbox_to_anchor=(-8, 1.02, 1., .102), loc=3,\n",
    "                       ncol=1, mode=\"expand\", borderaxespad=0.,fontsize=30)\"\"\"\n",
    "\n",
    "            plt.barh(list(range(0,len(dataK))),list(np.ones(len(dataK))),color=color,height=1)\n",
    "            plt.xticks([]);plt.yticks([])\n",
    "            plt.ylim(-.5,len(dataK)-.5)\n",
    "            plt.xlim(0,1)\n",
    "            ax11.set_xlabel('B-level', rotation=90, fontsize=20, labelpad=40)\n",
    "            plt.gca().invert_yaxis()\n",
    "\n",
    "\n",
    "            #C-level assignment\n",
    "\n",
    "            ax12 = f.add_subplot(gs[5:36,6])\n",
    "            col = sns.color_palette(\"Set3\", len(keysC));values = [i for i in col]\n",
    "            lutC = dict(zip(keysC,values))\n",
    "\n",
    "            C_list  = D_group.loc[idxlist]['C']\n",
    "            C_list.replace(C_list.index,idxlist)\n",
    "            color = C_list.map(lutC)\n",
    "            color = color.fillna('None')\n",
    "\n",
    "            plt.barh(list(range(0,len(dataK))),list(np.ones(len(dataK))),color=color,height=1)\n",
    "            plt.xticks([]);plt.yticks([])\n",
    "            plt.ylim(-.5,len(dataK)-.5)\n",
    "            plt.xlim(0,1)\n",
    "            ax12.set_xlabel('C-level', rotation=90, fontsize=20, labelpad=40)\n",
    "            plt.gca().invert_yaxis()\n",
    "\n",
    "            for x,y in lutC.items():\n",
    "                plt.bar(0,0,color=y,label=x,width=1)\n",
    "            if blev==5:\n",
    "                plt.legend(bbox_to_anchor=(-24, 1.02, 1., .102), loc=3,\n",
    "                       ncol=1, mode=\"expand\", borderaxespad=0.)#,fontsize=30)\n",
    "            else:\n",
    "                plt.legend(bbox_to_anchor=(-8, 1.02, 1., .102), loc=3,\n",
    "                           ncol=1, mode=\"expand\", borderaxespad=0.)#,fontsize=30)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            \"\"\"ax5 = f.add_subplot(gs[0,:])\n",
    "            ax5.text(0.5,0.5,'bla',fontsize=28,horizontalalignment='center',verticalalignment='center')\n",
    "            ax5.grid(False)\n",
    "            plt.yticks([])\n",
    "            plt.xticks([])\"\"\"\n",
    "\n",
    "            gs.update(wspace=.05,hspace=.05)\n",
    "            \n",
    "            #plt.gcf().subplots_adjust(bottom=0.1,top=0.11)\n",
    "            #f.set_figheight(30+0.2*len(dataK))\n",
    "            if blev == 5:\n",
    "                plt.gcf().subplots_adjust(left=0.4)\n",
    "                f.set_figheight(60)\n",
    "                f.set_figwidth(40)\n",
    "            else:\n",
    "                plt.gcf().subplots_adjust(left=0.3)\n",
    "                f.set_figheight(20)\n",
    "                f.set_figwidth(60)\n",
    "                \n",
    "            f.savefig('BrianDataSets/figures/heatmaps/BlevelOnly/Overview_heatmap_KEGGBlevel_'+completeness+'_'+keys[Blevel])\n",
    "            f.savefig('BrianDataSets/figures/heatmaps/BlevelOnly/pdfs/Overview_heatmap_KEGGBlevel_'+completeness+'_'+keys[Blevel]+'.pdf')\n",
    "            f.clf()\n",
    "            #return(optimal_Z_clust,clustindexlist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.6/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "plotClusterOverviewBlev_sel('conservative')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
