{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import csv\n",
    "import matplotlib.cm as cm\n",
    "import seaborn as sns\n",
    "from matplotlib import gridspec\n",
    "import matplotlib.animation as manimation\n",
    "import matplotlib.patches as patches\n",
    "import time\n",
    "from collections import Counter\n",
    "import sklearn.cluster as cluster\n",
    "import time\n",
    "import hdbscan\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import re\n",
    "\n",
    "#import plotly. as py\n",
    "#import plotly.graph_objs as go\n",
    "font = {'family' : 'DejaVu Sans',\n",
    "        'weight' : 'regular',\n",
    "        'size'   : 16}\n",
    "\n",
    "plt.rc('font', **font)\n",
    "plt.rc('lines',lw=2) \n",
    "sns.set(font_scale = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# dereks function for optimal leaf clustering\n",
    "def _optimal_order(data, **kwargs):\n",
    "    \"\"\" Optimal leaf ordering\n",
    "        **kwargs passed to pdist e.g. metric='correlation'\n",
    "    \"\"\"\n",
    "    d = pdist(data, **kwargs)\n",
    "    link = linkage(d, method='average')\n",
    "    optimal_order = optimal_leaf_ordering(link, d)\n",
    "    return optimal_order\n",
    "\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from fastcluster import linkage\n",
    "from polo import optimal_leaf_ordering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "KEGGmoddf = pd.read_pickle('KEGG_modules_df')\n",
    "alt4 = pd.read_pickle('ObsCom_alt4')\n",
    "alt = pd.read_pickle('ObsCom_alt')\n",
    "alt5 = pd.read_pickle('ObsCom_alt5')\n",
    "alt6 = pd.read_pickle('ObsCom_alt6')\n",
    "indexswitcher = pd.read_pickle('ObsCom_indexswitcher')\n",
    "altlength = pd.read_pickle('ObsCom_altlength')\n",
    "clusterRes_df = pd.read_pickle('ObsCom_clusterResDF')\n",
    "\n",
    "alt2 = pd.read_pickle('ObsCom_alt2')\n",
    "alt3 = pd.read_pickle('ObsCom_alt3')\n",
    "altlength = pd.read_pickle('ObsCom_altlength')\n",
    "altlength.groupby('sample').sum()['chipcount']\n",
    "\n",
    "mod_df = pd.read_pickle('KEGG_module_list')\n",
    "known_genomes = pd.read_pickle('known_genomes_KEGG_module_database')\n",
    "database4 = pd.read_pickle('KEGG_organism_list')\n",
    "mod_df['Module_combined'] = mod_df['Module']+', '+mod_df['Module_long']\n",
    "known_genomes_long = known_genomes.copy()\n",
    "known_genomes_long.index = mod_df['Module_combined']\n",
    "\n",
    "dataC = pd.read_pickle('all_genomes_dataC')\n",
    "dataK = pd.read_pickle('all_genomes_dataK')\n",
    "idx = list(open('idxfile.txt'))\n",
    "idxcl = list(open('idxclfile.txt'))\n",
    "idx = [int(i[:-1]) for i in idx]\n",
    "idxcl = [int(i[:-1]) for i in idxcl]\n",
    "\n",
    "Mmap_all_NU = pd.read_pickle('Mmap_all_NU')\n",
    "Mmap_cons_NU = pd.read_pickle('Mmap_cons_NU')\n",
    "Mmap_complete_NU = pd.read_pickle('Mmap_complete_NU')\n",
    "Mmap_all_NU_col = pd.read_pickle('Mmap_all_NU_col')\n",
    "Mmap_cons_NU_col = pd.read_pickle('Mmap_cons_NU_col')\n",
    "Mmap_complete_NU_col = pd.read_pickle('Mmap_complete_NU_col')\n",
    "\n",
    "Mmap_all_NU_grp = pd.read_pickle('Mmap_all_NU_grp')\n",
    "Mmap_cons_NU_grp = pd.read_pickle('Mmap_cons_NU_grp')\n",
    "Mmap_complete_NU_grp = pd.read_pickle('Mmap_complete_NU_grp')\n",
    "Mmap_all_NU_grp_col = pd.read_pickle('Mmap_all_NU_grp_col')\n",
    "Mmap_cons_NU_grp_col = pd.read_pickle('Mmap_cons_NU_grp_col')\n",
    "Mmap_complete_NU_grp_col = pd.read_pickle('Mmap_complete_NU_grp_col')\n",
    "\n",
    "known_genomes_all = pd.read_pickle('known_genomes_all')\n",
    "known_genomes_all_grp = pd.read_pickle('known_genomes_all_grp')\n",
    "known_genomes_cons = pd.read_pickle('known_genomes_cons')\n",
    "known_genomes_cons_grp = pd.read_pickle('known_genomes_cons_grp')\n",
    "known_genomes_complete = pd.read_pickle('known_genomes_complete')\n",
    "known_genomes_complete_grp = pd.read_pickle('known_genomes_complete_grp') \n",
    "\n",
    "known_genomes_all_col = pd.read_pickle('known_genomes_all_col')\n",
    "known_genomes_all_grp_col = pd.read_pickle('known_genomes_all_grp_col')\n",
    "known_genomes_cons_col = pd.read_pickle('known_genomes_cons_col')\n",
    "known_genomes_cons_grp_col = pd.read_pickle('known_genomes_cons_grp_col')\n",
    "known_genomes_complete_col = pd.read_pickle('known_genomes_complete_col')\n",
    "known_genomes_complete_grp_col = pd.read_pickle('known_genomes_complete_grp_col') \n",
    "\n",
    "# assign length to each grouped cluster, as well as ungrouped \n",
    "alt5['length'] = altlength.groupby('groupnum').sum()['length']\n",
    "alt5.loc[altlength[altlength.index.isin(alt5.index)].index,'length'] = altlength[altlength.index.isin(alt5.index)]['length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "domain                          Viruses\n",
       "phylum                    dsDNA viruses\n",
       "species                      Unassigned\n",
       "assembly                       megameta\n",
       "orig doms                VirusesViruses\n",
       "orig phyl    dsDNA virusesdsDNA viruses\n",
       "clustsum       Obs4_mega_11Obs4_meta_17\n",
       "sample                             Obs4\n",
       "length                      1.01041e+06\n",
       "Name: M51, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt5[alt5['phylum'].str.contains('Micr')]\n",
    "#alt5[alt5['clustsum'].str.contains('Obs2_meta_16')]\n",
    "alt5.loc['M51']\n",
    "#alt5.loc['Obs5_mini_8']\n",
    "#alt5"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Mmap_all_NU.loc['M00598'].fillna(0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#add superclusters - superclusters generally are similar clusters over different experiments\n",
    "\n",
    "#alt['SuperCluster']=''\n",
    "#alt[alt['groupnum'].isin(['M59','M34','M00'])]['SuperCluster']=1\n",
    "\"\"\"alt.loc[alt[alt['groupnum'].isin(['M59','M34','M00'])].index,'SuperCluster'] = 1\n",
    "alt.loc[alt[alt['idx'].isin(['Obs4_meta_8',\n",
    "                             'Obs6_meta_18','Obs2_mega_3'])].index,'SuperCluster'] = 2\n",
    "alt.loc[alt[alt['idx'].isin(['Obs6_meta_23'])].index,'SuperCluster'] = 3\n",
    "alt.loc[alt[alt['groupnum'].isin(['M50','M04','M26'])].index,'SuperCluster'] = 3\n",
    "alt.loc[alt[alt['groupnum'].isin(['M56','M75'])].index,'SuperCluster'] = 4\n",
    "alt.loc[alt[alt['groupnum'].isin(['M35'])].index,'SuperCluster'] = 5\n",
    "alt.loc[alt[alt['idx'].isin(['Obs5_mini_8'])].index,'SuperCluster'] = 5\"\"\"\n",
    "# these have a similar set of drug resistance genes\n",
    "alt[alt['groupnum'].isin(['M58','M60','M31','M19','M39'])]\n",
    "alt[alt['idx'].isin(['Obs5_mini_18'])]\n",
    "#unique functionality \n",
    "alt[alt['groupnum']=='M30']\n",
    "alt[alt['groupnum']=='M29']\n",
    "\n",
    "# save to supercluster \n",
    "alt.to_pickle('ObsCom_alt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "database_gen_combined = database4.set_index('code')[['kingdom','phylum']]\n",
    "database_gen_combined.columns = ['domain','phylum']\n",
    "altt = alt[['dominant domain','dominant phylum']]\n",
    "altt.columns = ['domain','phylum']\n",
    "database_gen_combined = database_gen_combined.append(altt[['domain','phylum']])\n",
    "\n",
    "database_gen_combined_grp = database4.set_index('code')[['kingdom','phylum']]\n",
    "database_gen_combined_grp.columns = ['domain','phylum']\n",
    "database_gen_combined_grp = database_gen_combined.append(alt5[['domain','phylum']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "############################################################################################\n",
    "### plot the overview plots with all KEGG hits and all genomes \n",
    "############################################################################################\n",
    "\n",
    "def plotClusterOverviewTot_gen(KEGGhitType,group,phylum):\n",
    "    \n",
    "    if group != \"YES\":\n",
    "        \n",
    "        if KEGGhitType == 'all':\n",
    "            dataframe = pd.read_pickle('known_genomes_all')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_all_col')\n",
    "            subtitle = 'all KEGG Module hits'\n",
    "        elif KEGGhitType == 'conservative':\n",
    "            dataframe = pd.read_pickle('known_genomes_cons')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_cons_col')\n",
    "            subtitle = 'complete KEGG Modules and Modules missing one block'\n",
    "        elif KEGGhitType == 'complete':\n",
    "            dataframe = pd.read_pickle('known_genomes_complete')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_complete_col')\n",
    "            subtitle = 'complete KEGG Modules only'\n",
    "        else:\n",
    "            import sys\n",
    "            sys.exit(\"aborted, enter correct df type\")\n",
    "        dataframe = dataframe.loc[:,list(database4[database4['phylum'].str.contains(phylum)].set_index('code').index)+list(alt[alt['dominant phylum'].str.contains(phylum)].index)]\n",
    "        dataframe_col = dataframe_col.loc[:,list(database4[database4['phylum'].str.contains(phylum)].set_index('code').index)+list(alt[alt['dominant phylum'].str.contains(phylum)].index)]\n",
    "        alt55 = database_gen_combined.loc[dataframe.columns,:]\n",
    "        grouped = ', ungrouped clusters,\\n'\n",
    "    else:\n",
    "        \n",
    "        if KEGGhitType == 'all':\n",
    "            dataframe = pd.read_pickle('known_genomes_all_grp')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_all_grp_col')\n",
    "            subtitle = 'all KEGG Module hits'\n",
    "        elif KEGGhitType == 'conservative':\n",
    "            dataframe = pd.read_pickle('known_genomes_cons_grp')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_cons_grp_col')\n",
    "            subtitle = 'complete KEGG Modules and Modules missing one block'\n",
    "        elif KEGGhitType == 'complete':\n",
    "            dataframe = pd.read_pickle('known_genomes_complete_grp')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_complete_grp_col')\n",
    "            subtitle = 'complete KEGG Modules only'\n",
    "        else:\n",
    "            import sys\n",
    "            sys.exit(\"aborted, enter correct df type\")\n",
    "        dataframe = dataframe.loc[:,list(database4[database4['phylum'].str.contains(phylum)].set_index('code').index)+list(alt5[alt5['phylum'].str.contains(phylum)].index)]\n",
    "        dataframe_col = dataframe_col.loc[:,list(database4[database4['phylum'].str.contains(phylum)].set_index('code').index)+list(alt5[alt5['phylum'].str.contains(phylum)].index)]\n",
    "        alt55 = database_gen_combined_grp.loc[dataframe.columns,:]\n",
    "        grouped = ', grouped clusters,\\n'\n",
    "    \n",
    "    \"\"\"\n",
    "    0 = Energy metabolism\n",
    "    1 = Carb and lipid metabolism\n",
    "    2 = Nucleic acid and aa metabolism\n",
    "    3 = Secondary metabolism\n",
    "    4 = Genetic info proc.\n",
    "    5 = Env. info proc.\n",
    "    6 = Metabolism\n",
    "    7 = Cellular processes\n",
    "    8 = Gene set\n",
    "    \"\"\"\n",
    "    \n",
    "    completeness = KEGGhitType\n",
    "    \n",
    "\n",
    "    ############################################################################################\n",
    "    # cluster correlation for KEGG B-level = environmental info processing\n",
    "    ############################################################################################\n",
    "\n",
    "    dataC = dataframe.fillna(0).corr().dropna(how='all').T.dropna(how='all')\n",
    "    clustindexlist = dataC.index\n",
    "    \n",
    "    D = pdist(dataC, 'euclidean')\n",
    "    Z = linkage(D, 'ward')\n",
    "    row_link = _optimal_order(dataC, metric='correlation')\n",
    "\n",
    "    optimal_Z = optimal_leaf_ordering(Z, D)\n",
    "\n",
    "    cgClust = sns.clustermap(dataC, row_linkage=row_link, col_linkage=row_link,figsize=(20,40))\n",
    "\n",
    "\n",
    "    ############################################################################################\n",
    "    # KEGG-clust correlation for KEGG B-level = environmental info processing\n",
    "    ############################################################################################\n",
    "\n",
    "    #dataK = alt6.ix[idxblevdf['column'].values,:].T.corr() #np.random.choice(10000, (n, 1), replace=False)\n",
    "    dataK = dataframe.fillna(0).T.corr()\n",
    "\n",
    "    dataK = dataK.dropna(how='all').T.dropna(how='all')\n",
    "\n",
    "    D = pdist(dataK, 'euclidean')\n",
    "    Z = linkage(D, 'ward')\n",
    "    row_link = _optimal_order(dataK, metric='correlation')\n",
    "\n",
    "    optimal_Z = optimal_leaf_ordering(Z, D)\n",
    "\n",
    "    cgKEGG_blevel = sns.clustermap(dataK, row_linkage=row_link, col_linkage=row_link,figsize=(10,10))\n",
    "\n",
    "    # sort by clustermap, \n",
    "    for blev in range(0,1):\n",
    "\n",
    "        lut = {'Archaea':'r','Bacteria':'b','Unassigned':'k','Eukaryota':'c','Viruses':'m'}\n",
    "\n",
    "        \n",
    "            \n",
    "        idx = cgKEGG_blevel.dendrogram_col.reordered_ind\n",
    "        idxcl = cgClust.dendrogram_col.reordered_ind\n",
    "            \n",
    "        \n",
    "        cmap=ListedColormap([\"#e74c3c\", \"#3498db\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#9b59b6\"])\n",
    "        cmap=ListedColormap([\"#e74c3c\", \"#9b59b6\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#3498db\"])#[red,lightblue,green,GREY,darkblue,purple]\n",
    "        #cmap = LinearSegmentedColormap.from_list('Custom', myColors, len(myColors))\n",
    "        \n",
    "        \n",
    "\n",
    "        \n",
    "        \n",
    "        f = plt.figure()\n",
    "        gs = gridspec.GridSpec(45,38)\n",
    "\n",
    "\n",
    "        dataK2 = dataK.copy()\n",
    "        dataK2.index = mod_df.set_index('Module').loc[dataK.index]['Module_combined']\n",
    "\n",
    "        ax1 = f.add_subplot(gs[4:36,0:11])\n",
    "        sns.heatmap(dataK2.iloc[idx,idx],cbar=False,ax=ax1,xticklabels=True,yticklabels=True,cmap=\"RdBu_r\",vmin=-1,vmax=1)\n",
    "        plt.xticks([]);plt.yticks(fontsize=35)\n",
    "        plt.xlabel('KEGG modules',fontsize=30,labelpad=40)\n",
    "\n",
    "        ###############################################\n",
    "        ######### the sorter #####################\n",
    "        ###############################################\n",
    "        #idx2 = alt.sort_values('assembly').index\n",
    "        ###############################################\n",
    "        ###############################################\n",
    "\n",
    "\n",
    "        ax2 = f.add_subplot(gs[4:36,13:32])\n",
    "        sns.heatmap(dataframe_col.fillna(0).loc[dataK.iloc[idx,:].index,dataC.iloc[idxcl,:].index]\n",
    "                    ,ax=ax2,cbar=False,linewidth=0.5,cmap=cmap,vmin=-1,vmax=2,xticklabels=True)\n",
    "        plt.yticks([])\n",
    "        plt. title(phylum+grouped+subtitle,fontsize=70,y=1.05)\n",
    "        #plt.xlabel('Cluster number')\n",
    "        #plt.xticks(rotation=0)\n",
    "\n",
    "        ax13 = f.add_subplot(gs[37:,13:32])\n",
    "        sns.heatmap(dataC.iloc[idxcl,idxcl],cbar=False,ax=ax13,cmap=\"RdBu_r\",vmin=-1,vmax=1)\n",
    "        plt.yticks([])\n",
    "        plt.xticks([])\n",
    "\n",
    "        ##############################################################################\n",
    "        #################### plot hbars on right #####################################\n",
    "        ##############################################################################\n",
    "        \n",
    "\n",
    "        database = dataframe.loc[dataK.iloc[idx,:].index,alt55[~alt55.index.str.contains('Obs')].index]\n",
    "        database = database.loc[:,alt55[~alt55.index.str.contains('M')].index].T.sum().fillna(0)\n",
    "        experim = dataframe.loc[dataK.iloc[idx,:].index,alt55[alt55.index.str.contains('M')].index]\n",
    "        experim = experim.join(dataframe.loc[dataK.iloc[idx,:].index,alt55[alt55.index.str.contains('Obs')].index],how='outer').T.sum()\n",
    "        #vir = dataframe.loc[dataK.iloc[idx,:].index,alt55[alt55['domain']=='Viruses'].index].T.sum()\n",
    "        #una = dataframe.loc[dataK.iloc[idx,:].index,alt55[alt55['domain']=='Unassigned'].index].T.sum()\n",
    "\n",
    "        ax3 = f.add_subplot(gs[4:36,32]) #genomes from KEGG database\n",
    "        database = database/database.sum()\n",
    "        database.plot.barh(ax=ax3,sharey=True, color='#34495e')\n",
    "        ax3.grid(False)\n",
    "        plt.title('Database',fontsize=25)\n",
    "        plt.xticks(rotation=90)\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        ax4 = f.add_subplot(gs[4:36,33]) #found modules in experiment \n",
    "        experim = experim/experim.sum()\n",
    "        experim.plot.barh(ax=ax4,sharey=True, color=\"#3498db\")\n",
    "        ax4.grid(False)\n",
    "        plt.title('Experiment',fontsize=25)\n",
    "        plt.xticks(rotation=90)\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        \"\"\"\n",
    "        ax5 = f.add_subplot(gs[4:36,34]) #Viruses\n",
    "        vir = vir/vir.sum()\n",
    "        vir.plot.barh(ax=ax5,sharey=True, color=\"#9b59b6\")\n",
    "        ax5.grid(False)\n",
    "        plt.title('Viruses',fontsize=25)\n",
    "        plt.xticks(rotation=90)\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\"\"\"\n",
    "\n",
    "        \"\"\"\n",
    "        ax6 = f.add_subplot(gs[4:36,35]) #Unassigned\n",
    "        una = una/una.sum()\n",
    "        una.plot.barh(ax=ax6,sharey=True, color=\"#3498db\")\n",
    "        ax6.grid(False)\n",
    "        plt.title('Unassigned',fontsize=25)\n",
    "        plt.xticks(rotation=90)\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\"\"\"\n",
    "\n",
    "        ax14 = f.add_subplot(gs[4:36,34:36]) #Archaea-Bacteria    \n",
    "        database = database/database.max()\n",
    "        experim = experim/experim.max()\n",
    "        diff = experim-database\n",
    "\n",
    "        diff.plot.barh(ax=ax14,sharey=True, color=\"#e79f3c\")\n",
    "        ax14.grid(False)\n",
    "        plt.title('Expt-Datab\\ndifference',fontsize=25)\n",
    "        plt.xticks(rotation=90)\n",
    "        plt.axvline(x=0);plt.xlim(-1,1)\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        ##############################################################################\n",
    "        ##########################   plot top bars   #################################\n",
    "        ##############################################################################\n",
    "\n",
    "        #ax7 = f.add_subplot(gs[0,13:32]) #phylum\n",
    "        \n",
    "        \"\"\"\n",
    "        col = sns.color_palette(\"cubehelix\", len(list(alt55.loc[dataC.index]['phylum'].unique())))\n",
    "        keys = list(alt55.loc[dataC.index]['phylum'].unique());values = [i for i in col]\n",
    "        lut = dict(zip(keys,values))\n",
    "        lut.update({'Unassigned':\"#3498db\"})\n",
    "        color = alt55['phylum'].map(lut)\n",
    "        print(color)\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(0., 1.02, 1., .102), loc=3,\n",
    "                   ncol=6, mode=\"expand\", borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax7.set_ylabel('Phylum', rotation=0, fontsize=30, labelpad=90)\n",
    "\n",
    "\n",
    "\n",
    "        #3498db\n",
    "        ax8 = f.add_subplot(gs[1,13:32]) #domain\n",
    "        lut = {'Archaea':\"#e74c3c\",'Bacteria':'#34495e','Unassigned':\"#3498db\",'Eukaryota':'c','Viruses':\"#9b59b6\"}\n",
    "\n",
    "        color = alt55['domain'].map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=2, borderaxespad=0.,fontsize=20)\n",
    "        \n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax8.set_ylabel('Kingdom', rotation=0, fontsize=30, labelpad=90)\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "        \"\"\"        \n",
    "        ax9 = f.add_subplot(gs[2,13:32]) #assembly type \n",
    "        \n",
    "\n",
    "            col = sns.color_palette(\"BrBG\", len(alt5['assembly'].unique()))\n",
    "            keys = np.sort(alt5['assembly'].unique())\n",
    "            keys = keys[[1,2,0,5,6,4,3]] # order from most bulk to most minimetagenomic assembly type (with mega being the lesser of two bulks)\n",
    "            values = [i for i in col]\n",
    "            lut = dict(zip(keys,values))\n",
    "            color = alt5['assembly'].map(lut)\n",
    "        else:\n",
    "            col = sns.color_palette(\"BrBG\", len(alt['assembly'].unique()))\n",
    "            keys = np.sort(alt['assembly'].unique())\n",
    "            values = [i for i in col]\n",
    "            lut = dict(zip(keys,values))\n",
    "            color = alt['assembly'].map(lut)\n",
    "                \n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=3, borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        ax9.set_ylabel('Assembly', rotation=0, fontsize=30, labelpad=90)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        ax10 = f.add_subplot(gs[3,13:32]) #experiment\n",
    "        col = sns.color_palette(\"Set2\", 5)\n",
    "        if group == 'YES':\n",
    "            lut = {'Obs2':col[0],'Obs3':col[1],'Obs4':col[2],'Obs5':col[3],'Obs6':col[4]}\n",
    "            color2 = alt5['sample'].map(lut)\n",
    "        else:\n",
    "            lut = {'Obsidian2':col[0],'Obsidian3':col[1],'Obsidian4':col[2],'Obsidian5':col[3],'Obsidian6':col[4]}\n",
    "            color2 = alt['sample'].map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=3, borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color2[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax10.set_ylabel('Experiment', rotation=0, fontsize=30, labelpad=100)\n",
    "        \"\"\"\n",
    "\n",
    "        ##############################################################################\n",
    "        ##########################   plot left side bars  ############################\n",
    "        ##############################################################################\n",
    "        Blevel = blev\n",
    "        \n",
    "        \"\"\"\n",
    "        0 = Energy metabolism\n",
    "        1 = Carb and lipid metabolism\n",
    "        2 = Nucleic acid and aa metabolism\n",
    "        3 = Secondary metabolism\n",
    "        4 = Genetic info proc.\n",
    "        5 = Env. info proc.\n",
    "        6 = Metabolism\n",
    "        7 = Cellular processes\n",
    "        8 = Gene set\"\"\"\n",
    "        \n",
    "        # B-level assignment\n",
    "        ax11 = f.add_subplot(gs[4:36,11])\n",
    "        col = sns.color_palette(\"gist_earth\", len(KEGGmoddf.B.unique())-1)\n",
    "        keys = list(KEGGmoddf.B.unique());keys = keys[0:-1];values = [i for i in col]\n",
    "        keysC = KEGGmoddf[KEGGmoddf['B']==keys[Blevel]].C.unique();\n",
    "\n",
    "        lut = dict(zip(keys,values))\n",
    "        idxlist = dataK.iloc[idx,:].index\n",
    "        #idxlist2 = [i[0:6] for i in idxlist]\n",
    "        D_group = KEGGmoddf.groupby('D-module').sum()\n",
    "\n",
    "        for i in D_group.index:\n",
    "            a = D_group.loc[i,'B']\n",
    "            for j in keys:\n",
    "                if a.find(j)!=-1:\n",
    "                    D_group.loc[i,'B'] = j\n",
    "            c = D_group.loc[i,'C']\n",
    "            for j in KEGGmoddf.C.unique():\n",
    "                if c.find(j)!=-1:\n",
    "                    D_group.loc[i,'C'] = j\n",
    "\n",
    "        B_list  = D_group.loc[idxlist]['B']\n",
    "        #B_list.replace(B_list.index,idxlist)\n",
    "        color = B_list.map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(-12, 1.02, 1., .102), loc=3,\n",
    "                   ncol=1, mode=\"expand\", borderaxespad=0.,fontsize=30)\n",
    "        \n",
    "        plt.barh(list(range(0,len(dataK))),list(np.ones(len(dataK))),color=color,height=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.ylim(-.5,len(dataK)-.5)\n",
    "        plt.xlim(0,1)\n",
    "        ax11.set_xlabel('B', rotation=0, fontsize=30, labelpad=40)\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "\n",
    "        #C-level assignment\n",
    "\n",
    "        ax12 = f.add_subplot(gs[4:36,12])\n",
    "        col = sns.color_palette(\"Set3\", len(keysC));values = [i for i in col]\n",
    "        lutC = dict(zip(keysC,values))\n",
    "\n",
    "        C_list  = D_group.loc[idxlist]['C']\n",
    "        C_list.replace(C_list.index,idxlist)\n",
    "        color = C_list.map(lutC)\n",
    "        color = color.fillna('None')\n",
    "\n",
    "        plt.barh(list(range(0,len(dataK))),list(np.ones(len(dataK))),color=color,height=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.ylim(-.5,len(dataK)-.5)\n",
    "        plt.xlim(0,1)\n",
    "        ax12.set_xlabel('C', rotation=0, fontsize=30, labelpad=40)\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        for x,y in lutC.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(-8, 1.02, 1., .102), loc=3,\n",
    "                   ncol=1, mode=\"expand\", borderaxespad=0.,fontsize=30)\n",
    "\n",
    "        \n",
    "        \n",
    "        gs.update(wspace=.05,hspace=.05)\n",
    "        \n",
    "        f.set_figheight(55)\n",
    "        #plt.show()\n",
    "        if group =='YES':\n",
    "            f.set_figwidth(60)\n",
    "            f.savefig('OverviewHeatmaps/Grouped/Known/Obsidian_grouped_KEGG_'+completeness+'_'+phylum+'_'+keys[Blevel])\n",
    "            #f.savefig('OverviewHeatmaps/Grouped/Known/pdfs/Obsidian_grouped_KEGG_'+completeness+'_'+keys[Blevel]+'.pdf')\n",
    "        else:\n",
    "            f.set_figwidth(45)\n",
    "            f.savefig('OverviewHeatmaps/Ungrouped/Known/Obsidian_ungrouped_KEGG_'+completeness+'_'+phylum+'_'+keys[Blevel])\n",
    "            #f.savefig('OverviewHeatmaps/Ungrouped/Known/pdfs/Obsidian_ungrouped_KEGG_'+completeness+'_'+keys[Blevel]+'.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plotClusterOverviewTot_gen('conservative','YES','Nitrospirae')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5.9160797830996155e-07"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "an = (1.5e-1*0.7e-2)/3e9\n",
    "np.sqrt(an)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Gammaproteobacteria - Enterobacteria',\n",
       "       'Gammaproteobacteria - Others', 'Betaproteobacteria',\n",
       "       'Epsilonproteobacteria', 'Deltaproteobacteria',\n",
       "       'Alphaproteobacteria', 'Other proteobacteria',\n",
       "       'Firmicutes - Bacilli', 'Firmicutes - Clostridia',\n",
       "       'Firmicutes - Others', 'Tenericutes', 'Actinobacteria',\n",
       "       'Cyanobacteria', 'Melainabacteria', 'Chloroflexi',\n",
       "       'Deinococcus-Thermus', 'Armatimonadetes',\n",
       "       'Unclassified Terrabacteria group', 'Chlamydiae', 'Verrucomicrobia',\n",
       "       'Planctomycetes', 'Kiritimatiellaeota', 'Spirochaetes',\n",
       "       'Acidobacteria', 'Elusimicrobia', 'Fusobacteria', 'Synergistetes',\n",
       "       'Fibrobacteres', 'Gemmatimonadetes', 'Bacteroidetes', 'Chlorobi',\n",
       "       'Cloacimonetes', 'Aquificae', 'Thermotogae', 'Caldiserica',\n",
       "       'Chrysiogenetes', 'Deferribacteres', 'Calditrichaeota',\n",
       "       'Dictyoglomi', 'Nitrospirae', 'Thermodesulfobacteria',\n",
       "       'Saccharibacteria', 'Peregrinibacteria', 'Other',\n",
       "       'Unclassified Bacteria', 'Euryarchaeota', 'Crenarchaeota',\n",
       "       'Thaumarchaeota', 'Nanoarchaeota', 'Nanohaloarchaeota',\n",
       "       'Micrarchaeota', 'Korarchaeota', 'Bathyarchaeota', 'Lokiarchaeota',\n",
       "       'Unclassified Archaea'], dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "database4[database4['domain']=='Prokaryotes']['phylum'].unique()\n",
    "\n",
    "\"\"\"for p in phylumlist:\n",
    "    phylum = p;#'Euryarchaeota'\n",
    "    plotClusterOverviewTot_gen('all','YES',phylum)\n",
    "    plotClusterOverviewTot_gen('complete','YES',phylum)\n",
    "    plotClusterOverviewTot_gen('conservative','YES',phylum)\n",
    "    plotClusterOverviewTot_gen('all','nope',phylum)\n",
    "    plotClusterOverviewTot_gen('complete','nope',phylum)\n",
    "    plotClusterOverviewTot_gen('conservative','nope',phylum)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Unassigned', 'Crenarchaeota', 'Nanoarchaeota', 'Aquificae',\n",
       "       'Proteobacteria', 'Thermodesulfobacteria', 'Thaumarchaeota',\n",
       "       'Ignavibacteriae', 'Fervidibacteria', 'Firmicutes', 'Dictyoglomi',\n",
       "       'Aigarchaeota', 'Planctomycetes', 'Chloroflexi', 'Chlamydiae',\n",
       "       'Euryarchaeota', 'Bacteroidetes', 'Candidatus Bathyarchaeota',\n",
       "       'dsDNA viruses', 'Fervidibacteria AND Ignavibacteriae',\n",
       "       'Thermotogae', 'Cyanobacteria', 'Aminicenantes', 'Parcubacteria',\n",
       "       'Nitrospirae', 'Candidatus Micrarchaeota',\n",
       "       'dsDNA viruses AND dsDNA viruses, no RNA stage'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt['dominant phylum'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly</th>\n",
       "      <th>orig doms</th>\n",
       "      <th>orig phyl</th>\n",
       "      <th>clustsum</th>\n",
       "      <th>sample</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M72</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>Candidatus Micrarchaeota</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>megameta</td>\n",
       "      <td>ArchaeaArchaea</td>\n",
       "      <td>Candidatus MicrarchaeotaCandidatus Micrarchaeota</td>\n",
       "      <td>Obs6_mega_12Obs6_meta_11</td>\n",
       "      <td>Obs6</td>\n",
       "      <td>3187531.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      domain                    phylum       species  assembly  \\\n",
       "M72  Archaea  Candidatus Micrarchaeota  unclassified  megameta   \n",
       "\n",
       "          orig doms                                         orig phyl  \\\n",
       "M72  ArchaeaArchaea  Candidatus MicrarchaeotaCandidatus Micrarchaeota   \n",
       "\n",
       "                     clustsum sample     length  \n",
       "M72  Obs6_mega_12Obs6_meta_11   Obs6  3187531.0  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt5[alt5['phylum'].str.contains('Micrarchaeota')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "phylumlist = ['Crenarchaeota', 'Thermotogae', 'Cyanobacteria', 'Nanoarchaeota', 'Aquificae',\n",
    "              'Nitrospirae','Thermodesulfobacteria', 'Thaumarchaeota','Chloroflexi', 'Chlamydiae','Aminicenantes', 'Parcubacteria',]\n",
    "phylumlist = 'Micrarchaeota'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "############################################################################################\n",
    "### plot the overview plots with all KEGG hits and all genomes \n",
    "############################################################################################\n",
    "\n",
    "def plotClusterOverviewTot_gen_names(KEGGhitType,group,phylum):\n",
    "    \n",
    "    if group != \"YES\":\n",
    "        \n",
    "        if KEGGhitType == 'all':\n",
    "            dataframe = pd.read_pickle('known_genomes_all')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_all_col')\n",
    "            subtitle = 'all KEGG Module hits'\n",
    "        elif KEGGhitType == 'conservative':\n",
    "            dataframe = pd.read_pickle('known_genomes_cons')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_cons_col')\n",
    "            subtitle = 'complete KEGG Modules and Modules missing one block'\n",
    "        elif KEGGhitType == 'complete':\n",
    "            dataframe = pd.read_pickle('known_genomes_complete')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_complete_col')\n",
    "            subtitle = 'complete KEGG Modules only'\n",
    "        else:\n",
    "            import sys\n",
    "            sys.exit(\"aborted, enter correct df type\")\n",
    "        dataframe = dataframe.loc[:,list(database4[database4['phylum'].str.contains(phylum)].set_index('code').index)+list(alt[alt['dominant phylum'].str.contains(phylum)].index)]\n",
    "        dataframe_col = dataframe_col.loc[:,list(database4[database4['phylum'].str.contains(phylum)].set_index('code').index)+list(alt[alt['dominant phylum'].str.contains(phylum)].index)]\n",
    "        alt55 = database_gen_combined.loc[dataframe.columns,:]\n",
    "        grouped = ', ungrouped clusters,\\n'\n",
    "    else:\n",
    "        \n",
    "        if KEGGhitType == 'all':\n",
    "            dataframe = pd.read_pickle('known_genomes_all_grp')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_all_grp_col')\n",
    "            subtitle = 'all KEGG Module hits'\n",
    "        elif KEGGhitType == 'conservative':\n",
    "            dataframe = pd.read_pickle('known_genomes_cons_grp')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_cons_grp_col')\n",
    "            subtitle = 'complete KEGG Modules and Modules missing one block'\n",
    "        elif KEGGhitType == 'complete':\n",
    "            dataframe = pd.read_pickle('known_genomes_complete_grp')\n",
    "            dataframe_col = pd.read_pickle('known_genomes_complete_grp_col')\n",
    "            subtitle = 'complete KEGG Modules only'\n",
    "        else:\n",
    "            import sys\n",
    "            sys.exit(\"aborted, enter correct df type\")\n",
    "        dataframe = dataframe.loc[:,list(database4[database4['phylum'].str.contains(phylum)].set_index('code').index)+list(alt5[alt5['phylum'].str.contains(phylum)].index)]\n",
    "        dataframe_col = dataframe_col.loc[:,list(database4[database4['phylum'].str.contains(phylum)].set_index('code').index)+list(alt5[alt5['phylum'].str.contains(phylum)].index)]\n",
    "        alt55 = database_gen_combined_grp.loc[dataframe.columns,:]\n",
    "        grouped = ', grouped clusters,\\n'\n",
    "    \n",
    "    \"\"\"\n",
    "    0 = Energy metabolism\n",
    "    1 = Carb and lipid metabolism\n",
    "    2 = Nucleic acid and aa metabolism\n",
    "    3 = Secondary metabolism\n",
    "    4 = Genetic info proc.\n",
    "    5 = Env. info proc.\n",
    "    6 = Metabolism\n",
    "    7 = Cellular processes\n",
    "    8 = Gene set\n",
    "    \"\"\"\n",
    "    \n",
    "    completeness = KEGGhitType\n",
    "    \n",
    "\n",
    "    ############################################################################################\n",
    "    # cluster correlation for KEGG B-level = environmental info processing\n",
    "    ############################################################################################\n",
    "\n",
    "    dataC = dataframe.fillna(0).corr().dropna(how='all').T.dropna(how='all')\n",
    "    clustindexlist = dataC.index\n",
    "    \n",
    "    D = pdist(dataC, 'euclidean')\n",
    "    Z = linkage(D, 'ward')\n",
    "    row_link = _optimal_order(dataC, metric='correlation')\n",
    "\n",
    "    optimal_Z = optimal_leaf_ordering(Z, D)\n",
    "\n",
    "    cgClust = sns.clustermap(dataC, row_linkage=row_link, col_linkage=row_link,figsize=(20,40))\n",
    "\n",
    "\n",
    "    ############################################################################################\n",
    "    # KEGG-clust correlation for KEGG B-level = environmental info processing\n",
    "    ############################################################################################\n",
    "\n",
    "    #dataK = alt6.ix[idxblevdf['column'].values,:].T.corr() #np.random.choice(10000, (n, 1), replace=False)\n",
    "    dataK = dataframe.fillna(0).T.corr()\n",
    "\n",
    "    dataK = dataK.dropna(how='all').T.dropna(how='all')\n",
    "\n",
    "    D = pdist(dataK, 'euclidean')\n",
    "    Z = linkage(D, 'ward')\n",
    "    row_link = _optimal_order(dataK, metric='correlation')\n",
    "\n",
    "    optimal_Z = optimal_leaf_ordering(Z, D)\n",
    "\n",
    "    cgKEGG_blevel = sns.clustermap(dataK, row_linkage=row_link, col_linkage=row_link,figsize=(10,10))\n",
    "\n",
    "    # sort by clustermap, \n",
    "    for blev in range(0,1):\n",
    "\n",
    "        lut = {'Archaea':'r','Bacteria':'b','Unassigned':'k','Eukaryota':'c','Viruses':'m'}\n",
    "\n",
    "        \n",
    "            \n",
    "        idx = cgKEGG_blevel.dendrogram_col.reordered_ind\n",
    "        idxcl = cgClust.dendrogram_col.reordered_ind\n",
    "            \n",
    "        \n",
    "        cmap=ListedColormap([\"#e74c3c\", \"#3498db\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#9b59b6\"])\n",
    "        cmap=ListedColormap([\"#e74c3c\", \"#9b59b6\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#3498db\"])#[red,lightblue,green,GREY,darkblue,purple]\n",
    "        #cmap = LinearSegmentedColormap.from_list('Custom', myColors, len(myColors))\n",
    "        \n",
    "        \n",
    "\n",
    "        \n",
    "        \n",
    "        f = plt.figure()\n",
    "        gs = gridspec.GridSpec(45,38)\n",
    "\n",
    "\n",
    "        dataK2 = dataK.copy()\n",
    "        dataK2.index = mod_df.set_index('Module').loc[dataK.index]['Module_combined']\n",
    "        dataframe_col = dataframe_col.fillna(0).loc[dataK.iloc[idx,:].index,dataC.iloc[idxcl,:].index]\n",
    "        \n",
    "        \n",
    "        dataframe_col.index = mod_df.set_index('Module').loc[dataframe_col.index,:]['Module_combined']\n",
    "        \n",
    "     \n",
    "        ax1 = f.add_subplot(gs[4:36,:1])\n",
    "        sns.heatmap(dataK2.iloc[idx,idx],cbar=False,ax=ax1,xticklabels=True,yticklabels=True,cmap=\"RdBu_r\",vmin=-1,vmax=1)\n",
    "        plt.xticks([]);plt.yticks(fontsize=15)\n",
    "        plt.xlabel('KEGG modules',fontsize=30,labelpad=40)\n",
    "\n",
    "        ###############################################\n",
    "        ######### the sorter #####################\n",
    "        ###############################################\n",
    "        #idx2 = alt.sort_values('assembly').index\n",
    "        ###############################################\n",
    "        ###############################################\n",
    "\n",
    "        \n",
    "        ax2 = f.add_subplot(gs[4:36,2:32])\n",
    "        sns.heatmap(dataframe_col\n",
    "                    ,ax=ax2,cbar=False,linewidth=0.5,cmap=cmap\n",
    "                    ,vmin=-1,vmax=2,xticklabels=True,yticklabels=True)\n",
    "        \n",
    "        #plt.ylabel()\n",
    "        plt.title(phylum+grouped+subtitle,fontsize=25,y=1.05)\n",
    "        \n",
    "        #plt.xlabel('Cluster number')\n",
    "        #plt.xticks(rotation=0)\n",
    "\n",
    "        \"\"\"        ax13 = f.add_subplot(gs[37:,13:32])\n",
    "        sns.heatmap(dataC.iloc[idxcl,idxcl],cbar=False,ax=ax13,cmap=\"RdBu_r\",vmin=-1,vmax=1)\n",
    "        plt.yticks([])\n",
    "        plt.xticks([])\"\"\"\n",
    "\n",
    "        ##############################################################################\n",
    "        #################### plot hbars on right #####################################\n",
    "        ##############################################################################\n",
    "        \n",
    "\n",
    "        database = dataframe.loc[dataK.iloc[idx,:].index,alt55[~alt55.index.str.contains('Obs')].index]\n",
    "        database = database.loc[:,alt55[~alt55.index.str.contains('M')].index].T.sum().fillna(0)\n",
    "        experim = dataframe.loc[dataK.iloc[idx,:].index,alt55[alt55.index.str.contains('M')].index]\n",
    "        experim = experim.join(dataframe.loc[dataK.iloc[idx,:].index,alt55[alt55.index.str.contains('Obs')].index],how='outer').T.sum()\n",
    "        #vir = dataframe.loc[dataK.iloc[idx,:].index,alt55[alt55['domain']=='Viruses'].index].T.sum()\n",
    "        #una = dataframe.loc[dataK.iloc[idx,:].index,alt55[alt55['domain']=='Unassigned'].index].T.sum()\n",
    "\n",
    "        ax3 = f.add_subplot(gs[4:36,32]) #genomes from KEGG database\n",
    "        database = database/database.sum()\n",
    "        database.plot.barh(ax=ax3,sharey=True, color='#34495e')\n",
    "        ax3.grid(False)\n",
    "        plt.title('Database',fontsize=15,rotation=45,y=1.05,x=1.05)\n",
    "        #plt.xticks(rotation=90)\n",
    "        plt.xticks([])\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        ax4 = f.add_subplot(gs[4:36,33]) #found modules in experiment \n",
    "        experim = experim/experim.sum()\n",
    "        experim.plot.barh(ax=ax4,sharey=True, color=\"#3498db\")\n",
    "        ax4.grid(False)\n",
    "        plt.title('Experiment',fontsize=15,rotation=45,y=1.055,x=1.05)\n",
    "        #plt.xticks(rotation=90)\n",
    "        plt.xticks([])\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        \"\"\"\n",
    "        ax5 = f.add_subplot(gs[4:36,34]) #Viruses\n",
    "        vir = vir/vir.sum()\n",
    "        vir.plot.barh(ax=ax5,sharey=True, color=\"#9b59b6\")\n",
    "        ax5.grid(False)\n",
    "        plt.title('Viruses',fontsize=25)\n",
    "        plt.xticks(rotation=90)\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\"\"\"\n",
    "\n",
    "        \"\"\"\n",
    "        ax6 = f.add_subplot(gs[4:36,35]) #Unassigned\n",
    "        una = una/una.sum()\n",
    "        una.plot.barh(ax=ax6,sharey=True, color=\"#3498db\")\n",
    "        ax6.grid(False)\n",
    "        plt.title('Unassigned',fontsize=25)\n",
    "        plt.xticks(rotation=90)\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\"\"\"\n",
    "\n",
    "        ax14 = f.add_subplot(gs[4:36,34:36]) #Archaea-Bacteria    \n",
    "        database = database/database.max()\n",
    "        experim = experim/experim.max()\n",
    "        diff = experim-database\n",
    "\n",
    "        diff.plot.barh(ax=ax14,sharey=True, color=\"#e79f3c\")\n",
    "        ax14.grid(False)\n",
    "        plt.title('Expt-Datab\\ndifference',fontsize=15,rotation=45,y=1.05,x=1.01)\n",
    "        #plt.xticks(rotation=90)\n",
    "        plt.axvline(x=0);plt.xlim(-1,1)\n",
    "        plt.gca().invert_yaxis()\n",
    "        plt.xticks([])\n",
    "\n",
    "        ##############################################################################\n",
    "        ##########################   plot top bars   #################################\n",
    "        ##############################################################################\n",
    "\n",
    "        #ax7 = f.add_subplot(gs[0,13:32]) #phylum\n",
    "        \n",
    "        \"\"\"\n",
    "        col = sns.color_palette(\"cubehelix\", len(list(alt55.loc[dataC.index]['phylum'].unique())))\n",
    "        keys = list(alt55.loc[dataC.index]['phylum'].unique());values = [i for i in col]\n",
    "        lut = dict(zip(keys,values))\n",
    "        lut.update({'Unassigned':\"#3498db\"})\n",
    "        color = alt55['phylum'].map(lut)\n",
    "        print(color)\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(0., 1.02, 1., .102), loc=3,\n",
    "                   ncol=6, mode=\"expand\", borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax7.set_ylabel('Phylum', rotation=0, fontsize=30, labelpad=90)\n",
    "\n",
    "\n",
    "\n",
    "        #3498db\n",
    "        ax8 = f.add_subplot(gs[1,13:32]) #domain\n",
    "        lut = {'Archaea':\"#e74c3c\",'Bacteria':'#34495e','Unassigned':\"#3498db\",'Eukaryota':'c','Viruses':\"#9b59b6\"}\n",
    "\n",
    "        color = alt55['domain'].map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=2, borderaxespad=0.,fontsize=20)\n",
    "        \n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax8.set_ylabel('Kingdom', rotation=0, fontsize=30, labelpad=90)\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "        \"\"\"        \n",
    "        ax9 = f.add_subplot(gs[2,13:32]) #assembly type \n",
    "        \n",
    "\n",
    "            col = sns.color_palette(\"BrBG\", len(alt5['assembly'].unique()))\n",
    "            keys = np.sort(alt5['assembly'].unique())\n",
    "            keys = keys[[1,2,0,5,6,4,3]] # order from most bulk to most minimetagenomic assembly type (with mega being the lesser of two bulks)\n",
    "            values = [i for i in col]\n",
    "            lut = dict(zip(keys,values))\n",
    "            color = alt5['assembly'].map(lut)\n",
    "        else:\n",
    "            col = sns.color_palette(\"BrBG\", len(alt['assembly'].unique()))\n",
    "            keys = np.sort(alt['assembly'].unique())\n",
    "            values = [i for i in col]\n",
    "            lut = dict(zip(keys,values))\n",
    "            color = alt['assembly'].map(lut)\n",
    "                \n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=3, borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        ax9.set_ylabel('Assembly', rotation=0, fontsize=30, labelpad=90)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        ax10 = f.add_subplot(gs[3,13:32]) #experiment\n",
    "        col = sns.color_palette(\"Set2\", 5)\n",
    "        if group == 'YES':\n",
    "            lut = {'Obs2':col[0],'Obs3':col[1],'Obs4':col[2],'Obs5':col[3],'Obs6':col[4]}\n",
    "            color2 = alt5['sample'].map(lut)\n",
    "        else:\n",
    "            lut = {'Obsidian2':col[0],'Obsidian3':col[1],'Obsidian4':col[2],'Obsidian5':col[3],'Obsidian6':col[4]}\n",
    "            color2 = alt['sample'].map(lut)\n",
    "\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=3, borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        plt.bar(list(range(0,len(dataC))),np.ones(len(dataC)),color=color2[dataC.iloc[idxcl,:].index],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(dataC)-.5);plt.ylim(0,1)\n",
    "        ax10.set_ylabel('Experiment', rotation=0, fontsize=30, labelpad=100)\n",
    "        \"\"\"\n",
    "\n",
    "        ##############################################################################\n",
    "        ##########################   plot left side bars  ############################\n",
    "        ##############################################################################\n",
    "        Blevel = blev\n",
    "        \n",
    "        \"\"\"\n",
    "        0 = Energy metabolism\n",
    "        1 = Carb and lipid metabolism\n",
    "        2 = Nucleic acid and aa metabolism\n",
    "        3 = Secondary metabolism\n",
    "        4 = Genetic info proc.\n",
    "        5 = Env. info proc.\n",
    "        6 = Metabolism\n",
    "        7 = Cellular processes\n",
    "        8 = Gene set\"\"\"\n",
    "        \n",
    "        # B-level assignment\n",
    "        #ax11 = f.add_subplot(gs[4:36,1])\n",
    "        col = sns.color_palette(\"gist_earth\", len(KEGGmoddf.B.unique())-1)\n",
    "        keys = list(KEGGmoddf.B.unique());keys = keys[0:-1];values = [i for i in col]\n",
    "        keysC = KEGGmoddf[KEGGmoddf['B']==keys[Blevel]].C.unique();\n",
    "\n",
    "        lut = dict(zip(keys,values))\n",
    "        idxlist = dataK.iloc[idx,:].index\n",
    "        #idxlist2 = [i[0:6] for i in idxlist]\n",
    "        D_group = KEGGmoddf.groupby('D-module').sum()\n",
    "\n",
    "        for i in D_group.index:\n",
    "            a = D_group.loc[i,'B']\n",
    "            for j in keys:\n",
    "                if a.find(j)!=-1:\n",
    "                    D_group.loc[i,'B'] = j\n",
    "            c = D_group.loc[i,'C']\n",
    "            for j in KEGGmoddf.C.unique():\n",
    "                if c.find(j)!=-1:\n",
    "                    D_group.loc[i,'C'] = j\n",
    "\n",
    "        B_list  = D_group.loc[idxlist]['B']\n",
    "        #B_list.replace(B_list.index,idxlist)\n",
    "        color = B_list.map(lut)\n",
    "\n",
    "        \"\"\"for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(-12, 1.02, 1., .102), loc=3,\n",
    "                   ncol=1, mode=\"expand\", borderaxespad=0.,fontsize=30)\"\"\"\n",
    "        \n",
    "        \"\"\"plt.barh(list(range(0,len(dataK))),list(np.ones(len(dataK))),color=color,height=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.ylim(-.5,len(dataK)-.5)\n",
    "        plt.xlim(0,1)\n",
    "        ax11.set_xlabel('B', rotation=0, fontsize=30, labelpad=40)\n",
    "        plt.gca().invert_yaxis()\"\"\"\n",
    "\n",
    "\n",
    "        #C-level assignment\n",
    "\n",
    "        ax12 = f.add_subplot(gs[4:36,1])\n",
    "        col = sns.color_palette(\"Set3\", len(keysC));values = [i for i in col]\n",
    "        lutC = dict(zip(keysC,values))\n",
    "\n",
    "        C_list  = D_group.loc[idxlist]['C']\n",
    "        C_list.replace(C_list.index,idxlist)\n",
    "        color = C_list.map(lutC)\n",
    "        color = color.fillna('None')\n",
    "\n",
    "        plt.barh(list(range(0,len(dataK))),list(np.ones(len(dataK))),color=color,height=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.ylim(-.5,len(dataK)-.5)\n",
    "        plt.xlim(0,1)\n",
    "        #ax12.set_xlabel('C', rotation=0, fontsize=30, labelpad=40)\n",
    "        plt.gca().invert_yaxis()\n",
    "        \n",
    "        for x,y in lutC.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(-7, 1.02, 1., .102), loc=3,\n",
    "                   ncol=2, mode=\"expand\", borderaxespad=0.,fontsize=20)\n",
    "        \n",
    "        \n",
    "        \n",
    "        gs.update(wspace=.05,hspace=.05)\n",
    "        \n",
    "        f.set_figheight(35)\n",
    "        #plt.show()\n",
    "        if group =='YES':\n",
    "            f.set_figwidth(35)\n",
    "            f.savefig('OverviewHeatmaps/Grouped/Known/Obsidian_grouped_name_KEGG_'+completeness+'_'+phylum+'_'+keys[Blevel])\n",
    "            #f.savefig('OverviewHeatmaps/Grouped/Known/pdfs/Obsidian_grouped_KEGG_'+completeness+'_'+keys[Blevel]+'.pdf')\n",
    "        else:\n",
    "            f.set_figwidth(35)\n",
    "            f.savefig('OverviewHeatmaps/Ungrouped/Known/Obsidian_ungrouped_name_KEGG_'+completeness+'_'+phylum+'_'+keys[Blevel])\n",
    "            #f.savefig('OverviewHeatmaps/Ungrouped/Known/pdfs/Obsidian_ungrouped_KEGG_'+completeness+'_'+keys[Blevel]+'.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.6/site-packages/matplotlib/pyplot.py:524: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  max_open_warning, RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "plotClusterOverviewTot_gen_names('conservative','nope',phylumlist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D-module</th>\n",
       "      <th>D</th>\n",
       "      <th>E-module</th>\n",
       "      <th>E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2340</th>\n",
       "      <td>Structural complex</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>M00597</td>\n",
       "      <td>Anoxygenic photosystem II [BR:ko00194]</td>\n",
       "      <td>K08928</td>\n",
       "      <td>pufL; photosynthetic reaction center L subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341</th>\n",
       "      <td>Structural complex</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>M00597</td>\n",
       "      <td>Anoxygenic photosystem II [BR:ko00194]</td>\n",
       "      <td>K08929</td>\n",
       "      <td>pufM; photosynthetic reaction center M subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2342</th>\n",
       "      <td>Structural complex</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>M00598</td>\n",
       "      <td>Anoxygenic photosystem I [BR:ko00194]</td>\n",
       "      <td>K08940</td>\n",
       "      <td>pscA; photosystem P840 reaction center large s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2343</th>\n",
       "      <td>Structural complex</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>M00598</td>\n",
       "      <td>Anoxygenic photosystem I [BR:ko00194]</td>\n",
       "      <td>K08941</td>\n",
       "      <td>pscB; photosystem P840 reaction center iron-su...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2344</th>\n",
       "      <td>Structural complex</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>M00598</td>\n",
       "      <td>Anoxygenic photosystem I [BR:ko00194]</td>\n",
       "      <td>K08942</td>\n",
       "      <td>pscC; photosystem P840 reaction center cytochr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2345</th>\n",
       "      <td>Structural complex</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>M00598</td>\n",
       "      <td>Anoxygenic photosystem I [BR:ko00194]</td>\n",
       "      <td>K08943</td>\n",
       "      <td>pscD; photosystem P840 reaction center protein...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4735</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Metabolic capacity</td>\n",
       "      <td>M00612</td>\n",
       "      <td>Anoxygenic photosynthesis in purple bacteria</td>\n",
       "      <td>M00597</td>\n",
       "      <td>Anoxygenic photosystem II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4736</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Metabolic capacity</td>\n",
       "      <td>M00612</td>\n",
       "      <td>Anoxygenic photosynthesis in purple bacteria</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4737</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Metabolic capacity</td>\n",
       "      <td>M00613</td>\n",
       "      <td>Anoxygenic photosynthesis in green nonsulfur b...</td>\n",
       "      <td>M00597</td>\n",
       "      <td>Anoxygenic photosystem II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4738</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Metabolic capacity</td>\n",
       "      <td>M00613</td>\n",
       "      <td>Anoxygenic photosynthesis in green nonsulfur b...</td>\n",
       "      <td>M00376</td>\n",
       "      <td>3-Hydroxypropionate bi-cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4739</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Metabolic capacity</td>\n",
       "      <td>M00614</td>\n",
       "      <td>Anoxygenic photosynthesis in green sulfur bact...</td>\n",
       "      <td>M00598</td>\n",
       "      <td>Anoxygenic photosystem I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4740</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Metabolic capacity</td>\n",
       "      <td>M00614</td>\n",
       "      <td>Anoxygenic photosynthesis in green sulfur bact...</td>\n",
       "      <td>M00173</td>\n",
       "      <td>Reductive citrate cycle (Arnon-Buchanan cycle)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       A                  B                   C D-module  \\\n",
       "2340  Structural complex  Energy metabolism      Photosynthesis   M00597   \n",
       "2341  Structural complex  Energy metabolism      Photosynthesis   M00597   \n",
       "2342  Structural complex  Energy metabolism      Photosynthesis   M00598   \n",
       "2343  Structural complex  Energy metabolism      Photosynthesis   M00598   \n",
       "2344  Structural complex  Energy metabolism      Photosynthesis   M00598   \n",
       "2345  Structural complex  Energy metabolism      Photosynthesis   M00598   \n",
       "4735    Signature module           Gene set  Metabolic capacity   M00612   \n",
       "4736    Signature module           Gene set  Metabolic capacity   M00612   \n",
       "4737    Signature module           Gene set  Metabolic capacity   M00613   \n",
       "4738    Signature module           Gene set  Metabolic capacity   M00613   \n",
       "4739    Signature module           Gene set  Metabolic capacity   M00614   \n",
       "4740    Signature module           Gene set  Metabolic capacity   M00614   \n",
       "\n",
       "                                                      D E-module  \\\n",
       "2340             Anoxygenic photosystem II [BR:ko00194]   K08928   \n",
       "2341             Anoxygenic photosystem II [BR:ko00194]   K08929   \n",
       "2342              Anoxygenic photosystem I [BR:ko00194]   K08940   \n",
       "2343              Anoxygenic photosystem I [BR:ko00194]   K08941   \n",
       "2344              Anoxygenic photosystem I [BR:ko00194]   K08942   \n",
       "2345              Anoxygenic photosystem I [BR:ko00194]   K08943   \n",
       "4735       Anoxygenic photosynthesis in purple bacteria   M00597   \n",
       "4736       Anoxygenic photosynthesis in purple bacteria   M00165   \n",
       "4737  Anoxygenic photosynthesis in green nonsulfur b...   M00597   \n",
       "4738  Anoxygenic photosynthesis in green nonsulfur b...   M00376   \n",
       "4739  Anoxygenic photosynthesis in green sulfur bact...   M00598   \n",
       "4740  Anoxygenic photosynthesis in green sulfur bact...   M00173   \n",
       "\n",
       "                                                      E  \n",
       "2340     pufL; photosynthetic reaction center L subunit  \n",
       "2341     pufM; photosynthetic reaction center M subunit  \n",
       "2342  pscA; photosystem P840 reaction center large s...  \n",
       "2343  pscB; photosystem P840 reaction center iron-su...  \n",
       "2344  pscC; photosystem P840 reaction center cytochr...  \n",
       "2345  pscD; photosystem P840 reaction center protein...  \n",
       "4735                          Anoxygenic photosystem II  \n",
       "4736   Reductive pentose phosphate cycle (Calvin cycle)  \n",
       "4737                          Anoxygenic photosystem II  \n",
       "4738                       3-Hydroxypropionate bi-cycle  \n",
       "4739                           Anoxygenic photosystem I  \n",
       "4740     Reductive citrate cycle (Arnon-Buchanan cycle)  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KEGGmoddf[KEGGmoddf['D'].str.contains('Anoxygenic')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CRISPRS</th>\n",
       "      <th>CRISPRS</th>\n",
       "      <th>NoCOG</th>\n",
       "      <th>assembly</th>\n",
       "      <th>chip</th>\n",
       "      <th>chip counts</th>\n",
       "      <th>cluster#</th>\n",
       "      <th>dominant domain</th>\n",
       "      <th>dominant phylum</th>\n",
       "      <th>dominant species</th>\n",
       "      <th>...</th>\n",
       "      <th>highest_lineage_name</th>\n",
       "      <th>length</th>\n",
       "      <th>majorCOG</th>\n",
       "      <th>mu_x</th>\n",
       "      <th>mu_y</th>\n",
       "      <th>num_contigs</th>\n",
       "      <th>number of cells</th>\n",
       "      <th>sample</th>\n",
       "      <th>sig_x</th>\n",
       "      <th>sig_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mega</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Candidatus Micrarchaeota</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>268227.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-24.559164</td>\n",
       "      <td>46.653121</td>\n",
       "      <td>40.0</td>\n",
       "      <td>5.270336</td>\n",
       "      <td>Obsidian6</td>\n",
       "      <td>3.893585</td>\n",
       "      <td>6.631651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>meta</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Candidatus Micrarchaeota</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>...</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>2919304.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-16.205998</td>\n",
       "      <td>46.712373</td>\n",
       "      <td>257.0</td>\n",
       "      <td>3.096265</td>\n",
       "      <td>Obsidian6</td>\n",
       "      <td>8.535439</td>\n",
       "      <td>4.909309</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CRISPRS CRISPRS NoCOG assembly  chip chip counts  cluster#  \\\n",
       "238       0.0     NaN   NaN     mega   5.0         NaN      12.0   \n",
       "271       0.0     NaN   NaN     meta   3.0         NaN      11.0   \n",
       "\n",
       "    dominant domain           dominant phylum dominant species    ...     \\\n",
       "238         Archaea  Candidatus Micrarchaeota     unclassified    ...      \n",
       "271         Archaea  Candidatus Micrarchaeota     unclassified    ...      \n",
       "\n",
       "     highest_lineage_name     length majorCOG       mu_x       mu_y  \\\n",
       "238                         268227.0      1.0 -24.559164  46.653121   \n",
       "271          unclassified  2919304.0      0.0 -16.205998  46.712373   \n",
       "\n",
       "     num_contigs  number of cells     sample     sig_x     sig_y  \n",
       "238         40.0         5.270336  Obsidian6  3.893585  6.631651  \n",
       "271        257.0         3.096265  Obsidian6  8.535439  4.909309  \n",
       "\n",
       "[2 rows x 23 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusterRes_df[clusterRes_df['dominant phylum'].str.contains('Micrarchaeo')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D-module</th>\n",
       "      <th>D</th>\n",
       "      <th>E-module</th>\n",
       "      <th>E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K00855</td>\n",
       "      <td>PRK, prkB; phosphoribulokinase [EC:2.7.1.19]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K01601</td>\n",
       "      <td>rbcL; ribulose-bisphosphate carboxylase large ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K01602</td>\n",
       "      <td>rbcS; ribulose-bisphosphate carboxylase small ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K00927</td>\n",
       "      <td>PGK, pgk; phosphoglycerate kinase [EC:2.7.2.3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K05298</td>\n",
       "      <td>GAPA; glyceraldehyde-3-phosphate dehydrogenase...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K00150</td>\n",
       "      <td>gap2; glyceraldehyde-3-phosphate dehydrogenase...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K00134</td>\n",
       "      <td>GAPDH, gapA; glyceraldehyde 3-phosphate dehydr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K01623</td>\n",
       "      <td>ALDO; fructose-bisphosphate aldolase, class I ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K01624</td>\n",
       "      <td>FBA, fbaA; fructose-bisphosphate aldolase, cla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K03841</td>\n",
       "      <td>FBP, fbp; fructose-1,6-bisphosphatase I [EC:3....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K02446</td>\n",
       "      <td>glpX; fructose-1,6-bisphosphatase II [EC:3.1.3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K11532</td>\n",
       "      <td>glpX-SEBP; fructose-1,6-bisphosphatase II / se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K01086</td>\n",
       "      <td>fbp-SEBP; fructose-1,6-bisphosphatase I / sedo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K04041</td>\n",
       "      <td>fbp3; fructose-1,6-bisphosphatase III [EC:3.1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K00615</td>\n",
       "      <td>E2.2.1.1, tktA, tktB; transketolase [EC:2.2.1.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K01100</td>\n",
       "      <td>E3.1.3.37; sedoheptulose-bisphosphatase [EC:3....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K01807</td>\n",
       "      <td>rpiA; ribose 5-phosphate isomerase A [EC:5.3.1.6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00165</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycl...</td>\n",
       "      <td>K01808</td>\n",
       "      <td>rpiB; ribose 5-phosphate isomerase B [EC:5.3.1.6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00166</td>\n",
       "      <td>Reductive pentose phosphate cycle, ribulose-5P...</td>\n",
       "      <td>K00855</td>\n",
       "      <td>PRK, prkB; phosphoribulokinase [EC:2.7.1.19]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00166</td>\n",
       "      <td>Reductive pentose phosphate cycle, ribulose-5P...</td>\n",
       "      <td>K01601</td>\n",
       "      <td>rbcL; ribulose-bisphosphate carboxylase large ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00166</td>\n",
       "      <td>Reductive pentose phosphate cycle, ribulose-5P...</td>\n",
       "      <td>K01602</td>\n",
       "      <td>rbcS; ribulose-bisphosphate carboxylase small ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00166</td>\n",
       "      <td>Reductive pentose phosphate cycle, ribulose-5P...</td>\n",
       "      <td>K00927</td>\n",
       "      <td>PGK, pgk; phosphoglycerate kinase [EC:2.7.2.3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00166</td>\n",
       "      <td>Reductive pentose phosphate cycle, ribulose-5P...</td>\n",
       "      <td>K05298</td>\n",
       "      <td>GAPA; glyceraldehyde-3-phosphate dehydrogenase...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00166</td>\n",
       "      <td>Reductive pentose phosphate cycle, ribulose-5P...</td>\n",
       "      <td>K00150</td>\n",
       "      <td>gap2; glyceraldehyde-3-phosphate dehydrogenase...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00166</td>\n",
       "      <td>Reductive pentose phosphate cycle, ribulose-5P...</td>\n",
       "      <td>K00134</td>\n",
       "      <td>GAPDH, gapA; glyceraldehyde 3-phosphate dehydr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00167</td>\n",
       "      <td>Reductive pentose phosphate cycle, glyceraldeh...</td>\n",
       "      <td>K01623</td>\n",
       "      <td>ALDO; fructose-bisphosphate aldolase, class I ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00167</td>\n",
       "      <td>Reductive pentose phosphate cycle, glyceraldeh...</td>\n",
       "      <td>K01624</td>\n",
       "      <td>FBA, fbaA; fructose-bisphosphate aldolase, cla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00167</td>\n",
       "      <td>Reductive pentose phosphate cycle, glyceraldeh...</td>\n",
       "      <td>K03841</td>\n",
       "      <td>FBP, fbp; fructose-1,6-bisphosphatase I [EC:3....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00167</td>\n",
       "      <td>Reductive pentose phosphate cycle, glyceraldeh...</td>\n",
       "      <td>K02446</td>\n",
       "      <td>glpX; fructose-1,6-bisphosphatase II [EC:3.1.3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Pathway module</td>\n",
       "      <td>Energy metabolism</td>\n",
       "      <td>Carbon fixation</td>\n",
       "      <td>M00167</td>\n",
       "      <td>Reductive pentose phosphate cycle, glyceraldeh...</td>\n",
       "      <td>K11532</td>\n",
       "      <td>glpX-SEBP; fructose-1,6-bisphosphatase II / se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4903</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00649</td>\n",
       "      <td>Multidrug resistance, efflux pump AdeABC [PATH...</td>\n",
       "      <td>K18144</td>\n",
       "      <td>adeR; two-component system, OmpR family, respo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4904</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00649</td>\n",
       "      <td>Multidrug resistance, efflux pump AdeABC [PATH...</td>\n",
       "      <td>K18145</td>\n",
       "      <td>adeA; membrane fusion protein, multidrug efflu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4905</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00649</td>\n",
       "      <td>Multidrug resistance, efflux pump AdeABC [PATH...</td>\n",
       "      <td>K18146</td>\n",
       "      <td>adeB; multidrug efflux pump</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4906</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00649</td>\n",
       "      <td>Multidrug resistance, efflux pump AdeABC [PATH...</td>\n",
       "      <td>K18147</td>\n",
       "      <td>adeC; outer membrane protein, multidrug efflux...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4907</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00696</td>\n",
       "      <td>Multidrug resistance, efflux pump AcrEF-TolC</td>\n",
       "      <td>K18140</td>\n",
       "      <td>envR, acrS; TetR/AcrR family transcriptional r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4908</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00696</td>\n",
       "      <td>Multidrug resistance, efflux pump AcrEF-TolC</td>\n",
       "      <td>K18141</td>\n",
       "      <td>acrE; membrane fusion protein, multidrug efflu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4909</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00696</td>\n",
       "      <td>Multidrug resistance, efflux pump AcrEF-TolC</td>\n",
       "      <td>K18142</td>\n",
       "      <td>acrF; multidrug efflux pump</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4910</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00696</td>\n",
       "      <td>Multidrug resistance, efflux pump AcrEF-TolC</td>\n",
       "      <td>K12340</td>\n",
       "      <td>tolC; outer membrane protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4911</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00697</td>\n",
       "      <td>Multidrug resistance, efflux pump MdtEF-TolC</td>\n",
       "      <td>K07690</td>\n",
       "      <td>evgA, bvgA; two-component system, NarL family,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4912</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00697</td>\n",
       "      <td>Multidrug resistance, efflux pump MdtEF-TolC</td>\n",
       "      <td>K18898</td>\n",
       "      <td>mdtE; membrane fusion protein, multidrug efflu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4913</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00697</td>\n",
       "      <td>Multidrug resistance, efflux pump MdtEF-TolC</td>\n",
       "      <td>K18899</td>\n",
       "      <td>mdtF; multidrug efflux pump</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4914</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00697</td>\n",
       "      <td>Multidrug resistance, efflux pump MdtEF-TolC</td>\n",
       "      <td>K12340</td>\n",
       "      <td>tolC; outer membrane protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4915</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00698</td>\n",
       "      <td>Multidrug resistance, efflux pump BpeEF-OprC</td>\n",
       "      <td>K18900</td>\n",
       "      <td>bpeT; LysR family transcriptional regulator, r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4916</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00698</td>\n",
       "      <td>Multidrug resistance, efflux pump BpeEF-OprC</td>\n",
       "      <td>K18901</td>\n",
       "      <td>bpeE; membrane fusion protein, multidrug efflu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4917</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00698</td>\n",
       "      <td>Multidrug resistance, efflux pump BpeEF-OprC</td>\n",
       "      <td>K18902</td>\n",
       "      <td>bpeF; multidrug efflux pump</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4918</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00698</td>\n",
       "      <td>Multidrug resistance, efflux pump BpeEF-OprC</td>\n",
       "      <td>K18903</td>\n",
       "      <td>oprC, opcM; outer membrane protein, multidrug ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00700</td>\n",
       "      <td>Multidrug resistance, efflux pump AbcA</td>\n",
       "      <td>K18906</td>\n",
       "      <td>mgrA; MarR family transcriptional regulator, m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4920</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00700</td>\n",
       "      <td>Multidrug resistance, efflux pump AbcA</td>\n",
       "      <td>K18907</td>\n",
       "      <td>norG; GntR family transcriptional regulator, r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4921</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00700</td>\n",
       "      <td>Multidrug resistance, efflux pump AbcA</td>\n",
       "      <td>K18104</td>\n",
       "      <td>abcA, bmrA; ATP-binding cassette, subfamily B,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4922</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00702</td>\n",
       "      <td>Multidrug resistance, efflux pump NorB</td>\n",
       "      <td>K18906</td>\n",
       "      <td>mgrA; MarR family transcriptional regulator, m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4923</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00702</td>\n",
       "      <td>Multidrug resistance, efflux pump NorB</td>\n",
       "      <td>K18907</td>\n",
       "      <td>norG; GntR family transcriptional regulator, r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4924</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00702</td>\n",
       "      <td>Multidrug resistance, efflux pump NorB</td>\n",
       "      <td>K08170</td>\n",
       "      <td>norB, norC; MFS transporter, DHA2 family, mult...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4925</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00714</td>\n",
       "      <td>Multidrug resistance, efflux pump QacA</td>\n",
       "      <td>K18938</td>\n",
       "      <td>qacR; TetR/AcrR family transcriptional regulat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4926</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00714</td>\n",
       "      <td>Multidrug resistance, efflux pump QacA</td>\n",
       "      <td>K08167</td>\n",
       "      <td>smvA, qacA, lfrA; MFS transporter, DHA2 family...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4927</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00705</td>\n",
       "      <td>Multidrug resistance, efflux pump MepA</td>\n",
       "      <td>K18909</td>\n",
       "      <td>mepR; MarR family transcriptional regulator, r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4928</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00705</td>\n",
       "      <td>Multidrug resistance, efflux pump MepA</td>\n",
       "      <td>K18908</td>\n",
       "      <td>mepA; multidrug efflux pump</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4929</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00746</td>\n",
       "      <td>Multidrug resistance, repression of porin OmpF</td>\n",
       "      <td>K13632</td>\n",
       "      <td>marA; AraC family transcriptional regulator, m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4930</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00746</td>\n",
       "      <td>Multidrug resistance, repression of porin OmpF</td>\n",
       "      <td>K18513</td>\n",
       "      <td>micF; small regulatory RNA MicF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4931</th>\n",
       "      <td>Signature module</td>\n",
       "      <td>Gene set</td>\n",
       "      <td>Drug resistance</td>\n",
       "      <td>M00746</td>\n",
       "      <td>Multidrug resistance, repression of porin OmpF</td>\n",
       "      <td>K09476</td>\n",
       "      <td>ompF; outer membrane pore protein F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4932</th>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4933 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     A                  B                C D-module  \\\n",
       "0       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "1       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "2       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "3       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "4       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "5       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "6       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "7       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "8       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "9       Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "10      Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "11      Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "12      Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "13      Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "14      Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "15      Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "16      Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "17      Pathway module  Energy metabolism  Carbon fixation   M00165   \n",
       "18      Pathway module  Energy metabolism  Carbon fixation   M00166   \n",
       "19      Pathway module  Energy metabolism  Carbon fixation   M00166   \n",
       "20      Pathway module  Energy metabolism  Carbon fixation   M00166   \n",
       "21      Pathway module  Energy metabolism  Carbon fixation   M00166   \n",
       "22      Pathway module  Energy metabolism  Carbon fixation   M00166   \n",
       "23      Pathway module  Energy metabolism  Carbon fixation   M00166   \n",
       "24      Pathway module  Energy metabolism  Carbon fixation   M00166   \n",
       "25      Pathway module  Energy metabolism  Carbon fixation   M00167   \n",
       "26      Pathway module  Energy metabolism  Carbon fixation   M00167   \n",
       "27      Pathway module  Energy metabolism  Carbon fixation   M00167   \n",
       "28      Pathway module  Energy metabolism  Carbon fixation   M00167   \n",
       "29      Pathway module  Energy metabolism  Carbon fixation   M00167   \n",
       "...                ...                ...              ...      ...   \n",
       "4903  Signature module           Gene set  Drug resistance   M00649   \n",
       "4904  Signature module           Gene set  Drug resistance   M00649   \n",
       "4905  Signature module           Gene set  Drug resistance   M00649   \n",
       "4906  Signature module           Gene set  Drug resistance   M00649   \n",
       "4907  Signature module           Gene set  Drug resistance   M00696   \n",
       "4908  Signature module           Gene set  Drug resistance   M00696   \n",
       "4909  Signature module           Gene set  Drug resistance   M00696   \n",
       "4910  Signature module           Gene set  Drug resistance   M00696   \n",
       "4911  Signature module           Gene set  Drug resistance   M00697   \n",
       "4912  Signature module           Gene set  Drug resistance   M00697   \n",
       "4913  Signature module           Gene set  Drug resistance   M00697   \n",
       "4914  Signature module           Gene set  Drug resistance   M00697   \n",
       "4915  Signature module           Gene set  Drug resistance   M00698   \n",
       "4916  Signature module           Gene set  Drug resistance   M00698   \n",
       "4917  Signature module           Gene set  Drug resistance   M00698   \n",
       "4918  Signature module           Gene set  Drug resistance   M00698   \n",
       "4919  Signature module           Gene set  Drug resistance   M00700   \n",
       "4920  Signature module           Gene set  Drug resistance   M00700   \n",
       "4921  Signature module           Gene set  Drug resistance   M00700   \n",
       "4922  Signature module           Gene set  Drug resistance   M00702   \n",
       "4923  Signature module           Gene set  Drug resistance   M00702   \n",
       "4924  Signature module           Gene set  Drug resistance   M00702   \n",
       "4925  Signature module           Gene set  Drug resistance   M00714   \n",
       "4926  Signature module           Gene set  Drug resistance   M00714   \n",
       "4927  Signature module           Gene set  Drug resistance   M00705   \n",
       "4928  Signature module           Gene set  Drug resistance   M00705   \n",
       "4929  Signature module           Gene set  Drug resistance   M00746   \n",
       "4930  Signature module           Gene set  Drug resistance   M00746   \n",
       "4931  Signature module           Gene set  Drug resistance   M00746   \n",
       "4932               nan                nan              nan      nan   \n",
       "\n",
       "                                                      D E-module  \\\n",
       "0     Reductive pentose phosphate cycle (Calvin cycl...   K00855   \n",
       "1     Reductive pentose phosphate cycle (Calvin cycl...   K01601   \n",
       "2     Reductive pentose phosphate cycle (Calvin cycl...   K01602   \n",
       "3     Reductive pentose phosphate cycle (Calvin cycl...   K00927   \n",
       "4     Reductive pentose phosphate cycle (Calvin cycl...   K05298   \n",
       "5     Reductive pentose phosphate cycle (Calvin cycl...   K00150   \n",
       "6     Reductive pentose phosphate cycle (Calvin cycl...   K00134   \n",
       "7     Reductive pentose phosphate cycle (Calvin cycl...   K01623   \n",
       "8     Reductive pentose phosphate cycle (Calvin cycl...   K01624   \n",
       "9     Reductive pentose phosphate cycle (Calvin cycl...   K03841   \n",
       "10    Reductive pentose phosphate cycle (Calvin cycl...   K02446   \n",
       "11    Reductive pentose phosphate cycle (Calvin cycl...   K11532   \n",
       "12    Reductive pentose phosphate cycle (Calvin cycl...   K01086   \n",
       "13    Reductive pentose phosphate cycle (Calvin cycl...   K04041   \n",
       "14    Reductive pentose phosphate cycle (Calvin cycl...   K00615   \n",
       "15    Reductive pentose phosphate cycle (Calvin cycl...   K01100   \n",
       "16    Reductive pentose phosphate cycle (Calvin cycl...   K01807   \n",
       "17    Reductive pentose phosphate cycle (Calvin cycl...   K01808   \n",
       "18    Reductive pentose phosphate cycle, ribulose-5P...   K00855   \n",
       "19    Reductive pentose phosphate cycle, ribulose-5P...   K01601   \n",
       "20    Reductive pentose phosphate cycle, ribulose-5P...   K01602   \n",
       "21    Reductive pentose phosphate cycle, ribulose-5P...   K00927   \n",
       "22    Reductive pentose phosphate cycle, ribulose-5P...   K05298   \n",
       "23    Reductive pentose phosphate cycle, ribulose-5P...   K00150   \n",
       "24    Reductive pentose phosphate cycle, ribulose-5P...   K00134   \n",
       "25    Reductive pentose phosphate cycle, glyceraldeh...   K01623   \n",
       "26    Reductive pentose phosphate cycle, glyceraldeh...   K01624   \n",
       "27    Reductive pentose phosphate cycle, glyceraldeh...   K03841   \n",
       "28    Reductive pentose phosphate cycle, glyceraldeh...   K02446   \n",
       "29    Reductive pentose phosphate cycle, glyceraldeh...   K11532   \n",
       "...                                                 ...      ...   \n",
       "4903  Multidrug resistance, efflux pump AdeABC [PATH...   K18144   \n",
       "4904  Multidrug resistance, efflux pump AdeABC [PATH...   K18145   \n",
       "4905  Multidrug resistance, efflux pump AdeABC [PATH...   K18146   \n",
       "4906  Multidrug resistance, efflux pump AdeABC [PATH...   K18147   \n",
       "4907       Multidrug resistance, efflux pump AcrEF-TolC   K18140   \n",
       "4908       Multidrug resistance, efflux pump AcrEF-TolC   K18141   \n",
       "4909       Multidrug resistance, efflux pump AcrEF-TolC   K18142   \n",
       "4910       Multidrug resistance, efflux pump AcrEF-TolC   K12340   \n",
       "4911       Multidrug resistance, efflux pump MdtEF-TolC   K07690   \n",
       "4912       Multidrug resistance, efflux pump MdtEF-TolC   K18898   \n",
       "4913       Multidrug resistance, efflux pump MdtEF-TolC   K18899   \n",
       "4914       Multidrug resistance, efflux pump MdtEF-TolC   K12340   \n",
       "4915       Multidrug resistance, efflux pump BpeEF-OprC   K18900   \n",
       "4916       Multidrug resistance, efflux pump BpeEF-OprC   K18901   \n",
       "4917       Multidrug resistance, efflux pump BpeEF-OprC   K18902   \n",
       "4918       Multidrug resistance, efflux pump BpeEF-OprC   K18903   \n",
       "4919             Multidrug resistance, efflux pump AbcA   K18906   \n",
       "4920             Multidrug resistance, efflux pump AbcA   K18907   \n",
       "4921             Multidrug resistance, efflux pump AbcA   K18104   \n",
       "4922             Multidrug resistance, efflux pump NorB   K18906   \n",
       "4923             Multidrug resistance, efflux pump NorB   K18907   \n",
       "4924             Multidrug resistance, efflux pump NorB   K08170   \n",
       "4925             Multidrug resistance, efflux pump QacA   K18938   \n",
       "4926             Multidrug resistance, efflux pump QacA   K08167   \n",
       "4927             Multidrug resistance, efflux pump MepA   K18909   \n",
       "4928             Multidrug resistance, efflux pump MepA   K18908   \n",
       "4929     Multidrug resistance, repression of porin OmpF   K13632   \n",
       "4930     Multidrug resistance, repression of porin OmpF   K18513   \n",
       "4931     Multidrug resistance, repression of porin OmpF   K09476   \n",
       "4932                                                nan      nan   \n",
       "\n",
       "                                                      E  \n",
       "0          PRK, prkB; phosphoribulokinase [EC:2.7.1.19]  \n",
       "1     rbcL; ribulose-bisphosphate carboxylase large ...  \n",
       "2     rbcS; ribulose-bisphosphate carboxylase small ...  \n",
       "3        PGK, pgk; phosphoglycerate kinase [EC:2.7.2.3]  \n",
       "4     GAPA; glyceraldehyde-3-phosphate dehydrogenase...  \n",
       "5     gap2; glyceraldehyde-3-phosphate dehydrogenase...  \n",
       "6     GAPDH, gapA; glyceraldehyde 3-phosphate dehydr...  \n",
       "7     ALDO; fructose-bisphosphate aldolase, class I ...  \n",
       "8     FBA, fbaA; fructose-bisphosphate aldolase, cla...  \n",
       "9     FBP, fbp; fructose-1,6-bisphosphatase I [EC:3....  \n",
       "10    glpX; fructose-1,6-bisphosphatase II [EC:3.1.3...  \n",
       "11    glpX-SEBP; fructose-1,6-bisphosphatase II / se...  \n",
       "12    fbp-SEBP; fructose-1,6-bisphosphatase I / sedo...  \n",
       "13    fbp3; fructose-1,6-bisphosphatase III [EC:3.1....  \n",
       "14     E2.2.1.1, tktA, tktB; transketolase [EC:2.2.1.1]  \n",
       "15    E3.1.3.37; sedoheptulose-bisphosphatase [EC:3....  \n",
       "16    rpiA; ribose 5-phosphate isomerase A [EC:5.3.1.6]  \n",
       "17    rpiB; ribose 5-phosphate isomerase B [EC:5.3.1.6]  \n",
       "18         PRK, prkB; phosphoribulokinase [EC:2.7.1.19]  \n",
       "19    rbcL; ribulose-bisphosphate carboxylase large ...  \n",
       "20    rbcS; ribulose-bisphosphate carboxylase small ...  \n",
       "21       PGK, pgk; phosphoglycerate kinase [EC:2.7.2.3]  \n",
       "22    GAPA; glyceraldehyde-3-phosphate dehydrogenase...  \n",
       "23    gap2; glyceraldehyde-3-phosphate dehydrogenase...  \n",
       "24    GAPDH, gapA; glyceraldehyde 3-phosphate dehydr...  \n",
       "25    ALDO; fructose-bisphosphate aldolase, class I ...  \n",
       "26    FBA, fbaA; fructose-bisphosphate aldolase, cla...  \n",
       "27    FBP, fbp; fructose-1,6-bisphosphatase I [EC:3....  \n",
       "28    glpX; fructose-1,6-bisphosphatase II [EC:3.1.3...  \n",
       "29    glpX-SEBP; fructose-1,6-bisphosphatase II / se...  \n",
       "...                                                 ...  \n",
       "4903  adeR; two-component system, OmpR family, respo...  \n",
       "4904  adeA; membrane fusion protein, multidrug efflu...  \n",
       "4905                        adeB; multidrug efflux pump  \n",
       "4906  adeC; outer membrane protein, multidrug efflux...  \n",
       "4907  envR, acrS; TetR/AcrR family transcriptional r...  \n",
       "4908  acrE; membrane fusion protein, multidrug efflu...  \n",
       "4909                        acrF; multidrug efflux pump  \n",
       "4910                       tolC; outer membrane protein  \n",
       "4911  evgA, bvgA; two-component system, NarL family,...  \n",
       "4912  mdtE; membrane fusion protein, multidrug efflu...  \n",
       "4913                        mdtF; multidrug efflux pump  \n",
       "4914                       tolC; outer membrane protein  \n",
       "4915  bpeT; LysR family transcriptional regulator, r...  \n",
       "4916  bpeE; membrane fusion protein, multidrug efflu...  \n",
       "4917                        bpeF; multidrug efflux pump  \n",
       "4918  oprC, opcM; outer membrane protein, multidrug ...  \n",
       "4919  mgrA; MarR family transcriptional regulator, m...  \n",
       "4920  norG; GntR family transcriptional regulator, r...  \n",
       "4921  abcA, bmrA; ATP-binding cassette, subfamily B,...  \n",
       "4922  mgrA; MarR family transcriptional regulator, m...  \n",
       "4923  norG; GntR family transcriptional regulator, r...  \n",
       "4924  norB, norC; MFS transporter, DHA2 family, mult...  \n",
       "4925  qacR; TetR/AcrR family transcriptional regulat...  \n",
       "4926  smvA, qacA, lfrA; MFS transporter, DHA2 family...  \n",
       "4927  mepR; MarR family transcriptional regulator, r...  \n",
       "4928                        mepA; multidrug efflux pump  \n",
       "4929  marA; AraC family transcriptional regulator, m...  \n",
       "4930                    micF; small regulatory RNA MicF  \n",
       "4931                ompF; outer membrane pore protein F  \n",
       "4932                                                nan  \n",
       "\n",
       "[4933 rows x 7 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KEGGmoddf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
