{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import csv\n",
    "import matplotlib.cm as cm\n",
    "import seaborn as sns\n",
    "from matplotlib import gridspec\n",
    "import matplotlib.animation as manimation\n",
    "import time\n",
    "import scipy.io as sio\n",
    "from random import shuffle\n",
    "import time\n",
    "from collections import Counter\n",
    "import sklearn.cluster as cluster\n",
    "import time\n",
    "import hdbscan\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "import os\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from pathlib import Path\n",
    "\n",
    "#import plotly. as py\n",
    "#import plotly.graph_objs as go\n",
    "font = {'family' : 'DejaVu Sans',\n",
    "        'weight' : 'regular',\n",
    "        'size'   : 16}\n",
    "\n",
    "plt.rc('font', **font)\n",
    "plt.rc('lines',lw=2)\n",
    "\n",
    "# function for optimal leaf ordering\n",
    "def _optimal_order(data, **kwargs):\n",
    "    \"\"\" Optimal leaf ordering\n",
    "        **kwargs passed to pdist e.g. metric='correlation'\n",
    "    \"\"\"\n",
    "    d = pdist(data, **kwargs)\n",
    "    link = linkage(d, method='average')\n",
    "    optimal_order = optimal_leaf_ordering(link, d)\n",
    "    return optimal_order\n",
    "\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "#from fastcluster import linkage\n",
    "#from polo import optimal_leaf_ordering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/bojk/miniconda3/envs/py360/lib/python3.6/site-packages/ipykernel_launcher.py:16: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  app.launch_new_instance()\n"
     ]
    }
   ],
   "source": [
    "mypath = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'\n",
    "foldernames = [f for f in listdir(mypath) if ('Knumbers') not in f and ('NN_MAGs') not in f]\n",
    "\n",
    "foldername = pd.DataFrame(index=range(len(foldernames)),columns=['folder_name','number','genome'])\n",
    "foldername['folder_name'] = [f for f in foldernames]\n",
    "foldername['number'] = [f[-10:] for f in foldernames]\n",
    "foldername['genome'] = [f[0:15] for f in foldernames]\n",
    "ktab = pd.DataFrame(index=range(10000),columns=list(foldername['genome']))\n",
    "ctab = pd.DataFrame(index=range(10000),columns=list(foldername['genome']))\n",
    "ctab_thr = pd.DataFrame(index=range(10000),columns=list(foldername['genome'])) # threshold hit value\n",
    "\n",
    "# make KO-list table and write to txt file in Knumbers folder\n",
    "for i in range(0,len(foldername)):\n",
    "    kofile = mypath+foldername.loc[i,'folder_name']+'/IMG_Data/'+foldername.loc[i,'number']+'/'+foldername.loc[i,'number']+'.ko.tab.txt'        \n",
    "    ktab[foldername.loc[i,'genome']] = pd.read_csv(kofile,'\\t')['ko_id'].str[3:]\n",
    "    ktab[foldername.loc[i,'genome']].to_csv(mypath+'Knumbers/'+foldername.loc[i,'genome']+'.txt',sep='\\t')\n",
    "# make COG-list table \n",
    "for i in range(0,len(foldername)):\n",
    "    cofile = mypath+foldername.loc[i,'folder_name']+'/IMG_Data/'+foldername.loc[i,'number']+'/'+foldername.loc[i,'number']+'.cog.tab.txt'        \n",
    "    ctab[foldername.loc[i,'genome']] = pd.read_csv(cofile,'\\t')['cog_id']\n",
    "    ctab_thr[foldername.loc[i,'genome']] = pd.read_csv(cofile,'\\t')['percent_identity']\n",
    "    #ctab[foldername.loc[i,'genome']].to_csv(mypath+'Knumbers/'+foldername.loc[i,'genome']+'.txt',sep='\\t')\n",
    "    \n",
    "# COG dfs with different percent identity threshold values \n",
    "ctab_30 = ctab[ctab_thr>30]\n",
    "ctab_40 = ctab[ctab_thr>40]\n",
    "ctab_50 = ctab[ctab_thr>50]\n",
    "ctab_60 = ctab[ctab_thr>60]\n",
    "ctab_70 = ctab[ctab_thr>70]\n",
    "ctab_80 = ctab[ctab_thr>80]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/bojk/miniconda3/envs/py360/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\\t'.\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] File b'COGreffiles/fun2003-2014.tab' does not exist: b'COGreffiles/fun2003-2014.tab'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-230959971625>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mCOGcat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'COGreffiles/fun2003-2014.tab'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mCOGtab\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'COGreffiles/cognames2003-2014.tab'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mCOGtab\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'NameNum'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCOGtab\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m' - '\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mCOGtab\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'COG'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mCOGtab\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'FuncCat'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mCOGcat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mCOGcat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'# Code'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mCOGtab\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'func'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mCOGtab\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFuncCat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue_counts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py360/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    695\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    696\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 697\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    698\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    699\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py360/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    422\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    423\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 424\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    425\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    426\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py360/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    888\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    889\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 890\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    891\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    892\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py360/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1115\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1116\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'c'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1117\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1118\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1119\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'python'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/py360/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1846\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'usecols'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1847\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1848\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1849\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1850\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] File b'COGreffiles/fun2003-2014.tab' does not exist: b'COGreffiles/fun2003-2014.tab'"
     ]
    }
   ],
   "source": [
    "COGcat = pd.read_table('COGreffiles/fun2003-2014.tab')\n",
    "COGtab = pd.read_table('COGreffiles/cognames2003-2014.tab')\n",
    "COGtab['NameNum'] = COGtab['name']+' - '+COGtab['COG']\n",
    "COGtab['FuncCat'] = [COGcat[COGcat['# Code']==i[0]]['Name'].values[0] for i in COGtab['func']]\n",
    "COGtab.FuncCat.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def COGmapper(df):\n",
    "    # gets ctab formatted df into mappable one\n",
    "    mapdf = pd.DataFrame(index=COGtab.set_index('COG').index,columns=df.columns)\n",
    "    for names in df.columns:\n",
    "        mapdf.loc[df[names].dropna(how='any').values,names] = 1\n",
    "    return(mapdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'COGtab' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-6a70da4aa70d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcmap_30\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCOGmapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mctab_30\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-5-b1408b9a24fd>\u001b[0m in \u001b[0;36mCOGmapper\u001b[0;34m(df)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mCOGmapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;31m# gets ctab formatted df into mappable one\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mmapdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mCOGtab\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'COG'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mnames\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mmapdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnames\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'any'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnames\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'COGtab' is not defined"
     ]
    }
   ],
   "source": [
    "cmap_30 = COGmapper(ctab_30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cmap_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from urllib.request import Request, urlopen, URLError\n",
    "from urllib.parse import urlencode\n",
    "from lxml import etree\n",
    "import lxml\n",
    "import html2text as h2t\n",
    "from html.parser import HTMLParser\n",
    "\n",
    "class myhtmlparser(HTMLParser):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        self.reset()\n",
    "\n",
    "        self.NEWTAGS = []\n",
    "        self.NEWATTRS = []\n",
    "        self.HTMLDATA = []\n",
    "    def handle_starttag(self, tag, attrs):\n",
    "        self.NEWTAGS.append(tag)\n",
    "        self.NEWATTRS.append(attrs)\n",
    "    def handle_data(self, data):\n",
    "        self.HTMLDATA.append(data)\n",
    "    def clean(self):\n",
    "        self.NEWTAGS = []\n",
    "        self.NEWATTRS = []\n",
    "        self.HTMLDATA = []\n",
    "\n",
    "def retrieveKEGGmodules(df,typedata,main_dir):\n",
    "    #typedata can be: 'all','complete', or 'conservative'\n",
    "    if typedata == 'conservative':\n",
    "        mode = 'complete+ng1'\n",
    "    else:\n",
    "        mode = typedata\n",
    "        \n",
    "    # make dataframe\n",
    "    Mlist = pd.DataFrame(columns=df.columns)\n",
    "    # make a datafame containing the KEGG entries for known hits. This is incomplete!\n",
    "    # as I have hits not showing up in the known genomes. \n",
    "    # edit Oct 2017: just make a range of len(max(Modules in KEGG database))\n",
    "    dta = list(range(0,845))\n",
    "    dta2 = ['M0000'+str(dta[s]) for s in range(0,\n",
    "                        10)]+['M000'+str(dta[s]) for s in range(10,\n",
    "                            100)]+['M00'+str(dta[s]) for s in range(100,len(dta))]\n",
    "    #Mmap = pd.DataFrame(index=known_genomes.index,columns=Klist.columns)\n",
    "    Mmap = pd.DataFrame(index=dta2,columns=df.columns)\n",
    "    \n",
    "    \n",
    "    a=0\n",
    "    start_time = time.time()   \n",
    "    for s in df.columns:\n",
    "        \n",
    "        #get list of KEGG genes for a given cluster (tab-delimited text)\n",
    "        with open(main_dir+'Knumbers/'+s+'.txt', 'rt') as f:\n",
    "            dat = f.read()\n",
    "\n",
    "        # retrieve data from KEGG database\n",
    "        url = 'http://www.genome.jp/kegg-bin/find_module_object'\n",
    "        params = urlencode({'unclassified': dat, 'mode': mode}).encode()\n",
    "        html = urlopen(url, params).read()\n",
    "\n",
    "\n",
    "        # save as html file for reference purposes (and look at detailed output)\n",
    "        with open(main_dir+'Knumbers/html/'+s+'_'+typedata+'.html', 'wb') as f:\n",
    "            f.write(html)\n",
    "\n",
    "        # convert data to list\n",
    "        pstring = html.decode('utf8') \n",
    "        parser = myhtmlparser()\n",
    "        parser.feed(pstring)\n",
    "\n",
    "        # Extract data from parser\n",
    "        tags  = parser.NEWTAGS\n",
    "        attrs = parser.NEWATTRS\n",
    "        data  = parser.HTMLDATA\n",
    "\n",
    "        # Clean the parser\n",
    "        parser.clean()\n",
    "\n",
    "        # extract Module numbers and feed to dataframe\n",
    "        indices = [i for i, s in enumerate(data) if 'M00' in s]\n",
    "        mdata = [data[s] for s in indices]\n",
    "        mdatadf = pd.DataFrame(mdata,columns=[s])\n",
    "        \n",
    "        if a==0:\n",
    "            Mlist = mdatadf.copy()\n",
    "            a=1\n",
    "        else:\n",
    "            Mlist = Mlist.join(mdatadf,how='outer')\n",
    "        Mmap.loc[mdatadf.set_index(s).index,s] = 1\n",
    "    \n",
    "    end_time = time.time()            \n",
    "    print('Retrieving modules from KEGG database took {:.2f} s'.format(end_time - start_time)) \n",
    "    Mlist.to_pickle(main_dir+'Knumbers/module_list_'+typedata)\n",
    "    Mmap.to_pickle(main_dir+'Knumbers/module_map_'+typedata)\n",
    "    \n",
    "    return (Mlist,Mmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'retrieveKEGGmodules' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-61b580e9b612>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mmypath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mMlist\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mMmap\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mretrieveKEGGmodules\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mktab\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'conservative'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmypath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'retrieveKEGGmodules' is not defined"
     ]
    }
   ],
   "source": [
    "mypath = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'\n",
    "Mlist,Mmap = retrieveKEGGmodules(ktab,'conservative',mypath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# if mapping has been done recently before:\n",
    "mypath = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'\n",
    "typedata = 'conservative'\n",
    "Mmap = pd.read_pickle(mypath+'Knumbers/module_map_'+typedata)\n",
    "\n",
    "typedata = 'complete'\n",
    "Mmap_comp = pd.read_pickle(mypath+'Knumbers/module_map_'+typedata)#specify the modules \n",
    "moddf = pd.read_pickle('KEGG_module_list')\n",
    "KEGGmoddf = pd.read_pickle('KEGG_modules_df')\n",
    "C_type_metab = 'Carbon fixation'\n",
    "modules = KEGGmoddf[KEGGmoddf.C.str.contains(C_type_metab)]['D-module'].unique()\n",
    "moddf[moddf['Module'].isin(modules)]['Module_combined'].values\n",
    "\n",
    "Mmap_plot = Mmap_sum.loc[modules_man_carb_fix,cols]\n",
    "Mmap_plot.index = moddf.set_index('Module').loc[Mmap_plot.index]['Module_combined']\n",
    "#specify the bins\n",
    "f = plt.figure()\n",
    "f.set_figwidth(8)\n",
    "f.set_figheight(30)\n",
    "gs = gridspec.GridSpec(1,1)\n",
    "ax1 = f.add_subplot(gs[0,0])\n",
    "sns.heatmap(Mmap_plot.T,xticklabels=True,yticklabels=True,cbar=False,cmap='RdBu_r'\n",
    "            ,vmin=-1,linewidths=.1,ax=ax1)\n",
    "plt.xlabel('')\n",
    "#plt.show()\n",
    "plt.gcf().subplots_adjust(left=0.4,right=0.99,bottom=0.3,top=0.99)\n",
    "f.savefig('writeup/KEGG/carbon_fixation.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def KEGGmoduleMapper(df,KEGGhitType,path,blevelonly,metric):\n",
    "    \n",
    "    completeness = KEGGhitType\n",
    "    KEGGmoddf = pd.read_pickle('KEGG_modules_df') # this is the KEGG database file\n",
    "    # load metadata for all 111 Obsidian MAGs\n",
    "    #lookup_obsidian = pd.read_table('Tree_Obsidian/FredV2/ObsV2/metadata/ObsidianPool_UNI56_allbins_mafftBMGE_nofilteringnoclustering - Sheet1.tsv')\n",
    "    mod_df = pd.read_pickle('KEGG_module_list').set_index('Module')\n",
    "    # add CRISPR column\n",
    "    \"\"\"lookup_obsidian['CRISPR'] = ''\n",
    "    for name in lookup_obsidian['Obsidian Bins ']:\n",
    "        num = lookup_obsidian.set_index('Obsidian Bins ').loc[name,'IMG_ID'].astype(str)\n",
    "        path = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/'+name+'_IMG_'+num+'/IMG_Data/'+num+'/'+num+'.crispr.txt'\n",
    "        file = Path(path)\n",
    "        if file.is_file():\n",
    "            lookup_obsidian.loc[lookup_obsidian[lookup_obsidian['Obsidian Bins ']==name].index,'CRISPR'] = 'Yes'\n",
    "        else:\n",
    "            lookup_obsidian.loc[lookup_obsidian[lookup_obsidian['Obsidian Bins ']==name].index,'CRISPR'] = 'No'\n",
    "    \"\"\"\n",
    "    \n",
    "    # assign values per superkingdom for heatmap coloring \n",
    "    lutnum = {'Archaea':-1,'Bacteria':.5}\n",
    "    \n",
    "    df = df.dropna(how='all').dropna(how='all',axis=1)\n",
    "    df_col = df.copy()\n",
    "    bincolor = lookup_obsidian.set_index('Obsidian Bins ')['Domain'].map(lutnum)\n",
    "    df_col = df_col.multiply(bincolor.loc[df_col.T.index],axis='columns')\n",
    "    df_col.index = mod_df[mod_df.index.isin(df_col.index)]['Module_combined']\n",
    "    heatmap_startpos = 5\n",
    "    \n",
    "    \"\"\"\n",
    "    0 = Energy metabolism\n",
    "    1 = Carb and lipid metabolism\n",
    "    2 = Nucleic acid and aa metabolism\n",
    "    3 = Secondary metabolism\n",
    "    4 = Genetic info proc.\n",
    "    5 = Env. info proc.\n",
    "    6 = Metabolism\n",
    "    7 = Cellular processes\n",
    "    8 = Gene set\n",
    "    \"\"\"\n",
    "    \n",
    "    row_index,col_index = polo(df.fillna(0),metric)\n",
    "    \n",
    "    # sort by clustermap, \n",
    "    for Blevel in range(0,9):\n",
    "        \n",
    "        if blevelonly=='Yes': \n",
    "            idxblevdf = pd.read_pickle('Obs_FINALMAGS_indexblevel_'+completeness+'_'+str(Blevel))\n",
    "            dfB = df.loc[mod_df.loc[idxblevdf['column'].values,:].index,:]\n",
    "            dfB = dfB.dropna(how='all').dropna(how='all',axis=1).fillna(0)\n",
    "            row_index,col_index = polo(dfB,metric)\n",
    "            df_col = df_col.loc[mod_df.loc[idxblevdf['column'].values,:]['Module_combined'],:]\n",
    "            df_col = df_col.dropna(how='all').dropna(how='all',axis=1).fillna(0)\n",
    "            \n",
    "        else:\n",
    "            df.fillna(0,inplace=True)\n",
    "            df_col.fillna(0,inplace=True)\n",
    "            \n",
    "        lut = {'Archaea':'r','Bacteria':'b','Unassigned':'k','Eukaryota':'c','Viruses':'m'}\n",
    "        cmap=ListedColormap([\"#e74c3c\", \"#9b59b6\", \"#2ecc71\", \"#95a5a6\", \"#34495e\",\"#3498db\"])#[red,lightblue,green,GREY,darkblue,purple]\n",
    "        \n",
    "        f = plt.figure()\n",
    "        gs = gridspec.GridSpec(72,36)\n",
    "\n",
    "        ax2 = f.add_subplot(gs[heatmap_startpos:,0:32])\n",
    "        sns.heatmap(df_col.iloc[row_index,col_index]\n",
    "                    ,ax=ax2,cbar=False,linewidth=0.5,cmap=cmap,vmin=-1,vmax=1\n",
    "                    ,yticklabels=True,xticklabels=True)\n",
    "        plt.ylabel('')\n",
    "\n",
    "        ##############################################################################\n",
    "        #################### plot hbars on right #####################################\n",
    "        ##############################################################################\n",
    "        \n",
    "        arc = df.loc[df.iloc[row_index,:].index,lookup_obsidian[lookup_obsidian['Domain']=='Archaea']['Obsidian Bins ']].T.sum()\n",
    "        bac = df.loc[df.iloc[row_index,:].index,lookup_obsidian[lookup_obsidian['Domain']=='Bacteria']['Obsidian Bins ']].T.sum()\n",
    "            \n",
    "        ax3 = f.add_subplot(gs[heatmap_startpos:,32]) #Archaea\n",
    "        arc = arc/arc.sum()\n",
    "        arc.plot.barh(ax=ax3,sharey=True, color=\"#e74c3c\")\n",
    "        ax3.grid(False)\n",
    "        plt.title('Arch')\n",
    "        plt.xticks([])\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        ax4 = f.add_subplot(gs[heatmap_startpos:,33]) #Bacteria\n",
    "        bac = bac/bac.sum()\n",
    "        bac.plot.barh(ax=ax4,sharey=True, color='#34495e')\n",
    "        ax4.grid(False)\n",
    "        plt.title('Bact')\n",
    "        plt.xticks([])\n",
    "        #ax3.set_xlabel('Module-cluster\\npresence (%)')\n",
    "        plt.gca().invert_yaxis()\n",
    "\n",
    "        ##############################################################################\n",
    "        #########################   plot Blevel bars   ###############################\n",
    "        ##############################################################################\n",
    "        # B-level assignment\n",
    "        keys = list(KEGGmoddf.B.unique());keys = keys[0:-1]\n",
    "        idxlist = df.iloc[row_index,:].index\n",
    "        D_group = KEGGmoddf.groupby('D-module').sum()\n",
    "        for i in D_group.index:\n",
    "            a = D_group.loc[i,'B']\n",
    "            for j in keys:\n",
    "                if a.find(j)!=-1:\n",
    "                    D_group.loc[i,'B'] = j\n",
    "            c = D_group.loc[i,'C']\n",
    "            for j in KEGGmoddf.C.unique():\n",
    "                if c.find(j)!=-1:\n",
    "                    D_group.loc[i,'C'] = j\n",
    "        B_list  = D_group.loc[idxlist]['B']\n",
    "        \n",
    "        \n",
    "        if blevelonly!='Yes':\n",
    "            ax11 = f.add_subplot(gs[heatmap_startpos:,35])\n",
    "            col = sns.color_palette(\"gist_earth\", len(KEGGmoddf.B.unique())-1)\n",
    "            keys = list(KEGGmoddf.B.unique());keys = keys[0:-1];values = [i for i in col]\n",
    "            lut = dict(zip(keys,values))\n",
    "            color = B_list.map(lut)\n",
    "            for x,y in lut.items():\n",
    "                plt.bar(0,0,color=y,label=x,width=1)\n",
    "            plt.legend(bbox_to_anchor=(-53, 1.15, 50, 5), loc=3,\n",
    "                       ncol=4, mode=\"expand\", borderaxespad=0.,fontsize=25)\n",
    "            plt.barh(list(range(0,len(df))),list(np.ones(len(df))),color=color,height=1)\n",
    "            plt.xticks([]);plt.yticks([])\n",
    "            plt.ylim(-.5,len(df)-.5)\n",
    "            plt.xlim(0,1)\n",
    "            ax11.set_xlabel('B', rotation=0, fontsize=30, labelpad=40)\n",
    "            plt.gca().invert_yaxis()\n",
    "\n",
    "        #C-level assignment\n",
    "        ax12 = f.add_subplot(gs[heatmap_startpos:,34])\n",
    "        keysC = KEGGmoddf[KEGGmoddf['B']==keys[Blevel]].C.unique()\n",
    "        col = sns.color_palette(\"Set3\", len(keysC));values = [i for i in col]\n",
    "        lutC = dict(zip(keysC,values))\n",
    "        C_list  = D_group.loc[idxlist]['C']\n",
    "        C_list.replace(C_list.index,idxlist)\n",
    "        color = C_list.map(lutC)\n",
    "        color = color.fillna('None')\n",
    "        plt.barh(list(range(0,len(df))),list(np.ones(len(df))),color=color,height=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.ylim(-.5,len(df)-.5)\n",
    "        plt.xlim(0,1)\n",
    "        ax12.set_xlabel('C', rotation=0, fontsize=30, labelpad=40)\n",
    "        plt.gca().invert_yaxis()\n",
    "        for x,y in lutC.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(-50, 1.02, 1., .102), loc=3,\n",
    "                   ncol=1, mode=\"expand\", borderaxespad=0.,fontsize=20)\n",
    "\n",
    "        ##############################################################################\n",
    "        ##########################   plot top bars   #################################\n",
    "        ##############################################################################\n",
    "        columns_ordered = df.T.iloc[col_index,:].index\n",
    "        \n",
    "        ax7 = f.add_subplot(gs[0,0:32]) #phylum\n",
    "        keys = list(lookup_obsidian.set_index('Obsidian Bins ').loc[columns_ordered]['Phylum '].unique());\n",
    "        col = sns.color_palette(\"cubehelix\", len(keys))\n",
    "        values = [i for i in col]\n",
    "        lut = dict(zip(keys,values))\n",
    "        lut.update({'unclassified':\"#3498db\"})\n",
    "        color = lookup_obsidian.set_index('Obsidian Bins ')['Phylum '].map(lut)\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(0., 1.02, 1., .102), loc=3,\n",
    "                   ncol=6, mode=\"expand\", borderaxespad=0.,fontsize=20)\n",
    "        plt.bar(list(range(0,len(df.T))),np.ones(len(df.T)),color=color[columns_ordered],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(df.T)-.5);plt.ylim(0,1)\n",
    "        ax7.set_ylabel('Phylum', rotation=0, fontsize=20, labelpad=90)\n",
    "\n",
    "\n",
    "        ax8 = f.add_subplot(gs[1,0:32]) #domain\n",
    "        lut = {'Archaea':\"#e74c3c\",'Bacteria':'#34495e'}#,'Unassigned':\"#3498db\",'Eukaryota':'c','Viruses':\"#9b59b6\"}\n",
    "        color = lookup_obsidian.set_index('Obsidian Bins ')['Domain'].map(lut)\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=1, borderaxespad=0.,fontsize=20)\n",
    "        plt.bar(list(range(0,len(df.T))),np.ones(len(df.T)),color=color[columns_ordered],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(df.T)-.5);plt.ylim(0,1)\n",
    "        ax8.set_ylabel('Kingdom', rotation=0, fontsize=20, labelpad=90)\n",
    "\n",
    "        ax9 = f.add_subplot(gs[2,0:32]) #GC \n",
    "        listGC = list(lookup_obsidian.set_index('Obsidian Bins ').T.loc['GC',columns_ordered])\n",
    "        sns.heatmap([listGC],ax=ax9,cbar=False,cmap=\"RdBu_r\",linewidth=0.5,vmin=0,vmax=1,xticklabels=False)\n",
    "        ax9.set_ylabel('GC-content', rotation=0, fontsize=20, labelpad=90)\n",
    "\n",
    "        ax55 = f.add_subplot(gs[3,0:32]) #CRISPR\n",
    "        lut = {'Yes':'g','No':'#CACFD2'}\n",
    "        color = lookup_obsidian.set_index('Obsidian Bins ')['CRISPR'].map(lut)\n",
    "        for x,y in lut.items():\n",
    "            plt.bar(0,0,color=y,label=x,width=1)\n",
    "        plt.legend(bbox_to_anchor=(1.01, 0.95, 1., .102), loc=2,\n",
    "                   ncol=2, borderaxespad=0.,fontsize=20)\n",
    "        plt.bar(list(range(0,len(df.T))),np.ones(len(df.T)),color=color[columns_ordered],width=1)\n",
    "        plt.xticks([]);plt.yticks([])\n",
    "        plt.xlim(-.5,len(df.T)-.5);plt.ylim(0,1)\n",
    "        ax55.set_ylabel('CRISPR', rotation=0, fontsize=20, labelpad=90)\n",
    "\n",
    "        \n",
    "        ax10 = f.add_subplot(gs[4,0:32]) #Genome size \n",
    "        listGC = list(lookup_obsidian.set_index('Obsidian Bins ').T.loc['Genome size',columns_ordered])\n",
    "        sns.heatmap([listGC],ax=ax10,cbar=False,cmap=\"RdBu_r\",linewidth=0.5,xticklabels=False)\n",
    "        ax10.set_ylabel('Genome size', rotation=0, fontsize=20, labelpad=90)\n",
    "        \n",
    "        keys = list(KEGGmoddf.B.unique());keys = keys[0:-1]\n",
    "        gs.update(wspace=.05,hspace=.05)\n",
    "        plt.gcf().subplots_adjust(left=0.3,right=0.97,bottom=0.1)\n",
    "        if blevelonly == 'Yes':\n",
    "            f.set_figheight(round(len(df)/4))\n",
    "            f.set_figwidth(40)\n",
    "            f.savefig('MAGs/Analysis/OverviewHeatmaps/Obsidian_KEGGmodule_Blevel_'+metric+'_'+completeness+'_'+keys[Blevel])\n",
    "        else:\n",
    "            f.set_figheight(60)\n",
    "            f.set_figwidth(40)\n",
    "            f.savefig('MAGs/Analysis/OverviewHeatmaps/Obsidian_KEGGmodule_'+metric+'_'+completeness+'_'+keys[Blevel])\n",
    "        #f.savefig('MAGs/Analysis/OverviewHeatmaps/Obsidian_KEGG_'+completeness+'_'+keys[Blevel]+'.pdf')\n",
    "        \n",
    "        f.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "path = '/home/datastorage/IMG_ANNOTATION_DATA/Obsidian_MAGs/Knumbers/'\n",
    "Metrics = ['braycurtis', 'canberra','chebyshev','cityblock','correlation',\n",
    "              'cosine','dice','euclidean','hamming','jaccard','kulsinski',\n",
    "             'mahalanobis','matching','minkowski','rogerstanimoto','russellrao',\n",
    "             'seuclidean','sokalmichener','sokalsneath','sqeuclidean','yule']\n",
    "                            \n",
    "\n",
    "KEGGmoduleMapper(Mmap,'conservative',path,'nope',Metrics[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
